---
title: "10X Chromium Single Cell CRISPR Screening Workflows with Voyager"
output: html_document
vignette: >
  %\VignetteIndexEntry{10xnuclei_landing}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
  
```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval = FALSE
)
```

The [Chromium Single Cell CRISPR Screening from 10x Genomics](https://www.10xgenomics.com/products/single-cell-crispr-screening) enables researchers to assess the effects of CRISPR-mediated pertubations on cellular phenotypes and gene expression. The workflow relies on harvesting cells that have been transduced with CRISPR guides and stimulated before single cell preparation and library generation. Workflows typically include 2-5 sgRNAs per gene and 100-250 cells can be recovered per guide. Currently,  the 5â€™ CRISPR screening assay is compatible with most pre-existing Cas9 guide RNA vectors while the 3' assay requires a specific capture sequence for capture of the guide. 


## Getting Started

The vignettes below provide examples of processing raw data using a workflow that includes [`seqspec`](https://github.com/IGVF/seqspec), [`gget`](https://github.com/pachterlab/gget), and [`kallisto`/`bustools`](https://www.kallistobus.tools) to generate a count matrix. We process the output of various transcriptomics technologies into a [`SpatialFeatureExperiment`](https://bioconductor.org/packages/devel/bioc/html/SpatialFeatureExperiment.html)(SFE) object for use with `Voyager`. 

| Vignette | Colab Notebook | Description |
|--------------------|:---------------:|---------------------------------|
| [Preprocess raw data](https://pachterlab.github.io/voyager/articles/create_mtx_ca.html) | [Colab Notebook](https://colab.research.google.com/github/pachterlab/voyager/blob/documentation-devel/vignettes/create_mtx_ca.ipynb) | Fetch reference data with `gget`, process raw data with `seqspec`, generate a count matrix with `kallisto-bustools`|
| [Create a `SFE` object](https://pachterlab.github.io/voyager/articles/create_sfe.html) | [Colab Notebook](https://colab.research.google.com/github/pachterlab/voyager/blob/documentation-devel/vignettes/create_sfe.ipynb) | Download data, create `SFE` object |

## Analysis Workflows

The analysis tasks include basic quality control, spatial exploratory data analysis, identification of spatially variable genes, and computation of global and local spatial statistics. Accompanying Colab notebooks are linked.

| Vignette          |   Colab Notebook   | Description                   |
  |-------------------|:------------------:|-------------------------------|
  | [10X CRISPR Basic QC](https://pachterlab.github.io/voyager/dev/articles/vig12_crispr.html) | [Colab Notebook](https://colab.research.google.com/github/pachterlab/voyager/blob/documentation-devel/vignettes/vig12_crispr.ipynb) | Basic qc and preprocessing |
  
