---
title: "10X Chromium Single Cell Multiome Processing Workflows with Voyager"
output: html_document
vignette: >
  %\VignetteIndexEntry{10xmultiome_landing}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
  
```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval = FALSE
)
```

The [Chromium Single Cell Multiome solution from 10x Genomics](https://www.10xgenomics.com/products/single-cell-multiome-atac-plus-gene-expression) allows researchers to simultaneously interrogate chromatin accessibility and gene expression at the single cell level. Many of the workflow objectives in the standalone [Single Cell ATAC](https://pachterlab.github.io/voyager/articles/10xatac_landing.html) assay are mirrored in the multiome assay. Both assays require a nuclei suspension that is incubated in a solution that includes transposase. Nuclei and fragmented DNA are partitioned into Gel Beads-in-emulsion (GEMs) where a pool of barcodes is used to index the transposed DNA of each individual nucleus. In the multiome assay, gel beads include a poly(dT) sequence that facilitates capture of  mRNA for gene expression and a spacer sequence that enables barcode attachment to transposed DNA fragments for ATAC library generation. Sequenced libraries support investigations into the connections between open chromatin peaks and gene expression.

## Getting Started

The vignettes below provide examples of processing raw data using a workflow that includes [`seqspec`](https://github.com/IGVF/seqspec), [`gget`](https://github.com/pachterlab/gget), and [`kallisto`/`bustools`](https://www.kallistobus.tools) to generate a count matrix. We process the output of various transcriptomics technologies into a [`SpatialFeatureExperiment`](https://bioconductor.org/packages/devel/bioc/html/SpatialFeatureExperiment.html)(SFE) object for use with `Voyager`. 

| Vignette | Colab Notebook | Description |
|--------------------|:---------------:|---------------------------------|
| [Preprocess raw data](https://pachterlab.github.io/voyager/articles/create_mtx_ca.html) | [Colab Notebook](https://colab.research.google.com/github/pachterlab/voyager/blob/documentation-devel/vignettes/create_mtx_ca.ipynb) | Fetch reference data with `gget`, process raw data with `seqspec`, generate a count matrix with `kallisto-bustools`|
| [Create a `SFE` object](https://pachterlab.github.io/voyager/articles/create_sfe.html) | [Colab Notebook](https://colab.research.google.com/github/pachterlab/voyager/blob/documentation-devel/vignettes/create_sfe.ipynb) | Download data, create `SFE` object |

## Analysis Workflows

The analysis tasks include basic quality control, spatial exploratory data analysis, identification of spatially variable genes, and computation of global and local spatial statistics. Accompanying Google Colab notebooks are linked.

| Vignette          |   Colab Notebook   | Description                   |
  |-------------------|:------------------:|-------------------------------|
  | [10X Multiome Basic QC](https://pachterlab.github.io/voyager/articles/vig14_10xmultiome.html) | [Colab Notebook](https://colab.research.google.com/github/pachterlab/voyager/blob/documentation-devel/vignettes/vig14_10xmultiome.ipynb) | Basic qc and preprocessing |
  
