{
 "cells": [
  {
   "cell_type": "raw",
   "id": "1ccb003d",
   "metadata": {},
   "source": [
    "---\n",
    "title: \"Visium Processing Workflows with Voyager\"\n",
    "output: html_document\n",
    "vignette: >\n",
    "  %\\VignetteIndexEntry{visium_landing}\n",
    "  %\\VignetteEngine{knitr::rmarkdown}\n",
    "  %\\VignetteEncoding{UTF-8}\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c529617",
   "metadata": {
    "tags": [
     "remove_cell"
    ]
   },
   "outputs": [],
   "source": [
    "knitr::opts_chunk$set(\n",
    "  collapse = TRUE,\n",
    "  comment = \"#>\",\n",
    "  eval = FALSE\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04870a29",
   "metadata": {},
   "source": [
    "The Visium spatial transcriptomics platform by 10X Genomics is based on [Spatial Transcriptomics (ST) technology](https://doi.org/10.1126/science.aaf2403) that was originally published in 2016. In both methods, mRNA from tissue sections is captured on spatially barcoded spots that are immobilized on a microarray slide. Following construction of a barcoded cDNA library, mRNA transcripts can be mapped to specific spots on the microarray slide and overlayed with a high-resolution image of the tissue, allowing for visualization and analysis of gene expression in a spatial context. \n",
    "\n",
    "## Getting Started\n",
    "\n",
    "#### Dowload Data and Create a `SpatialFeatureExperiment` object\n",
    "\n",
    "Several publicly available Visium datasets are available from 10X Genomics on their [website](https://www.10xgenomics.com/resources/datasets). The vignette below demonstrates how to read output from a typical Visium experiment into a [`SpatialFeatureExperiment`](https://bioconductor.org/packages/devel/bioc/html/SpatialFeatureExperiment.html) object.\n",
    "\n",
    "| Vignette | Colab Notebook | Description |\n",
    "|--------------------|:---------------:|---------------------------------|\n",
    "| [Create a `SpatialFeatureExperiment` object](https://pachterlab.github.io/voyager/articles/create_sfe.html) |                | Download Visium data, create `SFE` object, perform basic QC |\n",
    "\n",
    "## Analysis Workflows\n",
    "\n",
    "The vignettes below demonstrate workflows that can be implemented with `Voyager` using a variety of Visium datasets. The analysis tasks include basic quality control, spatial exploratory data analysis, identification of spatially variable genes, and computation of global and local spatial statistics. Accompanying Colab notebooks are linked when available.\n",
    "\n",
    "| Vignette | Colab Notebook | Description |\n",
    "|--------------------|:---------------:|---------------------------------|\n",
    "| [Basic Visium analysis with 10X example dataset](https://pachterlab.github.io/voyager/articles/visium_10x.html) |  | Perform basic QC and standard non-spatial scRNA-seq analysis, and some spatial visualization\n",
    "| [Spatial analysis with 10X example dataset](https://pachterlab.github.io/voyager/articles/visium_10x_spatial.html) | | Tissue segmentation, and Moran's I for QC metrics and gene expression\n",
    "| [Basic Visium exploratory data analysis](https://pachterlab.github.io/voyager/articles/vig1_visium_basic.html) | [Colab Notebook](https://colab.research.google.com/drive/1EPbPu71O7mvc1Hsp1VI3Jc3xI5mzJ0eM?usp=sharing) | Perform basic QC, data normalization, dimension reduction, compute Moran's I, identify (spatially) variable genes                                                       |\n",
    "| [Spatial Visium exploratory data analysis](https://pachterlab.github.io/voyager/articles/vig2_visium.html)     |                                                                                                | Perform basic QC, find spatial neighborhood graphs, compute global and local univariate spatial statistics, permutation-based significance testing, compute correlogram |"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "cell_metadata_filter": "tags,-all",
   "main_language": "R",
   "notebook_metadata_filter": "-all"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
