<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Voyager">
<title>Create a SpatialFeatureExperiment object • Voyager</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Create a SpatialFeatureExperiment object">
<meta property="og:description" content="Voyager">
<meta name="robots" content="noindex">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">Voyager</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">1.2.4</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/install.html">Install</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-technologies">Technologies</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-technologies">
    <a class="dropdown-item" href="../articles/visium_landing.html">Visium</a>
    <a class="dropdown-item" href="../articles/slideseqV2_landing.html">Slide-seq v2</a>
    <a class="dropdown-item" href="../articles/cosmx_landing.html">CosMX</a>
    <a class="dropdown-item" href="../articles/xenium_landing.html">Xenium</a>
    <a class="dropdown-item" href="../articles/merfish_landing.html">MERFISH</a>
    <a class="dropdown-item" href="../articles/seqfish_landing.html">seqFISH</a>
    <a class="dropdown-item" href="../articles/codex_landing.html">CODEX/PhenoCycler</a>
    <a class="dropdown-item" href="../articles/splitseq_landing.html">SPLiT-seq</a>
    <a class="dropdown-item" href="../articles/chromium_landing.html">10X Chromium V3</a>
    <a class="dropdown-item" href="../articles/10xnuclei_landing.html">10X Chromium Nuclei Isolation</a>
    <a class="dropdown-item" href="../articles/10xcrispr_landing.html">10X Chromium Single Cell CRISPR</a>
    <a class="dropdown-item" href="../articles/10xatac_landing.html">10X Single Cell ATAC</a>
    <a class="dropdown-item" href="../articles/10xmultiome_landing.html">10X Single Cell Multiome</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-methods">Methods</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-methods">
    <h6 class="dropdown-header" data-toc-skip>Univariate global</h6>
    <a class="dropdown-item" href="../articles/moran_landing.html">Moran's I</a>
    <a class="dropdown-item" href="../articles/geary_landing.html">Geary's C</a>
    <a class="dropdown-item" href="../articles/correlogram_landing.html">Correlogram</a>
    <a class="dropdown-item" href="../articles/variogram_landing.html">Variogram</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Univariate local</h6>
    <a class="dropdown-item" href="../articles/moranplot_landing.html">Moran scatter plot</a>
    <a class="dropdown-item" href="../articles/localmoran_landing.html">Local Moran's I</a>
    <a class="dropdown-item" href="../articles/losh_landing.html">Local spatial heteroscedasticity</a>
    <a class="dropdown-item" href="../articles/getisord_landing.html">Getis-Ord Gi*</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Bivariate</h6>
    <a class="dropdown-item" href="../articles/lee_landing.html">Lee's L</a>
    <a class="dropdown-item" href="../articles/crossvariogram_landing.html">Cross variogram</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Multivariate</h6>
    <a class="dropdown-item" href="../articles/multispati_landing.html">MULTISPATI PCA</a>
    <a class="dropdown-item" href="../articles/localc_landing.html">Multivariate local Geary's C</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../articles/sfemethod">Custom methods</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/pachterlab/voyager/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Create a SpatialFeatureExperiment object</h1>
                        <h4 data-toc-skip class="author">Lambda
Moses</h4>
            
            <h4 data-toc-skip class="date">2023-08-17</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pachterlab/voyager/blob/HEAD/vignettes/create_sfe.Rmd" class="external-link"><code>vignettes/create_sfe.Rmd</code></a></small>
      <div class="d-none name"><code>create_sfe.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/voyager" class="external-link">Voyager</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SpatialFeatureExperiment" class="external-link">SpatialFeatureExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/alexcb/rjson" class="external-link">rjson</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org" class="external-link">Matrix</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="visium-space-ranger-output">Visium Space Ranger output<a class="anchor" aria-label="anchor" href="#visium-space-ranger-output"></a>
</h2>
<p>10x Genomics Space Ranger output from a Visium experiment can be read
in a similar manner as in <code>SpatialExperiment</code>; the
<code>SpatialFeatureExperiment</code> SFE object has the
<code>spotPoly</code> column geometry for the spot polygons. If the
filtered matrix (i.e. only spots in the tissue) is read in, then a
column graph called <code>visium</code> will also be present for the
spatial neighborhood graph of the Visium spots on the tissue. The graph
is not computed if all spots are read in regardless of whether they are
on tissue.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Example from SpatialExperiment</span></span>
<span><span class="va">dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"10xVisium"</span><span class="op">)</span>, </span>
<span>  package <span class="op">=</span> <span class="st">"SpatialExperiment"</span><span class="op">)</span></span>
<span>  </span>
<span><span class="va">sample_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"section1"</span>, <span class="st">"section2"</span><span class="op">)</span></span>
<span><span class="op">(</span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir</span>, <span class="va">sample_ids</span>, <span class="st">"outs"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "/Users/runner/work/_temp/Library/SpatialExperiment/extdata/10xVisium/section1/outs"</span></span>
<span><span class="co">#&gt; [2] "/Users/runner/work/_temp/Library/SpatialExperiment/extdata/10xVisium/section2/outs"</span></span></code></pre></div>
<p>The results for each tissue capture should be in the
<code>outs</code> directory. Inside the <code>outs</code> directory
there are two directories: <code>raw_reature_bc_matrix</code> has the
unfiltered gene count matrix, and <code>spatial</code> has the spatial
information.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "raw_feature_bc_matrix" "spatial"</span></span></code></pre></div>
<p>The <a href="https://bioconductor.org/packages/release/bioc/html/DropletUtils.html" class="external-link"><code>DropletUtils</code></a>
package has a function <code><a href="https://rdrr.io/pkg/DropletUtils/man/read10xCounts.html" class="external-link">read10xCounts()</a></code> which reads the gene
count matrix. SPE reads in the spatial information, and SFE uses the
spatial information to construct Visium spot polygons and spatial
neighborhood graphs. Inside the <code>spatial</code> directory:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"spatial"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "scalefactors_json.json"    "tissue_lowres_image.png"  </span></span>
<span><span class="co">#&gt; [3] "tissue_positions_list.csv"</span></span></code></pre></div>
<p><code>tissue_lowres_image.png</code> is a low resolution image of the
tissue.</p>
<p>Inside the <code>scalefactors_json.json</code> file:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/rjson/man/fromJSON.html" class="external-link">fromJSON</a></span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"spatial"</span>, <span class="st">"scalefactors_json.json"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; $spot_diameter_fullres</span></span>
<span><span class="co">#&gt; [1] 89.44476</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $tissue_hires_scalef</span></span>
<span><span class="co">#&gt; [1] 0.1701114</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $fiducial_diameter_fullres</span></span>
<span><span class="co">#&gt; [1] 144.4877</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $tissue_lowres_scalef</span></span>
<span><span class="co">#&gt; [1] 0.05103343</span></span></code></pre></div>
<p><code>spot_diameter_fullres</code> is the diameter of each Visium
spot in the full resolution H&amp;E image in pixels.
<code>tissue_hires_scalef</code> and <code>tissue_lowres_scalef</code>
are the ratio of the size of the high resolution (but not full
resolution) and low resolution H&amp;E image to the full resolution
image. <code>fiducial_diameter_fullres</code> is the diameter of each
fiducial spot used to align the spots to the H&amp;E image in pixels in
the full resolution image.</p>
<p>The <code>tissue_positions_list.csv</code> file contains information
for the spatial coordinates of the spots and whether each spot is in
tissue as automatically detected by Space Ranger or manually annotated
in the Loupe browser. If the polygon of the tissue boundary is
available, whether from image processing or manual annotation, geometric
operations as supported by the SFE package, which is based on the
<code>sf</code> package, can be used to find which spots intersect with
the tissue and which spots are contained in the tissue. Geometric
operations can also find the polygons of the intersections between spots
and the tissue, but the results can get messy since the intersections
can have not only polygons but also points and lines.</p>
<p>Now we read in the toy data that is in the Space Ranger output
format. Since Bioconductor version 3.17 (Voyager version 1.2.0), the
image is read as a <code>SpatRaster</code> object with the <a href="https://rspatial.github.io/terra/index.html" class="external-link"><code>terra</code></a>
package, so it is not loaded into memory unless necessary. When plotting
a large image, it will be downsampled and thus not fully loaded into
memory. The unit can be set with the <code>unit</code> argument, and can
be either pixels in full resolution image or microns. The latter is
calculated from the former based on spacing between spots, which is
known to be 100 microns.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/read10xVisiumSFE.html" class="external-link">read10xVisiumSFE</a></span><span class="op">(</span><span class="va">samples</span>, dirs <span class="op">=</span> <span class="va">samples</span>, sample_id <span class="op">=</span> <span class="va">sample_ids</span>, </span>
<span>                          type <span class="op">=</span> <span class="st">"sparse"</span>, data <span class="op">=</span> <span class="st">"raw"</span>, images <span class="op">=</span> <span class="st">"lowres"</span>, </span>
<span>                          unit <span class="op">=</span> <span class="st">"full_res_image_pixel"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 50 99 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(50): ENSMUSG00000051951 ENSMUSG00000089699 ...</span></span>
<span><span class="co">#&gt;   ENSMUSG00000005886 ENSMUSG00000101476</span></span>
<span><span class="co">#&gt; rowData names(1): symbol</span></span>
<span><span class="co">#&gt; colnames(99): AAACAACGAATAGTTC-1 AAACAAGTATCTCCCA-1 ...</span></span>
<span><span class="co">#&gt;   AAAGTCGACCCTCAGT-1-1 AAAGTGCCATCAATTA-1-1</span></span>
<span><span class="co">#&gt; colData names(4): in_tissue array_row array_col sample_id</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : pxl_col_in_fullres pxl_row_in_fullres</span></span>
<span><span class="co">#&gt; imgData names(4): sample_id image_id data scaleFactor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: full_res_image_pixel</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: spotPoly (POLYGON) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; section1: </span></span>
<span><span class="co">#&gt; section2:</span></span></code></pre></div>
<p>Space Ranger output includes the gene count matrix, spot coordinates,
and spot diameter. The Space Ranger output does NOT include nuclei
segmentation or pathologist annotation of histological regions. Extra
image processing, such as with ImageJ and QuPath, are required for those
geometries.</p>
<div class="section level3">
<h3 id="vizgen-merfish-output">Vizgen MERFISH output<a class="anchor" aria-label="anchor" href="#vizgen-merfish-output"></a>
</h3>
<p>The commercialized MERFISH from Vizgen has a standard output format,
that can be read into SFE with <code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/readVizgen.html" class="external-link">readVizgen()</a></code>. Because the
cell segmentation from each field of view (FOV) has a separate HDF5 file
and a MERFISH dataset can have hundreds of FOVs, we strongly recommend
reading the MERFISH output on a server with a large number of CPU cores.
Alternatively, some but not all MERFISH datasets store cell segmentation
in a <code>parquet</code> file, which can be more easily read into R.
Here we read a toy dataset which is the first FOV from a real
dataset:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dir_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"vizgen"</span><span class="op">)</span>, </span>
<span>                       package <span class="op">=</span> <span class="st">"SpatialFeatureExperiment"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe_mer</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/readVizgen.html" class="external-link">readVizgen</a></span><span class="op">(</span><span class="va">dir_use</span>, z <span class="op">=</span> <span class="fl">0L</span>, image <span class="op">=</span> <span class="st">"PolyT"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 385 98 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(385): Comt Ldha ... Blank-36 Blank-37</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(98): 103327291694389284070574461648020091166</span></span>
<span><span class="co">#&gt;   105028411815552368766949841604861213395 ...</span></span>
<span><span class="co">#&gt;   99103300832376657987379734140330816574</span></span>
<span><span class="co">#&gt;   99471994882184799235845481075474519252</span></span>
<span><span class="co">#&gt; colData names(7): fov volume ... max_y sample_id</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : center_x center_y</span></span>
<span><span class="co">#&gt; imgData names(4): sample_id image_id data scaleFactor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: micron</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: centroids (POINT), cellSeg (POLYGON) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01:</span></span></code></pre></div>
<p>The unit is always in microns.</p>
</div>
</div>
<div class="section level2">
<h2 id="create-sfe-object-from-scratch">Create SFE object from scratch<a class="anchor" aria-label="anchor" href="#create-sfe-object-from-scratch"></a>
</h2>
<p>An SFE object can be constructed from scratch with the assay matrices
and metadata. In this toy example, <code>dgCMatrix</code> is used, but
since SFE inherits from SingleCellExperiment (SCE), other types of
arrays supported by SCE such as delayed arrays should also work.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visium barcode location from Space Ranger</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"visium_row_col"</span><span class="op">)</span></span>
<span><span class="va">coords1</span> <span class="op">&lt;-</span> <span class="va">visium_row_col</span><span class="op">[</span><span class="va">visium_row_col</span><span class="op">$</span><span class="va">col</span> <span class="op">&lt;</span> <span class="fl">6</span> <span class="op">&amp;</span> <span class="va">visium_row_col</span><span class="op">$</span><span class="va">row</span> <span class="op">&lt;</span> <span class="fl">6</span>,<span class="op">]</span></span>
<span><span class="va">coords1</span><span class="op">$</span><span class="va">row</span> <span class="op">&lt;-</span> <span class="va">coords1</span><span class="op">$</span><span class="va">row</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Random toy sparse matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">29</span><span class="op">)</span></span>
<span><span class="va">col_inds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">13</span>, <span class="fl">13</span><span class="op">)</span></span>
<span><span class="va">row_inds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">13</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">values</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">13</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/sparseMatrix.html" class="external-link">sparseMatrix</a></span><span class="op">(</span>i <span class="op">=</span> <span class="va">row_inds</span>, j <span class="op">=</span> <span class="va">col_inds</span>, x <span class="op">=</span> <span class="va">values</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">coords1</span><span class="op">$</span><span class="va">barcode</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">LETTERS</span>, <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p>This should be sufficient to create an SPE object, and an SFE object,
even though no <code>sf</code> data frame was constructed for the
geometries. The constructor behaves similarly to the SPE constructor.
The centroid coordinates of the Visium spots in the example can be
converted into spot polygons with the <code>spotDiameter</code>
argument, which can also be relevant to other technologies with round
spots or beads, such as Slide-seq. Spot diameter in pixels in full
resolution images can be found in the
<code>scalefactors_json.json</code> file in Space Ranger output.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/SpatialFeatureExperiment.html" class="external-link">SpatialFeatureExperiment</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">mat</span><span class="op">)</span>, colData <span class="op">=</span> <span class="va">coords1</span>,</span>
<span>                                spatialCoordsNames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"col"</span>, <span class="st">"row"</span><span class="op">)</span>,</span>
<span>                                spotDiameter <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span></code></pre></div>
<p>More geometries and spatial graphs can be added after calling the
constructor.</p>
<p>Geometries can also be supplied in the constructor.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Convert regular data frame with coordinates to sf data frame</span></span>
<span><span class="va">cg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/df2sf.html" class="external-link">df2sf</a></span><span class="op">(</span><span class="va">coords1</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"col"</span>, <span class="st">"row"</span><span class="op">)</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"col"</span>, <span class="st">"row"</span><span class="op">)</span>, spotDiameter <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cg</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span></span>
<span><span class="va">sfe3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/SpatialFeatureExperiment.html" class="external-link">SpatialFeatureExperiment</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">mat</span><span class="op">)</span>, colGeometries <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>foo <span class="op">=</span> <span class="va">cg</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="technology-specific-notes">Technology specific notes<a class="anchor" aria-label="anchor" href="#technology-specific-notes"></a>
</h3>
<p>Not all commercial technologies have a function to directly read the
outputs. We may implement more such functions in the next version of
<code>SpatialFeatureExperiment</code>. For now we show example code to
read output from CosMX and Xenium.</p>
<div class="section level4">
<h4 id="gene-count-matrix-and-cell-metadata">Gene count matrix and cell metadata<a class="anchor" aria-label="anchor" href="#gene-count-matrix-and-cell-metadata"></a>
</h4>
<p>The gene count matrix and cell metadata (including cell centroid
coordinates) from example datasets for technologies such as CosMX and
Vizgen are CSV files. We recommend the <a href="https://vroom.r-lib.org/" class="external-link"><code>vroom</code></a> package to
quickly read in large CSV files. The CSV files are read in as data
frames. For the gene count matrix, this can be converted to a matrix and
then a sparse <code>dgCMatrix</code>. The matrix may need to be
transposed so the genes are in rows and cells are in columns. While
smFISH based data tend to be less sparse than scRNA-seq data, using
sparse matrix is worthwhile since the matrix is still about 50%
zero.</p>
<p>For 10x Genomics’ new single cell resolution technology Xenium, the
gene count matrix is an <code>h5</code> file, which can be read into R
as an SCE object with <code><a href="https://rdrr.io/pkg/DropletUtils/man/read10xCounts.html" class="external-link">DropletUtils::read10xCounts()</a></code>. This
can then be converted to <code>SpatialExperiment</code>, and then
<code>SpatialFeatureExperiment</code>. The gene count matrix is a
<code>DelayedArray</code>, so the data is not all loaded into memory and
operations on this matrix are performed in chunks. The
<code>DelayedArray</code> has been converted into a
<code>dgCMatrix</code> in memory. While the cell metadata is available
in the CSV format, there’s also the <code>parquet</code> format which is
more compact on disk, which can be read into R as a data frame with
<code><a href="https://arrow.apache.org/docs/r/reference/read_parquet.html" class="external-link">arrow::read_parquet()</a></code>. Example code:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Download data from https://www.10xgenomics.com/products/xenium-in-situ/preview-dataset-human-breast</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"curl -O https://cf.10xgenomics.com/samples/xenium/1.0.1/Xenium_FFPE_Human_Breast_Cancer_Rep1/Xenium_FFPE_Human_Breast_Cancer_Rep1_outs.zip"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"unzip Xenium_FFPE_Human_Breast_Cancer_Rep1_outs.zip"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"mv outs outs_R1"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"curl -O https://cf.10xgenomics.com/samples/xenium/1.0.1/Xenium_FFPE_Human_Breast_Cancer_Rep2/Xenium_FFPE_Human_Breast_Cancer_Rep2_outs.zip"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"unzip Xenium_FFPE_Human_Breast_Cancer_Rep2_outs.zip"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"mv outs outs_R2"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drighelli/SpatialExperiment" class="external-link">SpatialExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">DropletUtils</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/apache/arrow/" class="external-link">arrow</a></span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DropletUtils/man/read10xCounts.html" class="external-link">read10xCounts</a></span><span class="op">(</span><span class="st">"outs_R1/cell_feature_matrix.h5"</span><span class="op">)</span></span>
<span><span class="va">cell_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://arrow.apache.org/docs/r/reference/read_parquet.html" class="external-link">read_parquet</a></span><span class="op">(</span><span class="st">"outs_R1/cells.parquet"</span><span class="op">)</span></span>
<span><span class="co"># Add the centroid coordinates to colData</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="va">cell_info</span><span class="op">[</span>,<span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-coercion.html" class="external-link">toSpatialExperiment</a></span><span class="op">(</span><span class="va">sce</span>, spatialCoordsNames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x_centroid"</span>, <span class="st">"y_centroid"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/SpatialFeatureExperiment-coercion.html" class="external-link">toSpatialFeatureExperiment</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="cell-polygons">Cell polygons<a class="anchor" aria-label="anchor" href="#cell-polygons"></a>
</h4>
<p>File format of cell polygons (if available) is in different formats
in different technology. The cell polygons should be <a href="https://r-spatial.github.io/sf/" class="external-link"><code>sf</code></a> data frames
to put into <code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/dimGeometries.html" class="external-link">colGeometries()</a></code> of the SFE object. This section
explains how to do that for a number of smFISH-based technologies.</p>
<p>In Xenium, the cell polygons come in CSV or parquet files that can be
directly read into R as a data frame, with 2 columns for x and y
coordinates, and one indicating which cell the coordinates belong to.
Change the name of the cell ID column into “ID”, and use
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/df2sf.html" class="external-link">SpatialFeatureExperiment::df2sf()</a></code> to convert the data frame
into an <code>sf</code> data frame with <code>POLYGON</code> geometry.
Example code:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/apache/arrow/" class="external-link">arrow</a></span><span class="op">)</span></span>
<span><span class="va">cell_poly</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://arrow.apache.org/docs/r/reference/read_parquet.html" class="external-link">read_parquet</a></span><span class="op">(</span><span class="st">"outs_R2/cell_boundaries.parquet"</span><span class="op">)</span></span>
<span><span class="co"># Here the first column is cell ID</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cell_poly</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"ID"</span></span>
<span><span class="co"># "vertex_x" and "vertex_y" are the column names for coordinates here</span></span>
<span><span class="va">cell_sf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/df2sf.html" class="external-link">df2sf</a></span><span class="op">(</span><span class="va">cell_poly</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"vertex_x"</span>, <span class="st">"vertex_y"</span><span class="op">)</span>, geometryType <span class="op">=</span> <span class="st">"POLYGON"</span><span class="op">)</span></span></code></pre></div>
<p>In CoxMX, cell polygons are in CSV files. Besides the two coordinates
columns, there’s a column for field of view (FOV) and another for cell
ID. However, unlike in Xenium, the cell IDs are only unique in each FOV,
so they should be concatenated to FOV to make them unique. Then
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/df2sf.html" class="external-link">df2sf()</a></code> can also be used to convert the regular data frame
into <code>sf</code>. Example code:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Download data from https://nanostring.com/products/cosmx-spatial-molecular-imager/ffpe-dataset/</span></span>
<span><span class="co"># stored here: https://www.dropbox.com/s/hl3peavrx92bluy/Lung5_Rep1-polygons.csv?dl=0</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"wget https://www.dropbox.com/s/hl3peavrx92bluy/Lung5_Rep1-polygons.csv?dl=1"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"mv Lung5_Rep1-polygons.csv?dl=1 Lung5_Rep1-polygons.csv"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://vroom.r-lib.org" class="external-link">vroom</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span>
<span><span class="va">cell_poly</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://vroom.r-lib.org/reference/vroom.html" class="external-link">vroom</a></span><span class="op">(</span><span class="st">"Lung5_Rep1-polygons.csv"</span><span class="op">)</span></span>
<span><span class="va">cell_poly</span> <span class="op">&lt;-</span> <span class="va">cell_poly</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://tidyr.tidyverse.org/reference/unite.html" class="external-link">unite</a></span><span class="op">(</span><span class="st">"ID"</span>, <span class="va">fov</span><span class="op">:</span><span class="va">cellID</span><span class="op">)</span></span>
<span><span class="va">cell_sf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/df2sf.html" class="external-link">df2sf</a></span><span class="op">(</span><span class="va">cell_poly</span>, spatialCoordsNames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x_global_px"</span>, <span class="st">"y_global_px"</span><span class="op">)</span>,</span>
<span>                 geometryType <span class="op">=</span> <span class="st">"POLYGON"</span><span class="op">)</span></span></code></pre></div>
<p>See <a href="https://github.com/pachterlab/SFEData/blob/main/inst/scripts/make-data.R" class="external-link">the
code used to construct the example datasets in <code>SFEData</code></a>
for more examples.</p>
<p>Use <code><a href="https://r-spatial.github.io/sf/reference/valid.html" class="external-link">sf::st_is_valid()</a></code> to check if the polygons are
valid. Polygons with self-intersection are not valid, and will throw an
error in geometric operations. A common reason why polygons are invalid
is a protruding line, which can be eliminated with
<code>sf::st_buffer(cell_sf, dist = 0)</code>. Use
<code>sf::st_is_valid(cell_sf, reason = TRUE)</code>, and plot the
invalid polygons, to find why some polygons are not valid.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">#&gt; Platform: x86_64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS Monterey 12.6.7</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] Matrix_1.6-1                   rjson_0.2.21                  </span></span>
<span><span class="co">#&gt; [3] SpatialFeatureExperiment_1.2.2 Voyager_1.2.4                 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] DBI_1.1.3                   bitops_1.0-7               </span></span>
<span><span class="co">#&gt;   [3] deldir_1.0-9                s2_1.1.4                   </span></span>
<span><span class="co">#&gt;   [5] rlang_1.1.1                 magrittr_2.0.3             </span></span>
<span><span class="co">#&gt;   [7] matrixStats_1.0.0           e1071_1.7-13               </span></span>
<span><span class="co">#&gt;   [9] compiler_4.3.1              DelayedMatrixStats_1.22.5  </span></span>
<span><span class="co">#&gt;  [11] systemfonts_1.0.4           vctrs_0.6.3                </span></span>
<span><span class="co">#&gt;  [13] stringr_1.5.0               pkgconfig_2.0.3            </span></span>
<span><span class="co">#&gt;  [15] SpatialExperiment_1.10.0    wk_0.7.3                   </span></span>
<span><span class="co">#&gt;  [17] crayon_1.5.2                fastmap_1.1.1              </span></span>
<span><span class="co">#&gt;  [19] magick_2.7.5                XVector_0.40.0             </span></span>
<span><span class="co">#&gt;  [21] scuttle_1.10.2              utf8_1.2.3                 </span></span>
<span><span class="co">#&gt;  [23] rmarkdown_2.24              tzdb_0.4.0                 </span></span>
<span><span class="co">#&gt;  [25] ragg_1.2.5                  bit_4.0.5                  </span></span>
<span><span class="co">#&gt;  [27] purrr_1.0.2                 bluster_1.10.0             </span></span>
<span><span class="co">#&gt;  [29] xfun_0.40                   zlibbioc_1.46.0            </span></span>
<span><span class="co">#&gt;  [31] cachem_1.0.8                beachmat_2.16.0            </span></span>
<span><span class="co">#&gt;  [33] GenomeInfoDb_1.36.1         jsonlite_1.8.7             </span></span>
<span><span class="co">#&gt;  [35] rhdf5filters_1.12.1         DelayedArray_0.26.7        </span></span>
<span><span class="co">#&gt;  [37] scico_1.5.0                 Rhdf5lib_1.22.0            </span></span>
<span><span class="co">#&gt;  [39] BiocParallel_1.34.2         terra_1.7-39               </span></span>
<span><span class="co">#&gt;  [41] cluster_2.1.4               parallel_4.3.1             </span></span>
<span><span class="co">#&gt;  [43] R6_2.5.1                    bslib_0.5.1                </span></span>
<span><span class="co">#&gt;  [45] stringi_1.7.12              limma_3.56.2               </span></span>
<span><span class="co">#&gt;  [47] boot_1.3-28.1               GenomicRanges_1.52.0       </span></span>
<span><span class="co">#&gt;  [49] jquerylib_0.1.4             assertthat_0.2.1           </span></span>
<span><span class="co">#&gt;  [51] Rcpp_1.0.11                 SummarizedExperiment_1.30.2</span></span>
<span><span class="co">#&gt;  [53] knitr_1.43                  R.utils_2.12.2             </span></span>
<span><span class="co">#&gt;  [55] IRanges_2.34.1              igraph_1.5.1               </span></span>
<span><span class="co">#&gt;  [57] tidyselect_1.2.0            abind_1.4-5                </span></span>
<span><span class="co">#&gt;  [59] yaml_2.3.7                  codetools_0.2-19           </span></span>
<span><span class="co">#&gt;  [61] lattice_0.21-8              tibble_3.2.1               </span></span>
<span><span class="co">#&gt;  [63] Biobase_2.60.0              evaluate_0.21              </span></span>
<span><span class="co">#&gt;  [65] desc_1.4.2                  sf_1.0-14                  </span></span>
<span><span class="co">#&gt;  [67] units_0.8-3                 spData_2.3.0               </span></span>
<span><span class="co">#&gt;  [69] proxy_0.4-27                pillar_1.9.0               </span></span>
<span><span class="co">#&gt;  [71] MatrixGenerics_1.12.3       KernSmooth_2.23-22         </span></span>
<span><span class="co">#&gt;  [73] stats4_4.3.1                generics_0.1.3             </span></span>
<span><span class="co">#&gt;  [75] vroom_1.6.3                 rprojroot_2.0.3            </span></span>
<span><span class="co">#&gt;  [77] sp_2.0-0                    RCurl_1.98-1.12            </span></span>
<span><span class="co">#&gt;  [79] ggplot2_3.4.3               S4Vectors_0.38.1           </span></span>
<span><span class="co">#&gt;  [81] munsell_0.5.0               scales_1.2.1               </span></span>
<span><span class="co">#&gt;  [83] sparseMatrixStats_1.12.2    class_7.3-22               </span></span>
<span><span class="co">#&gt;  [85] glue_1.6.2                  tools_4.3.1                </span></span>
<span><span class="co">#&gt;  [87] ggnewscale_0.4.9            BiocNeighbors_1.18.0       </span></span>
<span><span class="co">#&gt;  [89] RSpectra_0.16-1             locfit_1.5-9.8             </span></span>
<span><span class="co">#&gt;  [91] fs_1.6.3                    rhdf5_2.44.0               </span></span>
<span><span class="co">#&gt;  [93] grid_4.3.1                  spdep_1.2-8                </span></span>
<span><span class="co">#&gt;  [95] sfarrow_0.4.1               colorspace_2.1-0           </span></span>
<span><span class="co">#&gt;  [97] DropletUtils_1.20.0         edgeR_3.42.4               </span></span>
<span><span class="co">#&gt;  [99] SingleCellExperiment_1.22.0 patchwork_1.1.3            </span></span>
<span><span class="co">#&gt; [101] GenomeInfoDbData_1.2.10     HDF5Array_1.28.1           </span></span>
<span><span class="co">#&gt; [103] cli_3.6.1                   textshaping_0.3.6          </span></span>
<span><span class="co">#&gt; [105] fansi_1.0.4                 S4Arrays_1.0.5             </span></span>
<span><span class="co">#&gt; [107] arrow_12.0.1.1              dplyr_1.1.2                </span></span>
<span><span class="co">#&gt; [109] gtable_0.3.3                R.methodsS3_1.8.2          </span></span>
<span><span class="co">#&gt; [111] sass_0.4.7                  digest_0.6.33              </span></span>
<span><span class="co">#&gt; [113] BiocGenerics_0.46.0         classInt_0.4-9             </span></span>
<span><span class="co">#&gt; [115] dqrng_0.3.0                 memoise_2.0.1              </span></span>
<span><span class="co">#&gt; [117] htmltools_0.5.6             pkgdown_2.0.7              </span></span>
<span><span class="co">#&gt; [119] R.oo_1.25.0                 lifecycle_1.0.3            </span></span>
<span><span class="co">#&gt; [121] bit64_4.0.5</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Lambda Moses, Kayla Jackson, Laura Luebbert, Sina Booeshaghi, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
