<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Voyager">
<title>Vignettes using Moran's I • Voyager</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Vignettes using Moran's I">
<meta property="og:description" content="Voyager">
<meta name="robots" content="noindex">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">Voyager</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">1.2.4</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/install.html">Install</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-technologies">Technologies</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-technologies">
    <a class="dropdown-item" href="../articles/visium_landing.html">Visium</a>
    <a class="dropdown-item" href="../articles/slideseqV2_landing.html">Slide-seq v2</a>
    <a class="dropdown-item" href="../articles/cosmx_landing.html">CosMX</a>
    <a class="dropdown-item" href="../articles/xenium_landing.html">Xenium</a>
    <a class="dropdown-item" href="../articles/merfish_landing.html">MERFISH</a>
    <a class="dropdown-item" href="../articles/seqfish_landing.html">seqFISH</a>
    <a class="dropdown-item" href="../articles/codex_landing.html">CODEX/PhenoCycler</a>
    <a class="dropdown-item" href="../articles/splitseq_landing.html">SPLiT-seq</a>
    <a class="dropdown-item" href="../articles/chromium_landing.html">10X Chromium V3</a>
    <a class="dropdown-item" href="../articles/10xnuclei_landing.html">10X Chromium Nuclei Isolation</a>
    <a class="dropdown-item" href="../articles/10xcrispr_landing.html">10X Chromium Single Cell CRISPR</a>
    <a class="dropdown-item" href="../articles/10xatac_landing.html">10X Single Cell ATAC</a>
    <a class="dropdown-item" href="../articles/10xmultiome_landing.html">10X Single Cell Multiome</a>
  </div>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-methods">Methods</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-methods">
    <h6 class="dropdown-header" data-toc-skip>Univariate global</h6>
    <a class="dropdown-item" href="../articles/moran_landing.html">Moran's I</a>
    <a class="dropdown-item" href="../articles/geary_landing.html">Geary's C</a>
    <a class="dropdown-item" href="../articles/correlogram_landing.html">Correlogram</a>
    <a class="dropdown-item" href="../articles/variogram_landing.html">Variogram</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Univariate local</h6>
    <a class="dropdown-item" href="../articles/moranplot_landing.html">Moran scatter plot</a>
    <a class="dropdown-item" href="../articles/localmoran_landing.html">Local Moran's I</a>
    <a class="dropdown-item" href="../articles/losh_landing.html">Local spatial heteroscedasticity</a>
    <a class="dropdown-item" href="../articles/getisord_landing.html">Getis-Ord Gi*</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Bivariate</h6>
    <a class="dropdown-item" href="../articles/lee_landing.html">Lee's L</a>
    <a class="dropdown-item" href="../articles/crossvariogram_landing.html">Cross variogram</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Multivariate</h6>
    <a class="dropdown-item" href="../articles/multispati_landing.html">MULTISPATI PCA</a>
    <a class="dropdown-item" href="../articles/localc_landing.html">Multivariate local Geary's C</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../articles/sfemethod">Custom methods</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/pachterlab/voyager/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Vignettes using Moran's I</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/pachterlab/voyager/blob/HEAD/vignettes/moran_landing.Rmd" class="external-link"><code>vignettes/moran_landing.Rmd</code></a></small>
      <div class="d-none name"><code>moran_landing.Rmd</code></div>
    </div>

    
    
<p>Tobler’s first law of geography <span class="citation">(Tobler
1970)</span> states that</p>
<blockquote>
<p>Everything is related to everything else. But near things are more
related than distant things.</p>
</blockquote>
<p>This observation motivates the examination of spatial
autocorrelation. Positive spatial autocorrelation is evident when nearby
things tend to be similar, such as that weather in Pasadena and downtown
Los Angeles (as opposed to the weather in Pasadena and San Francisco).
Negative spatial autocorrelation is evident when nearby things tend to
be more dissimilar, like squares on a chessboard. Spatial
autocorrelation can arise from an intrinsic process such as diffusion or
communication by physical contact, or result from a covariate that has
such an intrinsic process, or in areal data, when the areal units of
observation are smaller than the scale of the spatial process.</p>
<p>Moran’s I <span class="citation">(Moran 1950)</span> is one of the
most commonly used statistic of spatial autocorrelation, defined as</p>
<p><span class="math display">\[
I = \frac{n}{\sum_{i=1}^n \sum_{j=1}^n w_{ij}} \frac{\sum_{i=1}^n
\sum_{j=1}^n w_{ij} (x_i - \bar{x})(x_j - \bar{x})}{\sum_{i=1}^n (x_i -
\bar{x})^2},
\]</span></p>
<p>where <span class="math inline">\(n\)</span> is the number of spots
or locations, <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span> are different locations, or spots in
the Visium context, <span class="math inline">\(x\)</span> is a variable
with values at each location, and <span class="math inline">\(w_{ij}\)</span> is a spatial weight, which can be
inversely proportional to distance between spots or an indicator of
whether two spots are neighbors, subject to various definitions of
neighborhood.</p>
<p>Moran’s I is similar to the Pearson correlation between the value at
each location and the average value at its neighbors (but not identical,
see <span class="citation">(Lee 2001)</span>). Just like Pearson
correlation, Moran’s I is generally bound between -1 and 1, where
positive value indicates positive spatial autocorrelation and negative
value indicates negative spatial autocorrelation.</p>
<p>Below is a list of vignettes that use Moran’s I. The links point to
the sections that use Moran’s I. The corresponding Google Colab
notebooks are also linked to. The list is sorted by technology.</p>
<table class="table">
<colgroup>
<col width="28%">
<col width="24%">
<col width="47%">
</colgroup>
<thead><tr class="header">
<th>Vignette</th>
<th align="center">Colab Notebook</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><a href="https://pachterlab.github.io/voyager/articles/visium_10x_spatial.html#spatial-autocorrelation-of-qc-metrics">Spatial
analysis with 10X example Visium dataset</a></td>
<td align="center"><a href="https://colab.research.google.com/github/pachterlab/voyager/blob/documentation-devel/vignettes/visium_10x.ipynb" class="external-link">Colab
Notebook</a></td>
<td>Perform Moran’s I on QC metrics and gene expression in mouse
olfactory bulb dataset from 10X website. Also apply Moran’s I to k
nearest neighbor graph in gene expression PCA space instead of
histological space.</td>
</tr>
<tr class="even">
<td><a href="https://pachterlab.github.io/voyager/articles/vig1_visium_basic.html#morans-i">Basic
Visium exploratory data analysis</a></td>
<td align="center"><a href="https://colab.research.google.com/github/pachterlab/voyager/blob/documentation-devel/vignettes/vig1_visium_basic.ipynb" class="external-link">Colab
Notebook</a></td>
<td>Perform Moran’s I on QC metrics and highly variable genes in mouse
skeletal muscle dataset</td>
</tr>
<tr class="odd">
<td><a href="https://pachterlab.github.io/voyager/articles/vig2_visium.html#morans-i">Spatial
Visium exploratory data analysis</a></td>
<td align="center"><a href="https://colab.research.google.com/github/pachterlab/voyager/blob/documentation-devel/vignettes/vig2_visium.ipynb" class="external-link">Colab
Notebook</a></td>
<td>Perform Moran’s I on QC metrics and highly variable genes in mouse
skeletal muscle dataset, and with permutation testing</td>
</tr>
<tr class="even">
<td><a href="https://pachterlab.github.io/voyager/dev/articles/vig3_slideseq_v2.html#morans-i">Slide-seqV2
exploratory data analysis</a></td>
<td align="center"><a href="https://colab.research.google.com/github/pachterlab/voyager/blob/documentation-devel/vignettes/vig3_slideseq_v2.ipynb" class="external-link">Colab
Notebook</a></td>
<td>Perform Moran’s I on highly variable genes in human melanoma
metastasis Slide-seq2 dataset</td>
</tr>
<tr class="odd">
<td><a href="https://pachterlab.github.io/voyager/articles/vig4_cosmx.html#morans-i">CosMX
NSCLC analysis</a></td>
<td align="center"><a href="https://colab.research.google.com/github/pachterlab/voyager/blob/documentation-devel/vignettes/vig4_cosmx.ipynb" class="external-link">Colab
Notebook</a></td>
<td>Perform Moran’s I on QC metrics and on all genes, with
multithreading, in a human non-small cell lung cancer dataset</td>
</tr>
<tr class="even">
<td><a href="https://pachterlab.github.io/voyager/articles/vig5_xenium.html#spatial-autocorrelation-of-qc-metrics">Xenium
breast cancer analysis</a></td>
<td align="center"><a href="https://colab.research.google.com/github/pachterlab/voyager/blob/documentation-devel/vignettes/vig5_xenium.ipynb" class="external-link">Colab
Notebook</a></td>
<td>Perform Moran’s I on QC metrics and all genes, with multithreading,
in a human breast cancer dataset</td>
</tr>
<tr class="odd">
<td><a href="https://pachterlab.github.io/voyager/articles/vig6_merfish.html#spatial-autocorrelation-of-qc-metrics">MERFISH
mouse liver analysis</a></td>
<td align="center"><a href="https://colab.research.google.com/github/pachterlab/voyager/blob/documentation-devel/vignettes/vig6_merfish.ipynb" class="external-link">Colab
Notebook</a></td>
<td>Perform Moran’s I on QC metrics and all genes, with multithreading,
in a mouse liver dataset</td>
</tr>
<tr class="even">
<td><a href="https://pachterlab.github.io/voyager/articles/multispati.html">MULTISPATI
PCA and negative spatial autocorrelation</a></td>
<td align="center"><a href="https://colab.research.google.com/github/pachterlab/voyager/blob/documentation-devel/vignettes/multispati.ipynb" class="external-link">Colab
Notebook</a></td>
<td>Run MULTISPATI PCA on MERFISH mouse liver dataset, and compare the
results to those from non-spatial PCA. Moran’s I was computed on QC
metrics and cell embeddings in non-spatial and MULTISPATI PCs.</td>
</tr>
<tr class="odd">
<td><a href="https://pachterlab.github.io/voyager/articles/vig7_seqfish.html#univariate-spatial-statistics">seqFISH
exploratory data analysis</a></td>
<td align="center"><a href="https://colab.research.google.com/github/pachterlab/voyager/blob/documentation-devel/vignettes/vig7_seqfish.ipynb" class="external-link">Colab
Notebook</a></td>
<td>Perform Moran’s I with permutation testing on top highly variable
genes in mouse gastrulation data</td>
</tr>
<tr class="even">
<td><a href="https://pachterlab.github.io/voyager/dev/articles/vig8_codex.html#spatial-eda">CODEX
colon analysis</a></td>
<td align="center"><a href="https://colab.research.google.com/github/pachterlab/voyager/blob/documentation-devel/vignettes/vig8_codex.ipynb" class="external-link">Colab
Notebook</a></td>
<td>Perform Moran’s I with permutation testing on genes profiled in
human colon with CODEX proteomics</td>
</tr>
<tr class="odd">
<td><a href="https://pachterlab.github.io/voyager/articles/nonspatial.html#morans-i-1">10X
v3 Basic</a></td>
<td align="center"><a href="https://colab.research.google.com/github/pachterlab/voyager/blob/documentation-devel/vignettes/nonspatial.ipynb" class="external-link">Colab
Notebook</a></td>
<td>Apply Moran’s I to non-spatial human PBMC scRNA-seq dataset, with k
nearest neighbor graph in gene expression PCA space rather than
histological space</td>
</tr>
</tbody>
</table>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Lee2001-tm" class="csl-entry">
Lee, Sang-Il. 2001. <span>“Developing a Bivariate Spatial Association
Measure: An Integration of Pearson’s r and Moran’s
<span>I</span>.”</span> <em>J. Geogr. Syst.</em> 3 (4): 369–85.
</div>
<div id="ref-Moran1950-aq" class="csl-entry">
Moran, P A P. 1950. <span>“Notes on Continuous Stochastic
Phenomena.”</span> <em>Biometrika</em> 37 (1-2): 17–23.
</div>
<div id="ref-Tobler1970-hz" class="csl-entry">
Tobler, W R. 1970. <span>“A Computer Movie Simulating Urban Growth in
the Detroit Region.”</span> <em>Econ. Geogr.</em> 46 (sup1): 234–40.
</div>
</div>
</div>
  </main>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Lambda Moses, Kayla Jackson, Laura Luebbert, Sina Booeshaghi, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
