<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Voyager">
<title>MULTISPATI PCA and negative spatial autocorrelation • Voyager</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="MULTISPATI PCA and negative spatial autocorrelation">
<meta property="og:description" content="Voyager">
<meta name="robots" content="noindex">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">Voyager</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">1.4.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/install.html">Install</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-technologies">Technologies</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-technologies">
    <a class="dropdown-item" href="../articles/visium_landing.html">Visium</a>
    <a class="dropdown-item" href="../articles/slideseqV2_landing.html">Slide-seq v2</a>
    <a class="dropdown-item" href="../articles/cosmx_landing.html">CosMX</a>
    <a class="dropdown-item" href="../articles/xenium_landing.html">Xenium</a>
    <a class="dropdown-item" href="../articles/merfish_landing.html">MERFISH</a>
    <a class="dropdown-item" href="../articles/seqfish_landing.html">seqFISH</a>
    <a class="dropdown-item" href="../articles/codex_landing.html">CODEX/PhenoCycler</a>
    <a class="dropdown-item" href="../articles/splitseq_landing.html">SPLiT-seq</a>
    <a class="dropdown-item" href="../articles/chromium_landing.html">10X Chromium V3</a>
    <a class="dropdown-item" href="../articles/10xnuclei_landing.html">10X Chromium Nuclei Isolation</a>
    <a class="dropdown-item" href="../articles/10xcrispr_landing.html">10X Chromium Single Cell CRISPR</a>
    <a class="dropdown-item" href="../articles/10xatac_landing.html">10X Single Cell ATAC</a>
    <a class="dropdown-item" href="../articles/10xmultiome_landing.html">10X Single Cell Multiome</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-methods">Methods</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-methods">
    <h6 class="dropdown-header" data-toc-skip>Univariate global</h6>
    <a class="dropdown-item" href="../articles/moran_landing.html">Moran's I</a>
    <a class="dropdown-item" href="../articles/geary_landing.html">Geary's C</a>
    <a class="dropdown-item" href="../articles/correlogram_landing.html">Correlogram</a>
    <a class="dropdown-item" href="../articles/variogram_landing.html">Variogram</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Univariate local</h6>
    <a class="dropdown-item" href="../articles/moranplot_landing.html">Moran scatter plot</a>
    <a class="dropdown-item" href="../articles/localmoran_landing.html">Local Moran's I</a>
    <a class="dropdown-item" href="../articles/losh_landing.html">Local spatial heteroscedasticity</a>
    <a class="dropdown-item" href="../articles/getisord_landing.html">Getis-Ord Gi*</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Bivariate</h6>
    <a class="dropdown-item" href="../articles/lee_landing.html">Lee's L</a>
    <a class="dropdown-item" href="../articles/crossvariogram_landing.html">Cross variogram</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Multivariate</h6>
    <a class="dropdown-item" href="../articles/multispati_landing.html">MULTISPATI PCA</a>
    <a class="dropdown-item" href="../articles/localc_landing.html">Multivariate local Geary's C</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../articles/sfemethod">Custom methods</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/pachterlab/voyager/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>MULTISPATI PCA and negative spatial autocorrelation</h1>
                        <h4 data-toc-skip class="author">Lambda
Moses</h4>
            
            <h4 data-toc-skip class="date">2023-12-10</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pachterlab/voyager/blob/HEAD/vignettes/multispati.Rmd" class="external-link"><code>vignettes/multispati.Rmd</code></a></small>
      <div class="d-none name"><code>multispati.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Due to the large number of genes quantified in single cell and
spatial transcriptomics, dimension reduction is part of the standard
workflow to analyze such data, to visualize, to help interpreting the
data, to distill relevant information and reduce noise, to facilitate
downstream analyses such as clustering and pseudotime, to project
different samples into a shared latent space for data integration, and
so on.</p>
<p>The first dimension reduction methods we learn about, such as good
old principal component analysis (PCA), tSNE, and UMAP, don’t use
spatial information. With the rise of spatial transcriptomics, some
dimension reduction methods that take spatial dependence into account
have been written. Some, such as <code>SpatialPCA</code> <span class="citation">(Shang and Zhou 2022)</span>, <code>NSF</code> <span class="citation">(Townes and Engelhardt 2023)</span>, and
<code>MEFISTO</code> <span class="citation">(Velten et al. 2022)</span>
use factor analysis or probabilistic PCA which is related to factor
analysis, and model the factors as Gaussian processes, with a spatial
kernel for the covariance matrix, so the factors have positive spatial
autocorrelation and can be used for downstream clustering where the
clusters can be more spatially coherent. Some use graph convolution
networks on a spatial neighborhood graph to find spatially informed
embeddings of the cells, such as <code>conST</code> <span class="citation">(Zong et al. 2022)</span> and <code>SpaceFlow</code>
<span class="citation">(Ren et al. 2022)</span>. <code>SpaSRL</code>
<span class="citation">(Zhang et al. 2023)</span> finds a low dimension
projection of spatial neighborhood augmented data.</p>
<p>Spatially informed dimension reduction is actually not new, and dates
back to at least 1985, with Wartenberg’s crossover of Moran’s I and PCA
<span class="citation">(Wartenberg 1985)</span>, which was generalized
and further developed as MULTISPATI PCA <span class="citation">(Dray,
Saı̈d, and Débias 2008)</span>, implemented in the <a href="https://cran.r-project.org/web/packages/adespatial/index.html" class="external-link"><code>adespatial</code></a>
package on CRAN. In short, while PCA tries to maximize the variance
explained by each PC, MULTISPATI maximizes the product of Moran’s I and
variance explained. Also, while all the eigenvalues from PCA are
non-negative, because the covariance matrix is positive semidefinite,
MULTISPATI can give negative eigenvalues, which represent negative
spatial autocorrelation, which can be present and interesting but is not
as common as positive spatial autocorrelation and is often masked by the
latter <span class="citation">(Griffith 2019)</span>.</p>
<p>In single cell -omics conventions, let <span class="math inline">\(X\)</span> denote a gene count matrix whose
columns are cells or Visium spots and whose rows are genes, with <span class="math inline">\(n\)</span> columns. Let <span class="math inline">\(W\)</span> denote the row normalized <span class="math inline">\(n\times n\)</span> adjacency matrix of the spatial
neighborhood graph of the cells or Visium spots, which does not have to
be symmetric. MULTISPATI diagonalizes a symmetric matrix</p>
<p><span class="math display">\[
H = \frac 1 {2n} X(W^t+W)X^t
\]</span></p>
<p>However, the implementation in <code>adespatial</code> is more
general and can be used for other multivariate analyses in the <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3265363/" class="external-link">duality
diagram</a> paradigm, such as correspondence analysis; the equation
above is simplified just for PCA, without having to introduce the
duality diagram here.</p>
<p>Voyager 1.2.0 (Bioconductor 3.17) has a much faster implementation of
MULTISPATI PCA based on <a href="https://cran.r-project.org/web/packages/RSpectra/index.html" class="external-link"><code>RSpectra</code></a>.
See benchmark <a href="https://lambdamoses.github.io/thevoyages/posts/2023-03-25-multispati-part-2/" class="external-link">here</a>.</p>
<p>In this vignette, we perform MULTISPATI PCA on the MERFISH mouse
liver dataset. See the first vignette using this dataset <a href="https://pachterlab.github.io/voyager/articles/vig6_merfish.html">here</a>.</p>
<p>Here we load the packages used:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/voyager" class="external-link">Voyager</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SFEData" class="external-link">SFEData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SpatialFeatureExperiment" class="external-link">SpatialFeatureExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scuttle</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org" class="external-link">stringr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/" class="external-link">tibble</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/LTLA/BiocSingular" class="external-link">BiocSingular</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-spatial.github.io/sf/" class="external-link">sf</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/VizgenLiverData.html" class="external-link">VizgenLiverData</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; see ?SFEData and browseVignettes('SFEData') for documentation</span></span>
<span><span class="co">#&gt; downloading 1 resources</span></span>
<span><span class="co">#&gt; retrieving 1 resource</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 385 395215 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(385): Comt Ldha ... Blank-36 Blank-37</span></span>
<span><span class="co">#&gt; rowData names(3): means vars cv2</span></span>
<span><span class="co">#&gt; colnames(395215): 10482024599960584593741782560798328923</span></span>
<span><span class="co">#&gt;   111551578131181081835796893618918348842 ...</span></span>
<span><span class="co">#&gt;   92389687374928708938472537234969690424</span></span>
<span><span class="co">#&gt;   96399783859933548456002372694492036651</span></span>
<span><span class="co">#&gt; colData names(9): fov volume ... nCounts nGenes</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : center_x center_y</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: full_res_image_pixels</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: centroids (POINT), cellSeg (POLYGON) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01:</span></span></code></pre></div>
<p>MULTISPATI PCA is one of the multivariate methods introduced in
<code>Voyager</code> 1.2.0. All multivariate methods in
<code>Voyager</code> are listed here:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/listSFEMethods.html">listSFEMethods</a></span><span class="op">(</span>variate <span class="op">=</span> <span class="st">"multi"</span><span class="op">)</span></span>
<span><span class="co">#&gt;                name                                      description</span></span>
<span><span class="co">#&gt; 1        multispati                                   MULTISPATI PCA</span></span>
<span><span class="co">#&gt; 2      localC_multi                     Multivariate local Geary's C</span></span>
<span><span class="co">#&gt; 3 localC_perm_multi Multivariate local Geary's C permutation testing</span></span></code></pre></div>
<p>When calling <code>calculate*variate()</code> or
<code>run*variate()</code>, the <code>type</code> (2nd) argument takes
either an <code>SFEMethod</code> object or a string that matches an
entry in the <code>name</code> column in the data frame returned by
<code><a href="../reference/listSFEMethods.html">listSFEMethods()</a></code>.</p>
</div>
<div class="section level2">
<h2 id="quality-control">Quality control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h2>
<p>QC was already performed in the <a href="https://pachterlab.github.io/voyager/articles/vig6_merfish.html">first
vignette</a>. We do the same QC here, but see the first vignette for
more details.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">is_blank</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span>, <span class="st">"^Blank-"</span><span class="op">)</span></span>
<span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/addPerCellQCMetrics.html" class="external-link">addPerCellQCMetrics</a></span><span class="op">(</span><span class="va">sfe</span>, subset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>blank <span class="op">=</span> <span class="va">is_blank</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">get_neg_ctrl_outliers</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">col</span>, <span class="va">sfe</span>, <span class="va">nmads</span> <span class="op">=</span> <span class="fl">3</span>, <span class="va">log</span> <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">inds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">nCounts</span> <span class="op">&gt;</span> <span class="fl">0</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[[</span><span class="va">col</span><span class="op">]</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">0</span></span>
<span>    <span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span><span class="va">inds</span>,<span class="op">]</span></span>
<span>    <span class="va">outlier_inds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html" class="external-link">isOutlier</a></span><span class="op">(</span><span class="va">df</span><span class="op">[[</span><span class="va">col</span><span class="op">]</span><span class="op">]</span>, type <span class="op">=</span> <span class="st">"higher"</span>, nmads <span class="op">=</span> <span class="va">nmads</span>, log <span class="op">=</span> <span class="va">log</span><span class="op">)</span></span>
<span>    <span class="va">outliers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span><span class="op">[</span><span class="va">outlier_inds</span><span class="op">]</span></span>
<span>    <span class="va">col2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html" class="external-link">str_remove</a></span><span class="op">(</span><span class="va">col</span>, <span class="st">"^subsets_"</span><span class="op">)</span></span>
<span>    <span class="va">col2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html" class="external-link">str_remove</a></span><span class="op">(</span><span class="va">col2</span>, <span class="st">"_percent$"</span><span class="op">)</span></span>
<span>    <span class="va">new_colname</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"is"</span>, <span class="va">col2</span>, <span class="st">"outlier"</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[[</span><span class="va">new_colname</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">outliers</span></span>
<span>    <span class="va">sfe</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu">get_neg_ctrl_outliers</span><span class="op">(</span><span class="st">"subsets_blank_percent"</span>, <span class="va">sfe</span>, log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Remove the outliers and empty cells:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">inds</span> <span class="op">&lt;-</span> <span class="op">!</span><span class="va">sfe</span><span class="op">$</span><span class="va">is_blank_outlier</span> <span class="op">&amp;</span> <span class="va">sfe</span><span class="op">$</span><span class="va">nCounts</span> <span class="op">&gt;</span> <span class="fl">0</span></span>
<span><span class="op">(</span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">[</span>, <span class="va">inds</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 385 390348 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(385): Comt Ldha ... Blank-36 Blank-37</span></span>
<span><span class="co">#&gt; rowData names(3): means vars cv2</span></span>
<span><span class="co">#&gt; colnames(390348): 10482024599960584593741782560798328923</span></span>
<span><span class="co">#&gt;   111551578131181081835796893618918348842 ...</span></span>
<span><span class="co">#&gt;   92389687374928708938472537234969690424</span></span>
<span><span class="co">#&gt;   96399783859933548456002372694492036651</span></span>
<span><span class="co">#&gt; colData names(16): fov volume ... total is_blank_outlier</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : center_x center_y</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: full_res_image_pixels</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: centroids (POINT), cellSeg (POLYGON) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01:</span></span></code></pre></div>
<p>There still are over 390,000 cells left after removing the
outliers.</p>
<p>Next we compute Moran’s I for QC metrics, which requires a spatial
neighborhood graph:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"knn5"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findSpatialNeighbors-SpatialFeatureExperiment-method.html" class="external-link">findSpatialNeighbors</a></span><span class="op">(</span><span class="va">sfe</span>, method <span class="op">=</span> <span class="st">"knearneigh"</span>, </span>
<span>                                                  dist_type <span class="op">=</span> <span class="st">"idw"</span>, k <span class="op">=</span> <span class="fl">5</span>, </span>
<span>                                                  style <span class="op">=</span> <span class="st">"W"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt;  32.773   0.182  32.972</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">features_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span>, <span class="st">"volume"</span><span class="op">)</span></span>
<span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">colDataUnivariate</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"moran.mc"</span>, <span class="va">features_use</span>, </span>
<span>                         colGraphName <span class="op">=</span> <span class="st">"knn5"</span>, nsim <span class="op">=</span> <span class="fl">49</span>, </span>
<span>                         BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotMoranMC.html">plotMoranMC</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">features_use</span><span class="op">)</span></span></code></pre></div>
<p><img src="multispati_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<p>Here Moran’s I is a little negative, but from the permutation
testing, it is significant, though that can also be the large number of
cells. The lower bound of Moran’s I given the spatial neighborhood graph
is usually closer to -0.5 than -1, while the upper bound is usually
around 1. The bounds given a specific spatial neighborhood graph can be
found with <code><a href="../reference/moranBounds.html">moranBounds()</a></code>, but because it double centers the
adjacency matrix, hence making it dense, there isn’t enough memory to
use it for the entire dataset. But we can look at the Moran bounds of a
small subset of data, which might be generalizable to the whole dataset
given that this tissue appears quite homogeneous in space.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bbox_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>xmin <span class="op">=</span> <span class="fl">6000</span>, xmax <span class="op">=</span> <span class="fl">7000</span>, ymin <span class="op">=</span> <span class="fl">4000</span>, ymax <span class="op">=</span> <span class="fl">5000</span><span class="op">)</span></span>
<span><span class="va">inds2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://r-spatial.github.io/sf/reference/geos_binary_pred.html" class="external-link">st_intersects</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/dimGeometries.html" class="external-link">cellSeg</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span>, <span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_as_sfc.html" class="external-link">st_as_sfc</a></span><span class="op">(</span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_bbox.html" class="external-link">st_bbox</a></span><span class="op">(</span><span class="va">bbox_use</span><span class="op">)</span><span class="op">)</span>, </span>
<span>                       sparse <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">sfe_sub</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">[</span>, <span class="va">inds2</span><span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">mb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/moranBounds.html">moranBounds</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe_sub</span>, <span class="st">"knn5"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;       Imin       Imax </span></span>
<span><span class="co">#&gt; -0.6079436  1.0608389</span></span></code></pre></div>
<p>So considering the bounds, the MOran’s I values of the QC metrics are
more like</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="fu"><a href="../reference/colFeatureData.html">colFeatureData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span>, <span class="st">"volume"</span><span class="op">)</span>, </span>
<span>                             <span class="st">"moran.mc_statistic_sample01"</span><span class="op">]</span> <span class="op">/</span> <span class="va">mb</span><span class="op">[</span><span class="st">"Imin"</span><span class="op">]</span>,</span>
<span>         <span class="va">features_use</span><span class="op">)</span></span>
<span><span class="co">#&gt;    nCounts     nGenes     volume </span></span>
<span><span class="co">#&gt; 0.17839356 0.15168017 0.03211427</span></span></code></pre></div>
<p>whose magnitudes seem more substantial for <code>nCounts</code> and
<code>nGenes</code> if it’s positive spatial autocorrelation. So there
may be mild to moderate negative spatial autocorrelation.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Normalize data</span></span>
<span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="hepatic-zonation">Hepatic zonation<a class="anchor" aria-label="anchor" href="#hepatic-zonation"></a>
</h2>
<p>This dataset comes from a relatively large piece of tissue and we
need to zoom into a smaller region to better see the local structures.
Here we specify a bounding box.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bbox_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>xmin <span class="op">=</span> <span class="fl">6100</span>, xmax <span class="op">=</span> <span class="fl">7100</span>, ymin <span class="op">=</span> <span class="fl">7500</span>, ymax <span class="op">=</span> <span class="fl">8500</span><span class="op">)</span></span></code></pre></div>
<p>A portal triad is shown near the top right of this bounding box. The
two large vessels on the left and bottom right are central veins. The
portal triad consists of the hepatic artery, portal vein which brings
blood from the intestine, and bile duct, so it’s more oxygenated. The
regions around the central vein is more deoxygenated. The different
oxygen and nutrient contents mean that hepatocytes play different
metabolic roles in the zones between the portal triad and the central
vein. Here we plot some zonation marker genes from <span class="citation">(Halpern et al. 2017)</span>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Axin2"</span>, <span class="st">"Cyp1a2"</span>, <span class="st">"Gstm3"</span>, <span class="st">"Psmd4"</span>, <span class="co"># Pericentral</span></span>
<span>             <span class="st">"Cyp2e1"</span>, <span class="st">"Asl"</span>, <span class="st">"Alb"</span>, <span class="st">"Ass1"</span>, <span class="co"># Monotonic but has intermediate</span></span>
<span>             <span class="st">"Hamp"</span>, <span class="st">"Igfbp2"</span>, <span class="st">"Cyp8b1"</span>, <span class="st">"Mup3"</span>, <span class="co"># Non-monotonic</span></span>
<span>             <span class="st">"Arg1"</span>, <span class="st">"Pck1"</span>, <span class="st">"C2"</span>, <span class="st">"Sdhd"</span><span class="op">)</span> <span class="co"># Periportal</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">inds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">markers</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1]  1  2 14</span></span></code></pre></div>
<p>Only 3 of these marker genes are present in this dataset. The first
two are pericentral (near the central vein), and the last one is
periportal (near the portal triad).</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">markers</span><span class="op">[</span><span class="va">inds</span><span class="op">]</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>,</span>
<span>                   ncol <span class="op">=</span> <span class="fl">3</span>, bbox <span class="op">=</span> <span class="va">bbox_use</span><span class="op">)</span></span></code></pre></div>
<p><img src="multispati_files/figure-html/unnamed-chunk-19-1.png" width="864"></p>
<p>Besides hepatocytes, the liver also has many endothelial cells and
Kupffer cells (macrophages). Marker genes of these cells from <span class="citation">(Bonnardel et al. 2019)</span> are plotted to visualize
these cell types in space:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Kuppfer cells</span></span>
<span><span class="va">kc_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Timd4"</span>, <span class="st">"Vsig4"</span>, <span class="st">"Clec4f"</span>, <span class="st">"Clec1b"</span>, <span class="st">"Il18bp"</span>, <span class="st">"C6"</span>, <span class="st">"Irf7"</span>,</span>
<span>              <span class="st">"Slc40a1"</span>, <span class="st">"Cdh5"</span>, <span class="st">"Nr1h3"</span>, <span class="st">"Dmpk"</span>, <span class="st">"Paqr9"</span>, <span class="st">"Pcolce2"</span>, <span class="st">"Kcna2"</span>,</span>
<span>              <span class="st">"Gbp8"</span>, <span class="st">"Iigp1"</span>, <span class="st">"Helz2"</span>, <span class="st">"Cd207"</span>, <span class="st">"Icos"</span>, <span class="st">"Adcy4"</span>, <span class="st">"Slc1a2"</span>,</span>
<span>              <span class="st">"Rsad2"</span>, <span class="st">"Slc16a9"</span>, <span class="st">"Cd209f"</span>, <span class="st">"Oasl1"</span>, <span class="st">"Fam167a"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">kc_genes</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 9</span></span></code></pre></div>
<p>Only one of the Kupffer cell markers is available in this
dataset.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">kc_genes</span><span class="op">[</span><span class="fl">9</span><span class="op">]</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>,</span>
<span>                   bbox <span class="op">=</span> <span class="va">bbox_use</span><span class="op">)</span></span></code></pre></div>
<p><img src="multispati_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
<p>Expression of this gene does not seem very spatially coherent.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Endothelial cells</span></span>
<span><span class="va">lec_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Rspo3"</span>, <span class="st">"Wnt2"</span>, <span class="st">"Wnt9b"</span>, <span class="st">"Pcdhgc5"</span>, <span class="st">"Ecm1"</span>, <span class="st">"Ltbp4"</span>, <span class="st">"Efnb2"</span><span class="op">)</span></span>
<span><span class="op">(</span><span class="va">inds_lec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">lec_genes</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 2 6 7</span></span></code></pre></div>
<p>Only 3 of these endothelial cell marker genes are available in this
dataset.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">lec_genes</span><span class="op">[</span><span class="va">inds_lec</span><span class="op">]</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>,</span>
<span>                   bbox <span class="op">=</span> <span class="va">bbox_use</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="multispati_files/figure-html/unnamed-chunk-23-1.png" width="864"></p>
<p>Wnt2 seems to be more pericentral, while Ltbp4 and Efnb2 seem more
periportal.</p>
<p>Some of these marker genes will show up in the top PC loadings
non-spatial and spatial PCA.</p>
</div>
<div class="section level2">
<h2 id="non-spatial-pca">Non-spatial PCA<a class="anchor" aria-label="anchor" href="#non-spatial-pca"></a>
</h2>
<p>First we run non-spatial PCA, to compare to MULTISPATI.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">29</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span></span>
<span>    <span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sfe</span>, ncomponents <span class="op">=</span> <span class="fl">20</span>, subset_row <span class="op">=</span> <span class="op">!</span><span class="va">is_blank</span>,</span>
<span>                  exprs_values <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>                  scale <span class="op">=</span> <span class="cn">TRUE</span>, BSPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/BiocSingularParam.html" class="external-link">IrlbaParam</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt;  19.600   1.033  20.656</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/gc.html" class="external-link">gc</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;             used   (Mb) gc trigger   (Mb) limit (Mb)  max used   (Mb)</span></span>
<span><span class="co">#&gt; Ncells  16042090  856.8   23199771 1239.1         NA  23199771 1239.1</span></span>
<span><span class="co">#&gt; Vcells 239267700 1825.5  536340089 4092.0      16384 536197849 4090.9</span></span></code></pre></div>
<p>That’s pretty quick for almost 400,000 cells, but there aren’t that
many genes here. Use the elbow plot to see variance explained by each
PC:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ElbowPlot.html">ElbowPlot</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span></code></pre></div>
<p><img src="multispati_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
<p>Plot top gene loadings in each PC</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotDimLoadings.html">plotDimLoadings</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span></code></pre></div>
<p><img src="multispati_files/figure-html/unnamed-chunk-26-1.png" width="700"></p>
<p>Many of these genes seem to be related to the endothelium. PC1 and
PC4 concern the Kupffer cells as well, as the Kupffer cell marker gene
Cdh5 has high loading.</p>
<p>Plot the first 4 PCs in space</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/spatialReducedDim.html">spatialReducedDim</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"PCA"</span>, <span class="fl">4</span>, colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>, scattermore <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                  divergent <span class="op">=</span> <span class="cn">TRUE</span>, diverge_center <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p><img src="multispati_files/figure-html/unnamed-chunk-27-1.png" width="864"></p>
<p>PC1 and PC4 highlight the major blood vessels, while PC2 and PC3 have
less spatial structure. While in the CosMX and Xenium datasets on this
website, the top PCs have clear spatial structures despite the absence
of spatial information in non-spatial PCA because of clear spatial
compartments for some cell types, which does not seem to be the case in
this dataset except for the blood vessels. We have seen above that some
genes have strong spatial structures.</p>
<p>While PC2 and PC3 don’t seem to have large scale spatial structure,
they may have more local spatial structure not obvious from plotting the
entire section, so we zoom into a bounding box which shows hepatic
zonation.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/spatialReducedDim.html">spatialReducedDim</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"PCA"</span>, ncomponents <span class="op">=</span> <span class="fl">4</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>,</span>
<span>                  bbox <span class="op">=</span> <span class="va">bbox_use</span>, divergent <span class="op">=</span> <span class="cn">TRUE</span>, diverge_center <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p><img src="multispati_files/figure-html/unnamed-chunk-28-1.png" width="864"></p>
<p>There’s some spatial structure at a smaller scale, perhaps some
negative spatial autocorrelation.</p>
</div>
<div class="section level2">
<h2 id="multispati-pca">MULTISPATI PCA<a class="anchor" aria-label="anchor" href="#multispati-pca"></a>
</h2>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateMultivariate.html">runMultivariate</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"multispati"</span>, colGraphName <span class="op">=</span> <span class="st">"knn5"</span>, nfposi <span class="op">=</span> <span class="fl">20</span>,</span>
<span>                       nfnega <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in asMethod(object): sparse-&gt;dense coercion: allocating vector of size</span></span>
<span><span class="co">#&gt; 1.1 GiB</span></span>
<span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt; 171.052  18.654 205.442</span></span></code></pre></div>
<p>Then plot the most positive and most negative eigenvalues. Note that
the eigenvalues here are not variance explained. Instead, they are the
product of variance explained and Moran’s I. So the most positive
eigenvalues correspond to eigenvectors that simultaneously explain more
variance and have large positive Moran’s I. The most negative
eigenvalues correspond to eigenvectors that simultaneously explain more
variance and have negative Moran’s I.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ElbowPlot.html">ElbowPlot</a></span><span class="op">(</span><span class="va">sfe</span>, nfnega <span class="op">=</span> <span class="fl">20</span>, reduction <span class="op">=</span> <span class="st">"multispati"</span><span class="op">)</span></span></code></pre></div>
<p><img src="multispati_files/figure-html/unnamed-chunk-30-1.png" width="700"></p>
<p>Here the positive eigenvalues drop sharply from PC1 to PC4, and there
is only one very negative eigenvalue which might be interesting, which
is unsurprising given the moderately negative Moran’s I for
<code>nCounts</code> and <code>nGenes</code>. However, from the <a href="https://pachterlab.github.io/voyager/articles/vig6_merfish.html#morans-i">first
MERFISH vignette</a>, none of the genes have very negative Moran’s I.
Perhaps the negative eigenvalue comes from negative spatial
autocorrelation in a gene program or “eigengene” and is not obvious from
individual genes. This is the beauty of multivariate analysis.</p>
<p>What do these components mean? Each component is a linear combination
of genes to maximize the product of variance explained and Moran’s I.
The second component maximizes this product provided that it’s
orthogonal to the first component, and so on. As the loss in variance
explained is usually not huge, these components can be considered axes
along which <em>spatially coherent</em> groups of spots are separated
from each other as much as possible according to expression of the
highly variable genes, so in theory, clustering with positive MULTISPATI
components should give more spatially coherent clusters. Because of the
spatial coherence, MULTISPATI might be more robust to outliers.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotDimLoadings.html">plotDimLoadings</a></span><span class="op">(</span><span class="va">sfe</span>, dims <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">40</span><span class="op">)</span>, reduction <span class="op">=</span> <span class="st">"multispati"</span><span class="op">)</span></span></code></pre></div>
<p><img src="multispati_files/figure-html/unnamed-chunk-31-1.png" width="700"></p>
<p>From gene loadings, PC40 seems to separate endothelial cells and
Kupffer cells from hepatocytes.</p>
<p>Plot the these PCs:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/spatialReducedDim.html">spatialReducedDim</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"multispati"</span>, components <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">40</span><span class="op">)</span>, </span>
<span>                  colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>, bbox <span class="op">=</span> <span class="va">bbox_use</span>,</span>
<span>                  divergent <span class="op">=</span> <span class="cn">TRUE</span>, diverge_center <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p><img src="multispati_files/figure-html/unnamed-chunk-32-1.png" width="864"></p>
<p>The first two PCs pick up zoning. PC3 seems to have smaller scale
spatial structure. PC”40” (should really be 300 something) is an example
of negative spatial autocorrelation in biology. That Kupffer cells and
endothelial cells are scattered among hepatocytes may play a functional
role. This does not mean that non-spatial PCA is bad. While MULTISPATI
tends not to lose too much variance explained in per PC with positive
eigenvalues, it identifies co-expressed genes with spatially structured
expression patterns. MULTISPATI tells a different story from non-spatial
PCA. PCA cell embeddings are often used for downstream analysis. Whether
to use MULTISPATI embeddings instead and which or how many PCs to use
depend on the questions asked in the further downstream analyses.</p>
</div>
<div class="section level2">
<h2 id="spatial-autocorrelation-of-principal-components">Spatial autocorrelation of principal components<a class="anchor" aria-label="anchor" href="#spatial-autocorrelation-of-principal-components"></a>
</h2>
<div class="section level3">
<h3 id="morans-i">Moran’s I<a class="anchor" aria-label="anchor" href="#morans-i"></a>
</h3>
<p>Here we compare Moran’s I for cell embeddings in each non-spatial and
MULTISPATI PC:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># non-spatial</span></span>
<span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">reducedDimMoransI</a></span><span class="op">(</span><span class="va">sfe</span>, dimred <span class="op">=</span> <span class="st">"PCA"</span>, components <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span>,</span>
<span>                         BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># spatial</span></span>
<span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">reducedDimMoransI</a></span><span class="op">(</span><span class="va">sfe</span>, dimred <span class="op">=</span> <span class="st">"multispati"</span>, components <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">40</span>,</span>
<span>                         BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df_moran</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>PCA <span class="op">=</span> <span class="fu"><a href="../reference/colFeatureData.html">reducedDimFeatureData</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"PCA"</span><span class="op">)</span><span class="op">$</span><span class="va">moran_sample01</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span>,</span>
<span>                   MULTISPATI_pos <span class="op">=</span> </span>
<span>                       <span class="fu"><a href="../reference/colFeatureData.html">reducedDimFeatureData</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"multispati"</span><span class="op">)</span><span class="op">$</span><span class="va">moran_sample01</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span>,</span>
<span>                   MULTISPATI_neg <span class="op">=</span> </span>
<span>                       <span class="fu"><a href="../reference/colFeatureData.html">reducedDimFeatureData</a></span><span class="op">(</span><span class="va">sfe</span>,<span class="st">"multispati"</span><span class="op">)</span><span class="op">$</span><span class="va">moran_sample01</span><span class="op">[</span><span class="fl">21</span><span class="op">:</span><span class="fl">40</span><span class="op">]</span> <span class="op">|&gt;</span> </span>
<span>                       <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>                   index <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"ditto_colors"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df_moran</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">index</span>, values_to <span class="op">=</span> <span class="st">"value"</span>, names_to <span class="op">=</span> <span class="st">"name"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">index</span>, <span class="va">value</span>, color <span class="op">=</span> <span class="va">name</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">ditto_colors</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">0</span>, color <span class="op">=</span> <span class="st">"gray"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="va">mb</span>, linetype <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="st">"gray"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/breaks_pretty.html" class="external-link">breaks_pretty</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/breaks_width.html" class="external-link">breaks_width</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Moran's I"</span>, color <span class="op">=</span> <span class="st">"Type"</span>, x <span class="op">=</span> <span class="st">"Component"</span><span class="op">)</span></span></code></pre></div>
<p><img src="multispati_files/figure-html/unnamed-chunk-36-1.png" width="700"></p>
<p>In MULTISPATI, Moran’s I is high in PC1 and PC2, but then sharply
drops. Moran’s I for the PC with the most negative eigenvalues is not
very negative, which means the large magnitude of that eigenvalue comes
from explaining more variance. However, considering the lower bound of
Moran’s I that is around -0.6 instead of -1, the magnitude of Moran’s I
for the PC with the most negative eigenvalue is not trivial.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">df_moran</span><span class="op">$</span><span class="va">MULTISPATI_neg</span><span class="op">)</span> <span class="op">/</span> <span class="va">mb</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="co">#&gt;      Imin </span></span>
<span><span class="co">#&gt; 0.1374483</span></span></code></pre></div>
<p>Non-spatial PCs are not sorted by Moran’s I; PC5 has surprising large
Moran’s I.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/spatialReducedDim.html">spatialReducedDim</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"PCA"</span>, component <span class="op">=</span> <span class="fl">5</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>, </span>
<span>                  divergent <span class="op">=</span> <span class="cn">TRUE</span>, diverge_center <span class="op">=</span> <span class="fl">0</span>, bbox <span class="op">=</span> <span class="va">bbox_use</span><span class="op">)</span></span></code></pre></div>
<p><img src="multispati_files/figure-html/unnamed-chunk-38-1.png" width="700"></p>
<p>PC5 must be about zonation. Also show a larger scale:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/spatialReducedDim.html">spatialReducedDim</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"PCA"</span>, components <span class="op">=</span> <span class="fl">5</span>, colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>, </span>
<span>                  divergent <span class="op">=</span> <span class="cn">TRUE</span>, diverge_center <span class="op">=</span> <span class="fl">0</span>, scattermore <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="multispati_files/figure-html/unnamed-chunk-39-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="moran-scatter-plot">Moran scatter plot<a class="anchor" aria-label="anchor" href="#moran-scatter-plot"></a>
</h3>
<p>Local positive and negative spatial autocorrelation can average out
in global Moran’s I. From the zoomed in plots and gene loadings above,
some PCs are about endothelial cells. The Moran scatter plot can help
discovering more local heterogeneity.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">reducedDimUnivariate</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"moran.plot"</span>, dimred <span class="op">=</span> <span class="st">"PCA"</span>, components <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">6</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="../reference/moranPlot.html">moranPlot</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"PC"</span>, <span class="va">i</span><span class="op">)</span>, binned <span class="op">=</span> <span class="cn">TRUE</span>, hex <span class="op">=</span> <span class="cn">TRUE</span>, plot_influential <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plts</span>, widths <span class="op">=</span> <span class="fl">1</span>, heights <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span>tag_levels <span class="op">=</span> <span class="st">"1"</span>, </span>
<span>                    title <span class="op">=</span> <span class="st">"Moran scatter plot for non-spatial PCs"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<p><img src="multispati_files/figure-html/unnamed-chunk-42-1.png" width="960"></p>
<p>PCs 1-3 have some fainter clusters outside the main cluster,
indicating heterogeneous spatial autocorrelation. Also make Moran
scatter plots for MULTISPATI</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">reducedDimUnivariate</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"moran.plot"</span>, dimred <span class="op">=</span> <span class="st">"multispati"</span>, </span>
<span>                            components <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">40</span><span class="op">)</span>, </span>
<span>                            <span class="co"># Not to overwrite non-spatial PCA moran plots</span></span>
<span>                            name <span class="op">=</span> <span class="st">"moran.plot2"</span><span class="op">)</span> </span></code></pre></div>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plts2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">40</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="../reference/moranPlot.html">moranPlot</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"PC"</span>, <span class="va">i</span><span class="op">)</span>, binned <span class="op">=</span> <span class="cn">TRUE</span>, hex <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>              plot_influential <span class="op">=</span> <span class="cn">FALSE</span>, name <span class="op">=</span> <span class="st">"moran.plot2"</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plts2</span>, widths <span class="op">=</span> <span class="fl">1</span>, heights <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span>tag_levels <span class="op">=</span> <span class="st">"1"</span>,</span>
<span>                    title <span class="op">=</span> <span class="st">"Moran scatter plot for MULTISPATI PCs"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<p><img src="multispati_files/figure-html/unnamed-chunk-45-1.png" width="960"></p>
<p>There are some interesting clusters.</p>
</div>
</div>
<div class="section level2">
<h2 id="clustering-with-multispati-pca">Clustering with MULTISPATI PCA<a class="anchor" aria-label="anchor" href="#clustering-with-multispati-pca"></a>
</h2>
<p>In the standard scRNA-seq data analysis workflow, a k nearest
neighbor graph is found in the PCA space, which is then used for graph
based clustering such as Louvain and Leiden, which is used to perform
differential expression. Spatial dimension reductions can similarly be
used to perform clustering, to identify spatial regions in the tissue,
as done in <span class="citation">(Shang and Zhou 2022; Ren et al. 2022;
Zhang et al. 2023)</span>. This type of studies often use a manual
segmentation as ground truth to compare different methods that identify
spatial regions.</p>
<p>The problem with this is that spatial region methods are meant to
help us to identify novel spatial regions based on new -omics data,
which might reveal what’s previously unknown from manual annotations. If
the output from a method doesn’t match manual annotations, it might
simply be pointing out a previously unknown aspect of the tissue rather
than wrong. Depending on the questions being asked, there can
simultaneously be multiple spatial partitions. This happens in
geographical space. For instance, there’s land use and neighborhood
boundaries, but equally valid are watershed boundaries and types of rock
formation. Which one is relevant depends on the questions asked.</p>
<p>Here we perform Leiden clustering with non-spatial and MULTISPATI PCA
and compare the results. For the k nearest neighbor graph, I used the
default k = 10.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">29</span><span class="op">)</span></span>
<span>    <span class="va">sfe</span><span class="op">$</span><span class="va">clusts_nonspatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/clusterCells.html" class="external-link">clusterCells</a></span><span class="op">(</span><span class="va">sfe</span>, use.dimred <span class="op">=</span> <span class="st">"PCA"</span>, </span>
<span>                                          BLUSPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/NNGraphParam-class.html" class="external-link">NNGraphParam</a></span><span class="op">(</span></span>
<span>                                              cluster.fun <span class="op">=</span> <span class="st">"leiden"</span>,</span>
<span>                                              cluster.args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>                                                  objective_function <span class="op">=</span> <span class="st">"modularity"</span>,</span>
<span>                                                  resolution_parameter <span class="op">=</span> <span class="fl">1</span></span>
<span>                                              <span class="op">)</span></span>
<span>                                          <span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in (function (to_check, X, clust_centers, clust_info, dtype, nn, :</span></span>
<span><span class="co">#&gt; detected tied distances to neighbors, see ?'BiocNeighbors-ties'</span></span>
<span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt; 802.856   4.460 808.790</span></span></code></pre></div>
<p>See if clustering with the positive MULTISPATI PCs give more
spatially coherent clusters</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">29</span><span class="op">)</span></span>
<span>    <span class="va">sfe</span><span class="op">$</span><span class="va">clusts_multispati</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/clusterRows.html" class="external-link">clusterRows</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"multispati"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span>, </span>
<span>                                          BLUSPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/NNGraphParam-class.html" class="external-link">NNGraphParam</a></span><span class="op">(</span></span>
<span>                                              cluster.fun <span class="op">=</span> <span class="st">"leiden"</span>,</span>
<span>                                              cluster.args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>                                                  objective_function <span class="op">=</span> <span class="st">"modularity"</span>,</span>
<span>                                                  resolution_parameter <span class="op">=</span> <span class="fl">1</span></span>
<span>                                              <span class="op">)</span></span>
<span>                                          <span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in (function (to_check, X, clust_centers, clust_info, dtype, nn, :</span></span>
<span><span class="co">#&gt; detected tied distances to neighbors, see ?'BiocNeighbors-ties'</span></span>
<span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt; 685.932   5.049 692.712</span></span></code></pre></div>
<p>Plot the clusters in space:</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"clusts_nonspatial"</span>, <span class="st">"clusts_multispati"</span><span class="op">)</span>, </span>
<span>                   colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>,</span>
<span>                   scattermore <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>     <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">2</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="multispati_files/figure-html/unnamed-chunk-48-1.png" width="1056"></p>
<p>The MULTISPATI clusters do look somewhat more spatially structured
than clusters from non-spatial PCA. Also zoom into a small area:</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"clusts_nonspatial"</span>, <span class="st">"clusts_multispati"</span><span class="op">)</span>, </span>
<span>                   colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span>,</span>
<span>                   bbox <span class="op">=</span> <span class="va">bbox_use</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>     <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="multispati_files/figure-html/unnamed-chunk-49-1.png" width="1056"></p>
<p>What do these clusters mean? Clusters are supposed to be groups of
different spots that are more similar within a group, sharing some
characteristics. Non-spatial and MULTISPATI PCA use different
characteristics for the clustering. Non-spatial PCA finds genes that are
good for telling cell types apart, although those genes may happen to be
very spatially structured. Non-spatial clustering aims to find these
groups only from gene expression, and cells with similar gene expression
can be surrounded by cells of other types in histological space. This is
just like mapping Art Deco buildings, which are often near Spanish
revival and Beaux Art buildings whose styles are quite different and
perform different functions, thus not necessarily forming a coherent
spatial region.</p>
<p>In contrast, MULTISPATI’s positive components find genes that must
characterize spatial regions in addition to distinguishing between
different cell types. Which genes are involved in each MULTISPATI
component may be as interesting as the clusters. It would be interesting
to perform gene set enrichment analysis, or to interpret this as some
sort of spatial patterns of spatially variable genes. This is like
mapping when the buildings were built, so Art Deco, Spanish revival,
Beaux Art popular in the 1920s and 1930s will end up in the same cluster
and form a more spatially coherent region, which can be found in DTLA
Historical Core and Jewelry District, and Old Pasadena. Hence
non-spatial clustering of spatial data isn’t necessarily bad. Rather, it
tells a different story and reveals different aspects of the data from
spatial clustering.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.2 (2023-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS Ventura 13.6</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] patchwork_1.1.3                sf_1.0-14                     </span></span>
<span><span class="co">#&gt;  [3] BiocParallel_1.36.0            BiocSingular_1.18.0           </span></span>
<span><span class="co">#&gt;  [5] bluster_1.12.0                 tibble_3.2.1                  </span></span>
<span><span class="co">#&gt;  [7] tidyr_1.3.0                    stringr_1.5.1                 </span></span>
<span><span class="co">#&gt;  [9] scran_1.30.0                   scater_1.30.1                 </span></span>
<span><span class="co">#&gt; [11] ggplot2_3.4.4                  scuttle_1.12.0                </span></span>
<span><span class="co">#&gt; [13] SingleCellExperiment_1.24.0    SummarizedExperiment_1.32.0   </span></span>
<span><span class="co">#&gt; [15] Biobase_2.62.0                 GenomicRanges_1.54.1          </span></span>
<span><span class="co">#&gt; [17] GenomeInfoDb_1.38.1            IRanges_2.36.0                </span></span>
<span><span class="co">#&gt; [19] S4Vectors_0.40.2               BiocGenerics_0.48.1           </span></span>
<span><span class="co">#&gt; [21] MatrixGenerics_1.14.0          matrixStats_1.1.0             </span></span>
<span><span class="co">#&gt; [23] SpatialFeatureExperiment_1.3.0 SFEData_1.4.0                 </span></span>
<span><span class="co">#&gt; [25] Voyager_1.4.0                 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] splines_4.3.2                 later_1.3.2                  </span></span>
<span><span class="co">#&gt;   [3] bitops_1.0-7                  filelock_1.0.2               </span></span>
<span><span class="co">#&gt;   [5] lifecycle_1.0.4               edgeR_4.0.2                  </span></span>
<span><span class="co">#&gt;   [7] lattice_0.22-5                magrittr_2.0.3               </span></span>
<span><span class="co">#&gt;   [9] limma_3.58.1                  sass_0.4.8                   </span></span>
<span><span class="co">#&gt;  [11] rmarkdown_2.25                jquerylib_0.1.4              </span></span>
<span><span class="co">#&gt;  [13] yaml_2.3.7                    metapod_1.10.0               </span></span>
<span><span class="co">#&gt;  [15] httpuv_1.6.13                 sp_2.1-2                     </span></span>
<span><span class="co">#&gt;  [17] RColorBrewer_1.1-3            DBI_1.1.3                    </span></span>
<span><span class="co">#&gt;  [19] abind_1.4-5                   zlibbioc_1.48.0              </span></span>
<span><span class="co">#&gt;  [21] purrr_1.0.2                   RCurl_1.98-1.13              </span></span>
<span><span class="co">#&gt;  [23] rappdirs_0.3.3                GenomeInfoDbData_1.2.11      </span></span>
<span><span class="co">#&gt;  [25] ggrepel_0.9.4                 irlba_2.3.5.1                </span></span>
<span><span class="co">#&gt;  [27] terra_1.7-55                  units_0.8-5                  </span></span>
<span><span class="co">#&gt;  [29] RSpectra_0.16-1               dqrng_0.3.2                  </span></span>
<span><span class="co">#&gt;  [31] pkgdown_2.0.7                 DelayedMatrixStats_1.24.0    </span></span>
<span><span class="co">#&gt;  [33] codetools_0.2-19              DelayedArray_0.28.0          </span></span>
<span><span class="co">#&gt;  [35] tidyselect_1.2.0              farver_2.1.1                 </span></span>
<span><span class="co">#&gt;  [37] ScaledMatrix_1.10.0           viridis_0.6.4                </span></span>
<span><span class="co">#&gt;  [39] BiocFileCache_2.10.1          jsonlite_1.8.8               </span></span>
<span><span class="co">#&gt;  [41] BiocNeighbors_1.20.0          e1071_1.7-14                 </span></span>
<span><span class="co">#&gt;  [43] ellipsis_0.3.2                systemfonts_1.0.5            </span></span>
<span><span class="co">#&gt;  [45] tools_4.3.2                   ggnewscale_0.4.9             </span></span>
<span><span class="co">#&gt;  [47] ragg_1.2.6                    Rcpp_1.0.11                  </span></span>
<span><span class="co">#&gt;  [49] glue_1.6.2                    gridExtra_2.3                </span></span>
<span><span class="co">#&gt;  [51] SparseArray_1.2.2             mgcv_1.9-0                   </span></span>
<span><span class="co">#&gt;  [53] xfun_0.41                     dplyr_1.1.4                  </span></span>
<span><span class="co">#&gt;  [55] HDF5Array_1.30.0              withr_2.5.2                  </span></span>
<span><span class="co">#&gt;  [57] BiocManager_1.30.22           fastmap_1.1.1                </span></span>
<span><span class="co">#&gt;  [59] boot_1.3-28.1                 rhdf5filters_1.14.1          </span></span>
<span><span class="co">#&gt;  [61] fansi_1.0.6                   spData_2.3.0                 </span></span>
<span><span class="co">#&gt;  [63] digest_0.6.33                 rsvd_1.0.5                   </span></span>
<span><span class="co">#&gt;  [65] R6_2.5.1                      mime_0.12                    </span></span>
<span><span class="co">#&gt;  [67] textshaping_0.3.7             colorspace_2.1-0             </span></span>
<span><span class="co">#&gt;  [69] wk_0.9.1                      scattermore_1.2              </span></span>
<span><span class="co">#&gt;  [71] RSQLite_2.3.4                 hexbin_1.28.3                </span></span>
<span><span class="co">#&gt;  [73] utf8_1.2.4                    generics_0.1.3               </span></span>
<span><span class="co">#&gt;  [75] class_7.3-22                  httr_1.4.7                   </span></span>
<span><span class="co">#&gt;  [77] S4Arrays_1.2.0                spdep_1.3-1                  </span></span>
<span><span class="co">#&gt;  [79] pkgconfig_2.0.3               scico_1.5.0                  </span></span>
<span><span class="co">#&gt;  [81] gtable_0.3.4                  blob_1.2.4                   </span></span>
<span><span class="co">#&gt;  [83] XVector_0.42.0                htmltools_0.5.7              </span></span>
<span><span class="co">#&gt;  [85] scales_1.3.0                  png_0.1-8                    </span></span>
<span><span class="co">#&gt;  [87] SpatialExperiment_1.12.0      knitr_1.45                   </span></span>
<span><span class="co">#&gt;  [89] rjson_0.2.21                  nlme_3.1-164                 </span></span>
<span><span class="co">#&gt;  [91] curl_5.2.0                    proxy_0.4-27                 </span></span>
<span><span class="co">#&gt;  [93] cachem_1.0.8                  rhdf5_2.46.1                 </span></span>
<span><span class="co">#&gt;  [95] BiocVersion_3.18.1            KernSmooth_2.23-22           </span></span>
<span><span class="co">#&gt;  [97] parallel_4.3.2                vipor_0.4.5                  </span></span>
<span><span class="co">#&gt;  [99] AnnotationDbi_1.64.1          desc_1.4.3                   </span></span>
<span><span class="co">#&gt; [101] s2_1.1.5                      pillar_1.9.0                 </span></span>
<span><span class="co">#&gt; [103] grid_4.3.2                    vctrs_0.6.5                  </span></span>
<span><span class="co">#&gt; [105] promises_1.2.1                dbplyr_2.4.0                 </span></span>
<span><span class="co">#&gt; [107] beachmat_2.18.0               xtable_1.8-4                 </span></span>
<span><span class="co">#&gt; [109] cluster_2.1.6                 beeswarm_0.4.0               </span></span>
<span><span class="co">#&gt; [111] evaluate_0.23                 magick_2.8.1                 </span></span>
<span><span class="co">#&gt; [113] cli_3.6.1                     locfit_1.5-9.8               </span></span>
<span><span class="co">#&gt; [115] compiler_4.3.2                rlang_1.1.2                  </span></span>
<span><span class="co">#&gt; [117] crayon_1.5.2                  labeling_0.4.3               </span></span>
<span><span class="co">#&gt; [119] classInt_0.4-10               fs_1.6.3                     </span></span>
<span><span class="co">#&gt; [121] ggbeeswarm_0.7.2              stringi_1.8.2                </span></span>
<span><span class="co">#&gt; [123] viridisLite_0.4.2             deldir_2.0-2                 </span></span>
<span><span class="co">#&gt; [125] munsell_0.5.0                 Biostrings_2.70.1            </span></span>
<span><span class="co">#&gt; [127] Matrix_1.6-4                  ExperimentHub_2.10.0         </span></span>
<span><span class="co">#&gt; [129] sparseMatrixStats_1.14.0      bit64_4.0.5                  </span></span>
<span><span class="co">#&gt; [131] Rhdf5lib_1.24.0               KEGGREST_1.42.0              </span></span>
<span><span class="co">#&gt; [133] statmod_1.5.0                 shiny_1.8.0                  </span></span>
<span><span class="co">#&gt; [135] highr_0.10                    interactiveDisplayBase_1.40.0</span></span>
<span><span class="co">#&gt; [137] AnnotationHub_3.10.0          igraph_1.5.1                 </span></span>
<span><span class="co">#&gt; [139] memoise_2.0.1                 bslib_0.6.1                  </span></span>
<span><span class="co">#&gt; [141] bit_4.0.5</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Bonnardel2019-le" class="csl-entry">
Bonnardel, Johnny, Wouter T’Jonck, Djoere Gaublomme, Robin Browaeys,
Charlotte L Scott, Liesbet Martens, Bavo Vanneste, et al. 2019.
<span>“Stellate Cells, Hepatocytes, and Endothelial Cells Imprint the
Kupffer Cell Identity on Monocytes Colonizing the Liver Macrophage
Niche.”</span> <em>Immunity</em> 51 (4): 638–654.e9.
</div>
<div id="ref-Dray2008-en" class="csl-entry">
Dray, Stéphane, Sonia Saı̈d, and Françis Débias. 2008. <span>“Spatial
Ordination of Vegetation Data Using a Generalization of Wartenberg’s
Multivariate Spatial Correlation.”</span> <em>J. Veg. Sci.</em> 19 (1):
45–56.
</div>
<div id="ref-Griffith2019-bo" class="csl-entry">
Griffith, Daniel A. 2019. <span>“Negative Spatial Autocorrelation: One
of the Most Neglected Concepts in Spatial Statistics.”</span>
<em>Stats</em> 2 (3): 388–415.
</div>
<div id="ref-Halpern2017-lq" class="csl-entry">
Halpern, Keren Bahar, Rom Shenhav, Orit Matcovitch-Natan, Beata Toth,
Doron Lemze, Matan Golan, Efi E Massasa, et al. 2017. <span>“Single-Cell
Spatial Reconstruction Reveals Global Division of Labour in the
Mammalian Liver.”</span> <em>Nature</em> 542 (7641): 352–56.
</div>
<div id="ref-Ren2022-qx" class="csl-entry">
Ren, Honglei, Benjamin L Walker, Zixuan Cang, and Qing Nie. 2022.
<span>“Identifying Multicellular Spatiotemporal Organization of Cells
with <span>SpaceFlow</span>.”</span> <em>Nat. Commun.</em> 13 (1): 4076.
</div>
<div id="ref-Shang2022-qy" class="csl-entry">
Shang, Lulu, and Xiang Zhou. 2022. <span>“Spatially Aware Dimension
Reduction for Spatial Transcriptomics.”</span> <em>Nat. Commun.</em> 13
(1): 7203.
</div>
<div id="ref-Townes2023-bi" class="csl-entry">
Townes, F William, and Barbara E Engelhardt. 2023. <span>“Nonnegative
Spatial Factorization Applied to Spatial Genomics.”</span> <em>Nat.
Methods</em> 20 (2): 229–38.
</div>
<div id="ref-Velten2022-gv" class="csl-entry">
Velten, Britta, Jana M Braunger, Ricard Argelaguet, Damien Arnol, Jakob
Wirbel, Danila Bredikhin, Georg Zeller, and Oliver Stegle. 2022.
<span>“Identifying Temporal and Spatial Patterns of Variation from
Multimodal Data Using <span>MEFISTO</span>.”</span> <em>Nat.
Methods</em> 19 (2): 179–86.
</div>
<div id="ref-Wartenberg1985-fk" class="csl-entry">
Wartenberg, Daniel. 1985. <span>“Multivariate Spatial Correlation: A
Method for Exploratory Geographical Analysis.”</span> <em>Geogr.
Anal.</em> 17 (4): 263–83.
</div>
<div id="ref-Zhang2023-kf" class="csl-entry">
Zhang, Chuanchao, Xinxing Li, Wendong Huang, Lequn Wang, and Qianqian
Shi. 2023. <span>“Spatially Aware Self-Representation Learning for
Tissue Structure Characterization and Spatial Functional Genes
Identification.”</span> <em>bioRxiv</em>.
</div>
<div id="ref-Zong2022-tb" class="csl-entry">
Zong, Yongshuo, Tingyang Yu, Xuesong Wang, Yixuan Wang, Zhihang Hu, and
Yu Li. 2022. <span>“<span class="nocase">conST</span>: An Interpretable
Multi-Modal Contrastive Learning Framework for Spatial
Transcriptomics.”</span> <em>bioRxiv</em>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Lambda Moses, Kayla Jackson, Laura Luebbert, Sina Booeshaghi, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
