<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Voyager">
<title>Basic Visium exploratory data analysis • Voyager</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Basic Visium exploratory data analysis">
<meta property="og:description" content="Voyager">
<meta name="robots" content="noindex">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">Voyager</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">1.0.7</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/install.html">Install</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-technologies">Technologies</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-technologies">
    <a class="dropdown-item" href="../articles/visium_landing.html">Visium</a>
    <a class="dropdown-item" href="../articles/slideseqV2_landing.html">Slide-seq v2</a>
    <a class="dropdown-item" href="../articles/cosmx_landing.html">CosMX</a>
    <a class="dropdown-item" href="../articles/xenium_landing.html">Xenium</a>
    <a class="dropdown-item" href="../articles/merfish_landing.html">MERFISH</a>
    <a class="dropdown-item" href="../articles/seqfish_landing.html">seqFISH</a>
    <a class="dropdown-item" href="../articles/chromium_landing.html">10X Chromium V3</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/pachterlab/voyager/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Basic Visium exploratory data analysis</h1>
                        <h4 data-toc-skip class="author">Lambda
Moses</h4>
            
            <h4 data-toc-skip class="date">2023-01-21</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pachterlab/voyager/blob/HEAD/vignettes/vig1_visium_basic.Rmd" class="external-link"><code>vignettes/vig1_visium_basic.Rmd</code></a></small>
      <div class="d-none name"><code>vig1_visium_basic.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In this introductory vignette for <a href="https://bioconductor.org/packages/devel/bioc/html/SpatialFeatureExperiment.html" class="external-link"><code>SpatialFeatureExperiment</code></a>
data representation and <a href="https://bioconductor.org/packages/devel/bioc/html/Voyager.html" class="external-link"><code>Voyager</code></a>
anlaysis package, we demonstrate a basic exploratory data analysis (EDA)
of spatial transcriptomics data. Basic knowledge of R and <a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html" class="external-link"><code>SingleCellExperiment</code></a>
is assumed.</p>
<p>This vignette showcases the packages with a Visium spatial gene
expression system dataset. The technology was chosen due to its
popularity, and therefore the availability of numerous publicly
available datasets for analysis <span class="citation">(Moses and
Pachter 2022)</span>. The code can be run on <a href="https://colab.research.google.com/drive/1EPbPu71O7mvc1Hsp1VI3Jc3xI5mzJ0eM?usp=sharing" class="external-link">Google
Colab</a>.</p>
<p><img src="https://pachterlab.github.io/LP_2021/04-current_files/figure-html/n-insts-1.png" alt='Bar chart showing the number of institutions using each spatial transcriptomics data collection method. Only methods used by at least 3 different institutions are shown. The bars are colored by category of the methods. 10X Visium which is based on sequence barcoding is used by over 100 institutions. Following Visium are GeoMX DSP and GeoMX WTA which are "regions of interest" (ROI) methods, along with 2016 ST and some other ROI selection methods.' style="display: block; margin: auto;"></p>
<p>While Voyager was developed with the goal of facilitating the use of
geospatial methods in spatial genomics, this introductory vignette is
restricted to non-spatial scRNA-seq EDA with the Visium dataset. For a
vignette illustrating univariate spatial analysis with the same dataset,
see the more advanced <a href="https://pachterlab.github.io/voyager/articles/vig2_visium.html" class="external-link">exploratory
spatial data analyis vignette</a> with the same dataset.</p>
<p>Here we load the packages used in this vignette.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/voyager" class="external-link">Voyager</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SpatialFeatureExperiment" class="external-link">SpatialFeatureExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: SummarizedExperiment</span></span>
<span><span class="co">#&gt; Loading required package: MatrixGenerics</span></span>
<span><span class="co">#&gt; Loading required package: matrixStats</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'MatrixGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span></span>
<span><span class="co">#&gt;     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span></span>
<span><span class="co">#&gt;     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span></span>
<span><span class="co">#&gt;     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span></span>
<span><span class="co">#&gt;     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span></span>
<span><span class="co">#&gt;     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span></span>
<span><span class="co">#&gt;     colWeightedMeans, colWeightedMedians, colWeightedSds,</span></span>
<span><span class="co">#&gt;     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span></span>
<span><span class="co">#&gt;     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span></span>
<span><span class="co">#&gt;     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span></span>
<span><span class="co">#&gt;     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span></span>
<span><span class="co">#&gt;     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span></span>
<span><span class="co">#&gt;     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span></span>
<span><span class="co">#&gt;     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span></span>
<span><span class="co">#&gt;     rowWeightedSds, rowWeightedVars</span></span>
<span><span class="co">#&gt; Loading required package: GenomicRanges</span></span>
<span><span class="co">#&gt; Loading required package: stats4</span></span>
<span><span class="co">#&gt; Loading required package: BiocGenerics</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'BiocGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">#&gt;     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">#&gt;     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,</span></span>
<span><span class="co">#&gt;     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span></span>
<span><span class="co">#&gt;     Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,</span></span>
<span><span class="co">#&gt;     table, tapply, union, unique, unsplit, which.max, which.min</span></span>
<span><span class="co">#&gt; Loading required package: S4Vectors</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'S4Vectors'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     expand.grid, I, unname</span></span>
<span><span class="co">#&gt; Loading required package: IRanges</span></span>
<span><span class="co">#&gt; Loading required package: GenomeInfoDb</span></span>
<span><span class="co">#&gt; Loading required package: Biobase</span></span>
<span><span class="co">#&gt; Welcome to Bioconductor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     Vignettes contain introductory material; view with</span></span>
<span><span class="co">#&gt;     'browseVignettes()'. To cite Bioconductor, see</span></span>
<span><span class="co">#&gt;     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'Biobase'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:MatrixGenerics':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     rowMedians</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyMissing, rowMedians</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drighelli/SpatialExperiment" class="external-link">SpatialExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: scuttle</span></span>
<span><span class="co">#&gt; Loading required package: ggplot2</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SFEData" class="external-link">SFEData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org" class="external-link">stringr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/const-ae/sparseMatrixStats" class="external-link">sparseMatrixStats</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="mouse-skeletal-muscle-dataset">Mouse skeletal muscle dataset<a class="anchor" aria-label="anchor" href="#mouse-skeletal-muscle-dataset"></a>
</h2>
<p>The dataset used in this vignette is from the paper <a href="https://doi.org/10.1038/s42003-021-02810-x" class="external-link">Large-scale
integration of single-cell transcriptomic data captures transitional
progenitor states in mouse skeletal muscle regeneration</a> <span class="citation">(McKellar et al. 2021)</span>. Notexin was injected
into the tibialis anterior muscle of mice to induce injury, and the
healing muscle was collected 2, 5, and 7 days post injury for Visium
analysis. The dataset in this vignette is from the timepoint at day 2.
The vignette starts with a <code>SpatialFeatureExperiment</code> (SFE)
object.</p>
<p>The gene count matrix was directly downloaded <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM4904759" class="external-link">from
GEO</a>. All 4992 spots, whether in tissue or not, are included. The
tissue boundary was found by thresholding the H&amp;E image in OpenCV,
and small polygons were removed as they are likely to be debris. Spot
polygons were constructed with the spot centroid coordinates and
diameter in the Space Ranger output. The <code>in_tissue</code> column
in <code>colData</code> indicates which spot polygons intersect the
tissue polygons, and is based on <code><a href="https://r-spatial.github.io/sf/reference/geos_binary_pred.html" class="external-link">st_intersects()</a></code>.</p>
<p>Tissue boundary, nuclei, myofiber, and Visium spot polygons are
stored as <code>sf</code> data frames in the SFE object. The Visium spot
polygons are called “spotPoly” in this SFE object. The
<code>SpatialFeatureExperiment</code> package has a few convenience
wrappers to get and set common types of geometries, including
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/dimGeometries.html" class="external-link">spotPoly()</a></code> for Visium (or other technologies when relevant)
spot polygons, <code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/dimGeometries.html" class="external-link">cellSeg()</a></code> for cell segmentation,
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/dimGeometries.html" class="external-link">nucSeg()</a></code> for nuclei segmentation, and
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/dimGeometries.html" class="external-link">centroids()</a></code> for cell centroids. Behind the scene are
specially named <code>sf</code> data frames. See <a href="https://bioconductor.org/packages/devel/bioc/vignettes/SpatialFeatureExperiment/inst/doc/SFE.html" class="external-link">the
vignette of <code>SpatialFeatureExperiment</code></a> for more details
on the structure of the SFE object.</p>
<p>The SFE object of this dataset is provided in the
<code>SFEData</code> package; we begin by downloading the data and
loading it into R.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/McKellarMuscleData.html" class="external-link">McKellarMuscleData</a></span><span class="op">(</span><span class="st">"full"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; snapshotDate(): 2022-10-31</span></span>
<span><span class="co">#&gt; see ?SFEData and browseVignettes('SFEData') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 15123 4992 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(15123): ENSMUSG00000025902 ENSMUSG00000096126 ...</span></span>
<span><span class="co">#&gt;   ENSMUSG00000064368 ENSMUSG00000064370</span></span>
<span><span class="co">#&gt; rowData names(6): Ensembl symbol ... vars cv2</span></span>
<span><span class="co">#&gt; colnames(4992): AAACAACGAATAGTTC AAACAAGTATCTCCCA ... TTGTTTGTATTACACG</span></span>
<span><span class="co">#&gt;   TTGTTTGTGTAAATTC</span></span>
<span><span class="co">#&gt; colData names(12): barcode col ... prop_mito in_tissue</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : imageX imageY</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: spotPoly (POLYGON) </span></span>
<span><span class="co">#&gt; annotGeometries: tissueBoundary (POLYGON), myofiber_full (POLYGON), myofiber_simplified (POLYGON), nuclei (POLYGON), nuclei_centroid (POINT) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; Vis5A:</span></span></code></pre></div>
<p>The authors provided the full resolution hematoxylin and eosin
(H&amp;E) image on GEO, which we downsized to facilitate its display:
<img src="tissue_lowres_5a.jpeg" alt="A cross section of mouse muscle is slightly off center to the lower left. In the middle of the tissue is the notexin injury site with leukocyte infiltration and fewer myofibers. The rest of the tissue section is tightly packed with myofibers." width="100%" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="quality-control">Quality control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h2>
<div class="section level3">
<h3 id="spots">Spots<a class="anchor" aria-label="anchor" href="#spots"></a>
</h3>
<p>We begin quality control (QC) by plotting various metrics both as
violin plots and in space. The QC metrics are pre-computed and stored in
<code>colData</code> (for spots) and <code>rowData</code> of the SFE
object.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "barcode"   "col"       "row"       "x"         "y"         "dia"      </span></span>
<span><span class="co">#&gt;  [7] "tissue"    "sample_id" "nCounts"   "nGenes"    "prop_mito" "in_tissue"</span></span></code></pre></div>
<p>Below we plot the total unique molecular identifier (UMI) counts per
spot. The commented out line of code shows how to compute the total UMI
counts.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># colData(sfe)$nCounts &lt;- colSums(counts(sfe))</span></span>
<span><span class="va">violin</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nCounts"</span>, x <span class="op">=</span> <span class="st">"in_tissue"</span>, colour_by <span class="op">=</span> <span class="st">"in_tissue"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"top"</span><span class="op">)</span></span>
<span><span class="va">spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nCounts"</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>,</span>
<span>                              annotGeometryName <span class="op">=</span> <span class="st">"tissueBoundary"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">violin</span> <span class="op">+</span> <span class="va">spatial</span></span></code></pre></div>
<p><img src="vig1_visium_basic_files/figure-html/unnamed-chunk-6-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Some spots in the injury site with leukocyte infiltration have high
total counts. Spatial autocorrelation of the total counts is apparent,
which will be discussed in a later section of this vignette.</p>
<p>Next we find number of genes detected per spot. The commented out
line of code shows how to find the number of genes detected.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># colData(sfe)$nGenes &lt;- colSums(counts(sfe) &gt; 0)</span></span>
<span><span class="va">violin</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nGenes"</span>, x <span class="op">=</span> <span class="st">"in_tissue"</span>, colour_by <span class="op">=</span> <span class="st">"in_tissue"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"top"</span><span class="op">)</span></span>
<span><span class="va">spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nGenes"</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>,</span>
<span>                              annotGeometryName <span class="op">=</span> <span class="st">"tissueBoundary"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">violin</span> <span class="op">+</span> <span class="va">spatial</span></span></code></pre></div>
<p><img src="vig1_visium_basic_files/figure-html/unnamed-chunk-7-1.png" width="700" style="display: block; margin: auto;"></p>
<p>As commonly done for scRNA-seq data, here we plot nCounts
vs. nGenes</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe</span>, x <span class="op">=</span> <span class="st">"nCounts"</span>, y <span class="op">=</span> <span class="st">"nGenes"</span>, colour_by <span class="op">=</span> <span class="st">"in_tissue"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig1_visium_basic_files/figure-html/unnamed-chunk-8-1.png" width="700" style="display: block; margin: auto;"></p>
<p>This plot has two branches for the spots in tissue, which turn out to
be related to myofiber size. See <a href="https://pachterlab.github.io/voyager/articles/vig2_visium.html" class="external-link">the
exploratory spatial data analysis (ESDA) Visium vignette</a>.</p>
<p>As is commonly done for scRNA-seq data, we plot the proportion of
mitochondrially encoded counts. The commented out code shows how to find
this proportion:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># mito_ind &lt;- str_detect(rowData(sfe)$symbol, "^Mt-")</span></span>
<span><span class="co"># colData(sfe)$prop_mito &lt;- colSums(counts(sfe)[mito_ind,]) / colData(sfe)$nCounts</span></span>
<span><span class="va">violin</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"prop_mito"</span>, x <span class="op">=</span> <span class="st">"in_tissue"</span>, colour_by <span class="op">=</span> <span class="st">"in_tissue"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"top"</span><span class="op">)</span></span>
<span><span class="va">spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"prop_mito"</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>,</span>
<span>                              annotGeometryName <span class="op">=</span> <span class="st">"tissueBoundary"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">violin</span> <span class="op">+</span> <span class="va">spatial</span></span></code></pre></div>
<p><img src="vig1_visium_basic_files/figure-html/unnamed-chunk-9-1.png" width="700" style="display: block; margin: auto;"></p>
<p>As expected, spots outside tissue have a higher proportion of
mitochondrial counts, because when the tissue is lysed, mitochondrial
transcripts are are less likely to degrade than cytosolic transcripts as
they are protected by a double membrane. However, spots on myofibers
also have a high proportion of mitochondrial counts, because of the
function of myofibers. The injury site with leukocyte infiltration has a
lower proportion of mitochondrial counts.</p>
<p>To see the relationship between the proportion of mitochondrial
counts and total UMI counts, we plot them against each other as is
commonly done in scRNA-seq analysis to identify low quality cells,
i.e. cells with few UMI counts and a high proportion of mitochondrial
counts.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe</span>, x <span class="op">=</span> <span class="st">"nCounts"</span>, y <span class="op">=</span> <span class="st">"prop_mito"</span>, colour_by <span class="op">=</span> <span class="st">"in_tissue"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig1_visium_basic_files/figure-html/unnamed-chunk-10-1.png" width="700" style="display: block; margin: auto;"></p>
<p>There are two clusters for the spots in tissue, which also turn out
to be related to myofiber size. See the <a href="https://pachterlab.github.io/voyager/articles/vig2_visium.html" class="external-link">ESDA
Visium vignette</a>.</p>
<p>So far we haven’t seen spots that are obvious outliers in these QC
metrics.</p>
<p>The following analyses only use spots in tissue, which are selected
as follows:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">in_tissue</span><span class="op">]</span></span>
<span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="va">sfe_tissue</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>,<span class="op">]</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="genes">Genes<a class="anchor" aria-label="anchor" href="#genes"></a>
</h3>
<p>As in scRNA-seq, gene expression variance in Visium measurements is
overdispersed compared to variance of counts that are Poisson
distributed.</p>
<p>To understand the mean-variance relationship, we compute the mean,
variance, and coefficient of variance (CV2) for each gene among spots in
tissue:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">means</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">vars</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowVars.html" class="external-link">rowVars</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Coefficient of variance</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">cv2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">vars</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">means</span><span class="op">^</span><span class="fl">2</span></span></code></pre></div>
<p>To avoid overplotting and better show point density on the plot, we
use a 2D histogram. The color of each bin indicates the number of points
in that bin.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotColDataBin2D.html">plotRowDataBin2D</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"means"</span>, <span class="st">"vars"</span>, bins <span class="op">=</span> <span class="fl">50</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>slope <span class="op">=</span> <span class="fl">1</span>, intercept <span class="op">=</span> <span class="fl">0</span>, color <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotation_logticks.html" class="external-link">annotation_logticks</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_equal</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig1_visium_basic_files/figure-html/unnamed-chunk-13-1.png" width="700" style="display: block; margin: auto;"></p>
<p>The red line, <span class="math inline">\(y = x\)</span> is what is
expected for Poisson distributed data, but we find that the variance is
higher for more highly expressed genes than expected from Poisson
distributed counts. The coefficient of variation shows the same.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotColDataBin2D.html">plotRowDataBin2D</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"means"</span>, <span class="st">"cv2"</span>, bins <span class="op">=</span> <span class="fl">50</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>slope <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, intercept <span class="op">=</span> <span class="fl">0</span>, color <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotation_logticks.html" class="external-link">annotation_logticks</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_equal</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig1_visium_basic_files/figure-html/unnamed-chunk-14-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2">
<h2 id="normalize-data">Normalize data<a class="anchor" aria-label="anchor" href="#normalize-data"></a>
</h2>
<p>We demonstrate the use of <code>scater</code> for normalization
below, although we note that it is not necessarily the best approach to
normalizing spatial transcriptomics data. The problem of when and how to
normalize spatial transcriptomics data is non-trivial because, as the
<code>nCounts</code> plot in space shows above, spatial autocorrelation
is evident. Furthemrore, in Visium, reverse transcription occurs in situ
on the spots, but PCR amplification occurs after the cDNA is dissociated
from the spots. Artifacts may be subsequently introduced from the
amplification step, and these would not be associated with spatial
origin. Spatial artifacts may arise from the diffusion of transcripts
and tissue permeablization. However, given how the total counts seem to
correspond to histological regions, the total counts may have a
biological component and hence should not be treated as a technical
artifact to be normalized away as in scRNA-seq data normalization
methods. In other words, the issue of normalization for spatial
transcriptomics data, and Visium in particular, is complex and is
currently unsolved.</p>
<p>There is no one way to normalize non-spatial scRNA-seq data. The
commented out code implements the <code>scran</code> method <span class="citation">(Lun, Bach, and Marioni 2016)</span>. To simplify the
matter, we only perform <code><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts()</a></code> in this
introductory vignette.</p>
<p>Note that <code>scater</code>’s <code><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts()</a></code> is quite
different from that in Seurat. Let <span class="math inline">\(N\)</span> denote the total UMI count in one
Visium spot, <span class="math inline">\(\bar N\)</span> the average
total UMI count in all spots in this dataset, and <span class="math inline">\(x\)</span> denote the UMI count of one gene in the
Visium spot of interest. Seurat performs log normalization as <span class="math inline">\(\mathrm{log}\left( \frac{x}{N/10000} + 1
\right)\)</span>, where the natural log is used. In contrast, with
default parameters, <code>scater</code> uses <span class="math inline">\(\mathrm{log_2}\left( \frac{x}{N/\bar N} + 1
\right)\)</span>. The pseudocount (default to 1), library size factors
(default to <span class="math inline">\(N/\bar N\)</span>), and
transform (default to log2) can be changed. Log 2 is used because
differences in values can be interpreted as log fold change.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># clusters &lt;- quickCluster(sfe_tissue)</span></span>
<span><span class="co"># sfe_tissue &lt;- computeSumFactors(sfe_tissue, clusters=clusters)</span></span>
<span><span class="co"># sfe_tissue &lt;- sfe_tissue[, sizeFactors(sfe_tissue) &gt; 0]</span></span>
<span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span></code></pre></div>
<p>Next, we identify highly variable genes (HVGs), which will be used
for principal component analysis (PCA) dimensionality reduction. Again,
there are different ways to identify HVGs, and <code>scater</code> does
so differently from Seurat. In both frameworks, the log normalized data
is used by default. In summary, in Seurat, with default parameters, a
Loess curve is fitted to the log transformed data (the log normalized
data is log transformed for fitting purposes), and the fitted values are
exponentiated as the expected variance for each gene. Then the expected
variance and the mean are used to standardize the log normalized gene
expression; the standardized values are used to calculate a standardized
variance for each gene. The top HVGs are the genes with the largest
standardized variance.</p>
<p>In <code>scater</code>, with default parameters, a parametric
non-linear curve of variance vs. mean for each gene of the log
normalized data. Then the log ratio of the actual variance to the fitted
variance from the curve calculated, and a Loess curve is fitted to this
log ratio vs. mean for each gene. The “technical” component of the
variance is the fitted values from the Loess curve. The “biological”
component is the difference between the actual variance and the Loess
fitted variance. The top HVGs are genes with the largest biological
component. See the documentation of <code><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar()</a></code>,
<code><a href="https://rdrr.io/pkg/scran/man/fitTrendVar.html" class="external-link">fitTrendVar()</a></code>, and <code><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs()</a></code> for more
details.</p>
<p>These differences can lead to different downstream results. While we
don’t comment on which way is better in this vignette, it’s important to
be aware of such differences.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span>
<span><span class="va">hvgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">dec</span>, n <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="dimension-reduction-and-clustering">Dimension reduction and clustering<a class="anchor" aria-label="anchor" href="#dimension-reduction-and-clustering"></a>
</h2>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, ncomponents <span class="op">=</span> <span class="fl">30</span>, subset_row <span class="op">=</span> <span class="va">hvgs</span>,</span>
<span>                     scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># scale as in Seurat</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ElbowPlot.html">ElbowPlot</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, ndims <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig1_visium_basic_files/figure-html/unnamed-chunk-18-1.png" width="700" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotDimLoadings.html">plotDimLoadings</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig1_visium_basic_files/figure-html/unnamed-chunk-19-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Do the clustering to show on the dimension reduction plots</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/clusterRows.html" class="external-link">clusterRows</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"PCA"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span>,</span>
<span>                                           BLUSPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/NNGraphParam-class.html" class="external-link">SNNGraphParam</a></span><span class="op">(</span></span>
<span>                                               cluster.fun <span class="op">=</span> <span class="st">"leiden"</span>,</span>
<span>                                               cluster.args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>                                                   resolution_parameter <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                                                   objective_function <span class="op">=</span> <span class="st">"modularity"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, ncomponents <span class="op">=</span> <span class="fl">3</span>, colour_by <span class="op">=</span> <span class="st">"cluster"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig1_visium_basic_files/figure-html/unnamed-chunk-21-1.png" width="672" style="display: block; margin: auto;"></p>
<p>The principal components (PCs) can be plotted in space. Due to
spatial autocorrelation of many genes and spatial regions of different
histological characters, even though spatial information is not used in
the PCA procedure, the PCs may show spatial structure.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/spatialReducedDim.html">spatialReducedDim</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"PCA"</span>, ncomponents <span class="op">=</span> <span class="fl">4</span>, </span>
<span>                  colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, divergent <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                  diverge_center <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig1_visium_basic_files/figure-html/unnamed-chunk-22-1.png" width="576" style="display: block; margin: auto;"></p>
<p>PC1, which explains far more variance than PC2, separates the injury
site leukocytes and myofibers close to the site from the visium
myofibers. PC2 highlight some myofibers near the edge. PC3 highlights
the muscle tendon junctions. PC4 does not seem to be informative; it
might have picked up an outlier.</p>
<p>It is also possible to run UMAP following the PCA, as is done for
scRNA-seq. We do not recommend producing a UMAP since the procedure
distorts distances, and does not respect either local or global
structure in the data <span class="citation">(Chari, Banerjee, and
Pachter 2021)</span>. However, for completeness, we show how to compute
a UMAP below:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">29</span><span class="op">)</span></span>
<span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, dimred <span class="op">=</span> <span class="st">"PCA"</span>, n_dimred <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, colour_by <span class="op">=</span> <span class="st">"cluster"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig1_visium_basic_files/figure-html/unnamed-chunk-24-1.png" width="700" style="display: block; margin: auto;"></p>
<p>More interesting in spatial transcriptomics, is to locate the
clusters in space, and this can be done as follows:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"cluster"</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig1_visium_basic_files/figure-html/unnamed-chunk-25-1.png" width="700" style="display: block; margin: auto;"></p>
<p>While spatial information is not explicitly used in clustering, due
to spatial autocorrelation of gene expression and the histological
regions, some of these clusters are spatially contiguous. There are many
methods to find spatially informed clusters, such as <a href="https://bioconductor.org/packages/release/bioc/html/BayesSpace.html" class="external-link"><code>BayesSpace</code></a>
<span class="citation">(E. Zhao et al. 2021)</span>, which is on
Bioconductor.</p>
<p>Remark on spatial regions: In geographical space, there is usually no
one single way to define spatial regions. For example, influenced by
both sociology and geology, LA county can be partitioned into regions
such as Eastside, Westside, South Central, San Fernado Valley, San
Gabriel Valley, Pomona Valley, Gateway Cities, South Bay, and etc., each
containing multiple smaller cities or parts of LA City, each of which
can be further divided into many neighborhoods, such as Koreatown,
Highland Park, Lincoln Heights, and etc. Definitions of some of these
regions are subject to dispute. Meanwhile, LA county can also be
partitioned into watersheds of the LA River, San Gabriel River, Ballona
Creek, and etc., as well as different rock formations. Which kind of
spatial region at which resolution is relevant depends on the question
being asked. There are also gray areas in spatial regions. For example,
the Whittier Narrows dam intercepts both the San Gabriel River and Rio
Hondo (a large tributary of the LA River), so whether the dam area
belongs to the watershed of San Gabriel River or LA River is
unclear.</p>
<p>Similarly, in spatial transcriptomics, while methods identifying
spatial regions currently generally only aim to give one result,
multiple results at different resolutions depending on the question
asked may be relevant. Furthermore, methods for spatial region
demarcation to be used for spatial -omics would ideally provide
uncertainty assessments for assignment of cells or Visium spots. An
existing geospatial method that accounts for such uncertainty is <a href="https://cran.r-project.org/web/packages/geocmeans/" class="external-link"><code>geocmeans</code></a>
<span class="citation">(F. Zhao, Jiao, and Liu 2013)</span>, which is on
CRAN.</p>
<p>In both the geographical and histological space, there conflicting
views on spatial variation. On the one hand, methods that identify
spatially variable genes such as SpatialDE often assume that gene
expression vary smoothly and continuously in space. On the other hand,
methods identifying spatial regions attempt to identify discrete
regions. The continuous variation in features might be why definitions
of geographical neighborhoods are often subject to dispute. Some
existing methods attempt to harmonize the two views. For example, the
spatially variable gene method <a href="https://github.com/raphael-group/belayer" class="external-link"><code>belayer</code></a>
<span class="citation">(Ma et al. 2022)</span> takes discrete tissue
layers into account.</p>
</div>
<div class="section level2">
<h2 id="non-spatial-differential-expression">Non-spatial differential expression<a class="anchor" aria-label="anchor" href="#non-spatial-differential-expression"></a>
</h2>
<p>Cluster marker genes can be found using differential analysis methods
as is commonly done for scRNA-seq. Below is an example with the Wilcoxon
rank sum test:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/findMarkers.html" class="external-link">findMarkers</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span>,</span>
<span>                       test.type <span class="op">=</span> <span class="st">"wilcox"</span>, pval.type <span class="op">=</span> <span class="st">"all"</span>, direction <span class="op">=</span> <span class="st">"up"</span><span class="op">)</span></span></code></pre></div>
<p>The result is sorted by p-values:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">markers</span><span class="op">[[</span><span class="fl">6</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt; DataFrame with 15043 rows and 8 columns</span></span>
<span><span class="co">#&gt;                        p.value         FDR summary.AUC     AUC.1     AUC.2</span></span>
<span><span class="co">#&gt;                      &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; ENSMUSG00000098178 1.33968e-10 2.01527e-06    0.736941  0.736941  0.932721</span></span>
<span><span class="co">#&gt; ENSMUSG00000033707 6.35679e-02 1.00000e+00    0.510526  0.510526  0.510526</span></span>
<span><span class="co">#&gt; ENSMUSG00000028031 1.04324e-01 1.00000e+00    0.521479  0.533069  0.539945</span></span>
<span><span class="co">#&gt; ENSMUSG00000049353 1.42040e-01 1.00000e+00    0.505263  0.505263  0.505263</span></span>
<span><span class="co">#&gt; ENSMUSG00000045319 1.42040e-01 1.00000e+00    0.505263  0.505263  0.505263</span></span>
<span><span class="co">#&gt; ...                        ...         ...         ...       ...       ...</span></span>
<span><span class="co">#&gt; ENSMUSG00000043969           1           1         0.5       0.5  0.500000</span></span>
<span><span class="co">#&gt; ENSMUSG00000091378           1           1         0.5       0.5  0.500000</span></span>
<span><span class="co">#&gt; ENSMUSG00000072437           1           1         0.5       0.5  0.500000</span></span>
<span><span class="co">#&gt; ENSMUSG00000003228           1           1         0.5       0.5  0.495283</span></span>
<span><span class="co">#&gt; ENSMUSG00000094649           1           1         0.5       0.5  0.497642</span></span>
<span><span class="co">#&gt;                        AUC.3     AUC.4     AUC.5</span></span>
<span><span class="co">#&gt;                    &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; ENSMUSG00000098178  0.775869  0.910004  0.861166</span></span>
<span><span class="co">#&gt; ENSMUSG00000033707  0.510526  0.510526  0.510526</span></span>
<span><span class="co">#&gt; ENSMUSG00000028031  0.527655  0.535115  0.521479</span></span>
<span><span class="co">#&gt; ENSMUSG00000049353  0.505263  0.505263  0.505263</span></span>
<span><span class="co">#&gt; ENSMUSG00000045319  0.505263  0.505263  0.505263</span></span>
<span><span class="co">#&gt; ...                      ...       ...       ...</span></span>
<span><span class="co">#&gt; ENSMUSG00000043969  0.500000  0.496183  0.500000</span></span>
<span><span class="co">#&gt; ENSMUSG00000091378  0.500000  0.496183  0.500000</span></span>
<span><span class="co">#&gt; ENSMUSG00000072437  0.500000  0.492366  0.500000</span></span>
<span><span class="co">#&gt; ENSMUSG00000003228  0.477679  0.473282  0.490991</span></span>
<span><span class="co">#&gt; ENSMUSG00000094649  0.500000  0.500000  0.500000</span></span></code></pre></div>
<p>Significant markers for each cluster can be obtained as follows:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">vapply</a></span><span class="op">(</span><span class="va">markers</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, FUN.VALUE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotExpression.html" class="external-link">plotExpression</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">[</span><span class="va">genes_use</span>, <span class="st">"symbol"</span><span class="op">]</span>, x <span class="op">=</span> <span class="st">"cluster"</span>,</span>
<span>               colour_by <span class="op">=</span> <span class="st">"cluster"</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig1_visium_basic_files/figure-html/unnamed-chunk-28-1.png" width="700" style="display: block; margin: auto;"></p>
<p>These genes are interesting to view in spatial context:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="va">genes_use</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig1_visium_basic_files/figure-html/unnamed-chunk-29-1.png" width="864" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="morans-i">Moran’s I<a class="anchor" aria-label="anchor" href="#morans-i"></a>
</h2>
<p>Tobler’s first law of geography states that</p>
<blockquote>
<p>Everything is related to everything else. But near things are more
related than distant things.</p>
</blockquote>
<p>This observation motivates the examination of spatial
autocorrelation. Positive spatial autocorrelation is evident when nearby
things tend to be similar, such as that weather in Pasadena and downtown
Los Angeles (as opposed to the weather in Pasadena and San Francisco).
Negative spatial autocorrelation is evident when nearby things tend to
be more dissimilar, like squares on a chessboard. Spatial
autocorrelation can be arise from an intrinsic process such as diffusion
or communication by physical contact, or result from a covariate that
has such an intrinsic process, or in areal data, when the areal units of
observation are smaller than the scale of the spatial process.</p>
<p>The most commonly used measure of spatial autocorrelation is Moran’s
I, defined as</p>
<p><span class="math display">\[
I = \frac{n}{\sum_{i=1}^n \sum_{j=1}^n w_{ij}} \frac{\sum_{i=1}^n
\sum_{j=1}^n w_{ij} (x_i - \bar{x})(x_j - \bar{x})}{\sum_{i=1}^n (x_i -
\bar{x})^2},
\]</span></p>
<p>where <span class="math inline">\(n\)</span> is the number of spots
or locations, <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span> are different locations, or spots in
the Visium context, <span class="math inline">\(x\)</span> is a variable
with values at each location, and <span class="math inline">\(w_{ij}\)</span> is a spatial weight, which can be
inversely proportional to distance between spots or an indicator of
whether two spots are neighbors, subject to various definitions of
neighborhood and whether to normalize the number of neighbors. The <a href="https://r-spatial.github.io/spdep/index.html" class="external-link"><code>spdep</code></a>
package uses the neighborhood.</p>
<p>Moran’s I is similar to the Pearson correlation between the value at
each location and the average value at its neighbors (but not identical,
see <span class="citation">(Lee 2001)</span>). Just like Pearson
correlation, Moran’s I is generally bound between -1 and 1, where
positive value indicates positive spatial autocorrelation and negative
value indicates negative spatial autocorrelation. Spatial dependence
analysis in <code>spdep</code> requires a spatial neighborhood graph.
The graph for adjacent Visium spot can be found with</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"visium"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findVisiumGraph.html" class="external-link">findVisiumGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span></code></pre></div>
<p>We mentioned that spatial autocorrelation is apparent in total UMI
counts. Here’s what Moran’s I shows:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/calculateUnivariate.html">calculateMoransI</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>, </span>
<span>                 listw <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"visium"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 2 rows and 2 columns</span></span>
<span><span class="co">#&gt;             moran         K</span></span>
<span><span class="co">#&gt;         &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; nCounts  0.528705   3.00082</span></span>
<span><span class="co">#&gt; nGenes   0.384028   3.88036</span></span></code></pre></div>
<p>K means kurtosis. The positive values of Moran’s I indicate positive
spatial autocorrelation.</p>
<div class="section level3">
<h3 id="spatially-variable-genes">Spatially variable genes<a class="anchor" aria-label="anchor" href="#spatially-variable-genes"></a>
</h3>
<p>A spatially variable gene is a gene whose expression depends on
spatial locations, rather than being spatially random, like salt grains
spread on a soup. Spatially variable genes can be identified by spatial
autocorrelation signatures, and sometimes Moran’s I is used to compare
and assess spatially variable genes identified with different methods.
Below <code>BPPARAM</code> is used to parallelize the computation of
Moran’s I for 2000 highly variable genes, and 2 cores are used with the
SNOW backend.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">runMoransI</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, features <span class="op">=</span> <span class="va">hvgs</span>, colGraphName <span class="op">=</span> <span class="st">"visium"</span>,</span>
<span>                         BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html" class="external-link">SnowParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: &lt;anonymous&gt;: ... may be used in an incorrect context: 'fun(x[i, ], listw, ...)'</span></span></code></pre></div>
<p>The results are stored in <code>rowData</code></p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 15043 rows and 8 columns</span></span>
<span><span class="co">#&gt;                               Ensembl      symbol            type      means</span></span>
<span><span class="co">#&gt;                           &lt;character&gt; &lt;character&gt;     &lt;character&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; ENSMUSG00000025902 ENSMUSG00000025902       Sox17 Gene Expression 0.03969957</span></span>
<span><span class="co">#&gt; ENSMUSG00000096126 ENSMUSG00000096126     Gm22307 Gene Expression 0.00107296</span></span>
<span><span class="co">#&gt; ENSMUSG00000033845 ENSMUSG00000033845      Mrpl15 Gene Expression 0.38197425</span></span>
<span><span class="co">#&gt; ENSMUSG00000025903 ENSMUSG00000025903      Lypla1 Gene Expression 0.28755365</span></span>
<span><span class="co">#&gt; ENSMUSG00000033813 ENSMUSG00000033813       Tcea1 Gene Expression 0.26502146</span></span>
<span><span class="co">#&gt; ...                               ...         ...             ...        ...</span></span>
<span><span class="co">#&gt; ENSMUSG00000064360 ENSMUSG00000064360      mt-Nd3 Gene Expression  56.445279</span></span>
<span><span class="co">#&gt; ENSMUSG00000064363 ENSMUSG00000064363      mt-Nd4 Gene Expression 123.991416</span></span>
<span><span class="co">#&gt; ENSMUSG00000064367 ENSMUSG00000064367      mt-Nd5 Gene Expression  14.645923</span></span>
<span><span class="co">#&gt; ENSMUSG00000064368 ENSMUSG00000064368      mt-Nd6 Gene Expression   0.109442</span></span>
<span><span class="co">#&gt; ENSMUSG00000064370 ENSMUSG00000064370     mt-Cytb Gene Expression 121.273605</span></span>
<span><span class="co">#&gt;                           vars       cv2 moran_Vis5A   K_Vis5A</span></span>
<span><span class="co">#&gt;                      &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; ENSMUSG00000025902  0.04460915  28.30429          NA        NA</span></span>
<span><span class="co">#&gt; ENSMUSG00000096126  0.00107296 932.00000          NA        NA</span></span>
<span><span class="co">#&gt; ENSMUSG00000033845  0.47048031   3.22458          NA        NA</span></span>
<span><span class="co">#&gt; ENSMUSG00000025903  0.34686963   4.19497          NA        NA</span></span>
<span><span class="co">#&gt; ENSMUSG00000033813  0.32388797   4.61140   0.0489758   19.2181</span></span>
<span><span class="co">#&gt; ...                        ...       ...         ...       ...</span></span>
<span><span class="co">#&gt; ENSMUSG00000064360 2.47976e+03  0.778314    0.410657  11.31069</span></span>
<span><span class="co">#&gt; ENSMUSG00000064363 1.45282e+04  0.944991    0.546964  13.62886</span></span>
<span><span class="co">#&gt; ENSMUSG00000064367 2.34858e+02  1.094895    0.480634   3.75345</span></span>
<span><span class="co">#&gt; ENSMUSG00000064368 1.31941e-01 11.015664          NA        NA</span></span>
<span><span class="co">#&gt; ENSMUSG00000064370 1.48225e+04  1.007833    0.621060  10.71784</span></span></code></pre></div>
<p>The <code>NA</code>’s are for genes that are not highly variable and
Moran’s I was not computed for those genes. We rank the genes by Moran’s
I and plot them in space as follows:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">[</span><span class="va">hvgs</span>,<span class="op">]</span></span>
<span><span class="va">ord</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">moran_Vis5A</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">df</span><span class="op">[</span><span class="va">ord</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"symbol"</span>, <span class="st">"moran_Vis5A"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">#&gt; DataFrame with 2000 rows and 2 columns</span></span>
<span><span class="co">#&gt;                         symbol moran_Vis5A</span></span>
<span><span class="co">#&gt;                    &lt;character&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; ENSMUSG00000064351      mt-Co1    0.764044</span></span>
<span><span class="co">#&gt; ENSMUSG00000050335      Lgals3    0.741474</span></span>
<span><span class="co">#&gt; ENSMUSG00000029304        Spp1    0.734937</span></span>
<span><span class="co">#&gt; ENSMUSG00000021939        Ctsb    0.708362</span></span>
<span><span class="co">#&gt; ENSMUSG00000004207        Psap    0.706552</span></span>
<span><span class="co">#&gt; ...                        ...         ...</span></span>
<span><span class="co">#&gt; ENSMUSG00000039911       Spsb1  -0.0333357</span></span>
<span><span class="co">#&gt; ENSMUSG00000015711       Prune  -0.0354638</span></span>
<span><span class="co">#&gt; ENSMUSG00000042675       Ypel3  -0.0369055</span></span>
<span><span class="co">#&gt; ENSMUSG00000090262       Mpv17  -0.0412250</span></span>
<span><span class="co">#&gt; ENSMUSG00000020964       Sel1l  -0.0443975</span></span></code></pre></div>
<p>We see that some genes that have strong positive spatial
autocorrelation, but don’t observe strong negative spatial
autocorrelation. We can plot the genes with the strongest positive
spatial autocorrelation in space:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig1_visium_basic_files/figure-html/unnamed-chunk-35-1.png" width="864" style="display: block; margin: auto;"></p>
<p>These genes do indeed look spatially variable. However, such spatial
variability can simply be due to the histological regions in space, or
in other words, spatial distribution of different cell types. There are
many methods to identify spatially variable genes, often involving
Gaussian process modeling, which are far more complex than Moran’s I,
such as <a href="https://bioconductor.org/packages/release/bioc/html/spatialDE.html" class="external-link"><code>SpatialDE</code></a>
<span class="citation">(Svensson, Teichmann, and Stegle 2018)</span>.
However, such methods usually don’t account for the histological
regions, except for <code>C-SIDE</code> <span class="citation">(Cable et
al. 2022)</span>, which identifies spatially variable genes within cell
types. This leads to the question of what is really meant by “cell
type”. It remains to see how spatial methods made specifically for
identifying spatially variable genes compare with methods that don’t
explicitly use spatial information but simply perform differential
analysis between cell types which often are in spatially defined
histological regions.</p>
<p>Another consideration in using Moran’s I is the extent to which the
strength of spatial autocorrelation varies in space. What if a gene
exhibits strong spatial autocorrelation in one region, but not in
another? Should the different histological regions be analyzed
separately in some cases?</p>
<p>There are ways to see whether Moran’s I is statistically significant,
and many other methods to explore spatial autocorrelation. These are
discussed in the more advanced <a href="https://pachterlab.github.io/voyager/articles/vig2_visium.html" class="external-link">ESDA
Visium vignette</a>.</p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.2.2 (2022-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS Big Sur ... 10.16</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRblas.0.dylib</span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] sparseMatrixStats_1.10.0       stringr_1.5.0                 </span></span>
<span><span class="co">#&gt;  [3] BiocParallel_1.32.5            SFEData_1.0.2                 </span></span>
<span><span class="co">#&gt;  [5] bluster_1.8.0                  patchwork_1.1.2               </span></span>
<span><span class="co">#&gt;  [7] scran_1.26.2                   scater_1.27.3                 </span></span>
<span><span class="co">#&gt;  [9] ggplot2_3.4.0                  scuttle_1.8.4                 </span></span>
<span><span class="co">#&gt; [11] SpatialExperiment_1.8.0        SingleCellExperiment_1.20.0   </span></span>
<span><span class="co">#&gt; [13] SummarizedExperiment_1.28.0    Biobase_2.58.0                </span></span>
<span><span class="co">#&gt; [15] GenomicRanges_1.50.2           GenomeInfoDb_1.34.6           </span></span>
<span><span class="co">#&gt; [17] IRanges_2.32.0                 S4Vectors_0.36.1              </span></span>
<span><span class="co">#&gt; [19] BiocGenerics_0.44.0            MatrixGenerics_1.10.0         </span></span>
<span><span class="co">#&gt; [21] matrixStats_0.63.0             SpatialFeatureExperiment_1.0.3</span></span>
<span><span class="co">#&gt; [23] Voyager_1.0.7                 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] snow_0.4-4                    AnnotationHub_3.6.0          </span></span>
<span><span class="co">#&gt;   [3] BiocFileCache_2.6.0           systemfonts_1.0.4            </span></span>
<span><span class="co">#&gt;   [5] igraph_1.3.5                  sp_1.6-0                     </span></span>
<span><span class="co">#&gt;   [7] digest_0.6.31                 htmltools_0.5.4              </span></span>
<span><span class="co">#&gt;   [9] viridis_0.6.2                 magick_2.7.3                 </span></span>
<span><span class="co">#&gt;  [11] fansi_1.0.3                   magrittr_2.0.3               </span></span>
<span><span class="co">#&gt;  [13] memoise_2.0.1                 ScaledMatrix_1.6.0           </span></span>
<span><span class="co">#&gt;  [15] cluster_2.1.4                 limma_3.54.0                 </span></span>
<span><span class="co">#&gt;  [17] Biostrings_2.66.0             R.utils_2.12.2               </span></span>
<span><span class="co">#&gt;  [19] pkgdown_2.0.7                 colorspace_2.0-3             </span></span>
<span><span class="co">#&gt;  [21] rappdirs_0.3.3                blob_1.2.3                   </span></span>
<span><span class="co">#&gt;  [23] ggrepel_0.9.2                 textshaping_0.3.6            </span></span>
<span><span class="co">#&gt;  [25] xfun_0.36                     dplyr_1.0.10                 </span></span>
<span><span class="co">#&gt;  [27] crayon_1.5.2                  RCurl_1.98-1.9               </span></span>
<span><span class="co">#&gt;  [29] jsonlite_1.8.4                glue_1.6.2                   </span></span>
<span><span class="co">#&gt;  [31] gtable_0.3.1                  zlibbioc_1.44.0              </span></span>
<span><span class="co">#&gt;  [33] XVector_0.38.0                DelayedArray_0.24.0          </span></span>
<span><span class="co">#&gt;  [35] scico_1.3.1                   BiocSingular_1.14.0          </span></span>
<span><span class="co">#&gt;  [37] DropletUtils_1.18.1           Rhdf5lib_1.20.0              </span></span>
<span><span class="co">#&gt;  [39] HDF5Array_1.26.0              scales_1.2.1                 </span></span>
<span><span class="co">#&gt;  [41] DBI_1.1.3                     edgeR_3.40.2                 </span></span>
<span><span class="co">#&gt;  [43] Rcpp_1.0.9                    xtable_1.8-4                 </span></span>
<span><span class="co">#&gt;  [45] viridisLite_0.4.1             spData_2.2.1                 </span></span>
<span><span class="co">#&gt;  [47] units_0.8-1                   dqrng_0.3.0                  </span></span>
<span><span class="co">#&gt;  [49] bit_4.0.5                     spdep_1.2-7                  </span></span>
<span><span class="co">#&gt;  [51] rsvd_1.0.5                    proxy_0.4-27                 </span></span>
<span><span class="co">#&gt;  [53] httr_1.4.4                    metapod_1.6.0                </span></span>
<span><span class="co">#&gt;  [55] FNN_1.1.3.1                   RColorBrewer_1.1-3           </span></span>
<span><span class="co">#&gt;  [57] ellipsis_0.3.2                wk_0.7.1                     </span></span>
<span><span class="co">#&gt;  [59] farver_2.1.1                  pkgconfig_2.0.3              </span></span>
<span><span class="co">#&gt;  [61] R.methodsS3_1.8.2             uwot_0.1.14                  </span></span>
<span><span class="co">#&gt;  [63] dbplyr_2.3.0                  sass_0.4.4                   </span></span>
<span><span class="co">#&gt;  [65] deldir_1.0-6                  locfit_1.5-9.7               </span></span>
<span><span class="co">#&gt;  [67] utf8_1.2.2                    labeling_0.4.2               </span></span>
<span><span class="co">#&gt;  [69] AnnotationDbi_1.60.0          later_1.3.0                  </span></span>
<span><span class="co">#&gt;  [71] tidyselect_1.2.0              rlang_1.0.6                  </span></span>
<span><span class="co">#&gt;  [73] munsell_0.5.0                 BiocVersion_3.16.0           </span></span>
<span><span class="co">#&gt;  [75] tools_4.2.2                   cachem_1.0.6                 </span></span>
<span><span class="co">#&gt;  [77] cli_3.6.0                     dbscan_1.1-11                </span></span>
<span><span class="co">#&gt;  [79] ExperimentHub_2.6.0           generics_0.1.3               </span></span>
<span><span class="co">#&gt;  [81] RSQLite_2.2.20                evaluate_0.20                </span></span>
<span><span class="co">#&gt;  [83] fastmap_1.1.0                 yaml_2.3.6                   </span></span>
<span><span class="co">#&gt;  [85] ragg_1.2.5                    knitr_1.41                   </span></span>
<span><span class="co">#&gt;  [87] bit64_4.0.5                   fs_1.5.2                     </span></span>
<span><span class="co">#&gt;  [89] purrr_1.0.1                   s2_1.1.2                     </span></span>
<span><span class="co">#&gt;  [91] KEGGREST_1.38.0               mime_0.12                    </span></span>
<span><span class="co">#&gt;  [93] R.oo_1.25.0                   compiler_4.2.2               </span></span>
<span><span class="co">#&gt;  [95] png_0.1-8                     interactiveDisplayBase_1.36.0</span></span>
<span><span class="co">#&gt;  [97] filelock_1.0.2                curl_5.0.0                   </span></span>
<span><span class="co">#&gt;  [99] beeswarm_0.4.0                e1071_1.7-12                 </span></span>
<span><span class="co">#&gt; [101] tibble_3.1.8                  statmod_1.5.0                </span></span>
<span><span class="co">#&gt; [103] bslib_0.4.2                   stringi_1.7.12               </span></span>
<span><span class="co">#&gt; [105] highr_0.10                    desc_1.4.2                   </span></span>
<span><span class="co">#&gt; [107] lattice_0.20-45               Matrix_1.5-3                 </span></span>
<span><span class="co">#&gt; [109] classInt_0.4-8                vctrs_0.5.1                  </span></span>
<span><span class="co">#&gt; [111] pillar_1.8.1                  lifecycle_1.0.3              </span></span>
<span><span class="co">#&gt; [113] rhdf5filters_1.10.0           BiocManager_1.30.19          </span></span>
<span><span class="co">#&gt; [115] jquerylib_0.1.4               BiocNeighbors_1.16.0         </span></span>
<span><span class="co">#&gt; [117] cowplot_1.1.1                 bitops_1.0-7                 </span></span>
<span><span class="co">#&gt; [119] irlba_2.3.5.1                 httpuv_1.6.8                 </span></span>
<span><span class="co">#&gt; [121] R6_2.5.1                      promises_1.2.0.1             </span></span>
<span><span class="co">#&gt; [123] KernSmooth_2.23-20            gridExtra_2.3                </span></span>
<span><span class="co">#&gt; [125] vipor_0.4.5                   codetools_0.2-18             </span></span>
<span><span class="co">#&gt; [127] boot_1.3-28.1                 assertthat_0.2.1             </span></span>
<span><span class="co">#&gt; [129] rhdf5_2.42.0                  rprojroot_2.0.3              </span></span>
<span><span class="co">#&gt; [131] rjson_0.2.21                  withr_2.5.0                  </span></span>
<span><span class="co">#&gt; [133] GenomeInfoDbData_1.2.9        parallel_4.2.2               </span></span>
<span><span class="co">#&gt; [135] grid_4.2.2                    beachmat_2.14.0              </span></span>
<span><span class="co">#&gt; [137] class_7.3-20                  rmarkdown_2.20               </span></span>
<span><span class="co">#&gt; [139] DelayedMatrixStats_1.20.0     ggnewscale_0.4.8             </span></span>
<span><span class="co">#&gt; [141] sf_1.0-9                      shiny_1.7.4                  </span></span>
<span><span class="co">#&gt; [143] ggbeeswarm_0.7.1</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Cable2022-ma" class="csl-entry">
Cable, Dylan M, Evan Murray, Vignesh Shanmugam, Simon Zhang, Luli S Zou,
Michael Diao, Haiqi Chen, Evan Z Macosko, Rafael A Irizarry, and Fei
Chen. 2022. <span>“Cell Type-Specific Inference of Differential
Expression in Spatial Transcriptomics.”</span> <em>Nat. Methods</em> 19
(9): 1076–87.
</div>
<div id="ref-Chari2021-hb" class="csl-entry">
Chari, Tara, Joeyta Banerjee, and Lior Pachter. 2021. <span>“The
Specious Art of <span>Single-Cell</span> Genomics.”</span>
<em>bioRxiv</em>.
</div>
<div id="ref-Lee2001-tm" class="csl-entry">
Lee, Sang-Il. 2001. <span>“Developing a Bivariate Spatial Association
Measure: An Integration of Pearson’s r and Moran’s
<span>I</span>.”</span> <em>J. Geogr. Syst.</em> 3 (4): 369–85.
</div>
<div id="ref-Lun2016-yq" class="csl-entry">
Lun, Aaron T L, Karsten Bach, and John C Marioni. 2016. <span>“Pooling
Across Cells to Normalize Single-Cell <span>RNA</span> Sequencing Data
with Many Zero Counts.”</span> <em>Genome Biol.</em> 17 (April): 75.
</div>
<div id="ref-Ma2022-dc" class="csl-entry">
Ma, Cong, Uthsav Chitra, Shirley Zhang, and Benjamin J Raphael. 2022.
<span>“Belayer: Modeling Discrete and Continuous Spatial Variation in
Gene Expression from Spatially Resolved Transcriptomics.”</span>
<em>Cell Syst</em> 13 (10): 786–797.e13.
</div>
<div id="ref-McKellar2021-ek" class="csl-entry">
McKellar, David W, Lauren D Walter, Leo T Song, Madhav Mantri, Michael F
Z Wang, Iwijn De Vlaminck, and Benjamin D Cosgrove. 2021.
<span>“Large-Scale Integration of Single-Cell Transcriptomic Data
Captures Transitional Progenitor States in Mouse Skeletal Muscle
Regeneration.”</span> <em>Commun Biol</em> 4 (1): 1280.
</div>
<div id="ref-Moses2022-xz" class="csl-entry">
Moses, Lambda, and Lior Pachter. 2022. <span>“Publisher Correction:
Museum of Spatial Transcriptomics.”</span> <em>Nat. Methods</em> 19 (5):
628.
</div>
<div id="ref-Svensson2018-sx" class="csl-entry">
Svensson, Valentine, Sarah A Teichmann, and Oliver Stegle. 2018.
<span>“<span>SpatialDE</span>: Identification of Spatially Variable
Genes.”</span> <em>Nat. Methods</em> 15 (5): 343–46.
</div>
<div id="ref-Zhao2021-vb" class="csl-entry">
Zhao, Edward, Matthew R Stone, Xing Ren, Jamie Guenthoer, Kimberly S
Smythe, Thomas Pulliam, Stephen R Williams, et al. 2021. <span>“Spatial
Transcriptomics at Subspot Resolution with
<span>BayesSpace</span>.”</span> <em>Nat. Biotechnol.</em> 39 (11):
1375–84.
</div>
<div id="ref-Zhao2013-tw" class="csl-entry">
Zhao, Feng, Licheng Jiao, and Hanqiang Liu. 2013. <span>“Kernel
Generalized Fuzzy c-Means Clustering with Spatial Information for Image
Segmentation.”</span> <em>Digit. Signal Process.</em> 23 (1): 184–99.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Lambda Moses, Kayla Jackson, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
