<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Voyager">
<title>Xenium breast cancer dataset • Voyager</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Xenium breast cancer dataset">
<meta property="og:description" content="Voyager">
<meta name="robots" content="noindex">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">Voyager</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">1.2.4</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/install.html">Install</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-technologies">Technologies</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-technologies">
    <a class="dropdown-item" href="../articles/pros_cons_landing.html">Pros and cons</a>
    <a class="dropdown-item" href="../articles/visium_landing.html">Visium</a>
    <a class="dropdown-item" href="../articles/slideseqV2_landing.html">Slide-seq v2</a>
    <a class="dropdown-item" href="../articles/cosmx_landing.html">CosMX</a>
    <a class="dropdown-item" href="../articles/xenium_landing.html">Xenium</a>
    <a class="dropdown-item" href="../articles/merfish_landing.html">MERFISH</a>
    <a class="dropdown-item" href="../articles/seqfish_landing.html">seqFISH</a>
    <a class="dropdown-item" href="../articles/codex_landing.html">CODEX/PhenoCycler</a>
    <a class="dropdown-item" href="../articles/splitseq_landing.html">SPLiT-seq</a>
    <a class="dropdown-item" href="../articles/chromium_landing.html">10X Chromium V3</a>
    <a class="dropdown-item" href="../articles/10xnuclei_landing.html">10X Chromium Nuclei Isolation</a>
    <a class="dropdown-item" href="../articles/10xcrispr_landing.html">10X Chromium Single Cell CRISPR</a>
    <a class="dropdown-item" href="../articles/10xatac_landing.html">10X Single Cell ATAC</a>
    <a class="dropdown-item" href="../articles/10xmultiome_landing.html">10X Single Cell Multiome</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-methods">Methods</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-methods">
    <h6 class="dropdown-header" data-toc-skip>Univariate global</h6>
    <a class="dropdown-item" href="../articles/moran_landing.html">Moran's I</a>
    <a class="dropdown-item" href="../articles/geary_landing.html">Geary's C</a>
    <a class="dropdown-item" href="../articles/correlogram_landing.html">Correlogram</a>
    <a class="dropdown-item" href="../articles/variogram_landing.html">Variogram</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Univariate local</h6>
    <a class="dropdown-item" href="../articles/moranplot_landing.html">Moran scatter plot</a>
    <a class="dropdown-item" href="../articles/localmoran_landing.html">Local Moran's I</a>
    <a class="dropdown-item" href="../articles/losh_landing.html">Local spatial heteroscedasticity</a>
    <a class="dropdown-item" href="../articles/getisord_landing.html">Getis-Ord Gi*</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Bivariate</h6>
    <a class="dropdown-item" href="../articles/lee_landing.html">Lee's L</a>
    <a class="dropdown-item" href="../articles/crossvariogram_landing.html">Cross variogram</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Multivariate</h6>
    <a class="dropdown-item" href="../articles/multispati_landing.html">MULTISPATI PCA</a>
    <a class="dropdown-item" href="../articles/localc_landing.html">Multivariate local Geary's C</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../articles/sfemethod">Custom methods</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/pachterlab/voyager/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Xenium breast cancer dataset</h1>
                        <h4 data-toc-skip class="author">Lambda
Moses</h4>
            
            <h4 data-toc-skip class="date">2023-07-31</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pachterlab/voyager/blob/HEAD/vignettes/vig5_xenium.Rmd" class="external-link"><code>vignettes/vig5_xenium.Rmd</code></a></small>
      <div class="d-none name"><code>vig5_xenium.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Xenium is a new technology from 10X genomics for single cell
resolution smFISH based spatial transcriptomics. The first Xenium
dataset is for formalin fixed paraffin embedded (FFPE) human breast
tumor, reported in <span class="citation">(Janesick et al. 2022)</span>
and downloaded from the <a href="https://www.10xgenomics.com/products/xenium-in-situ/preview-dataset-human-breast" class="external-link">10X
website</a>.</p>
<p>The gene count matrix was downloaded as an HDF5 file and read into R
as a <code>SingleCellExperiment</code> (SCE) object with
<code><a href="https://rdrr.io/pkg/DropletUtils/man/read10xCounts.html" class="external-link">DropletUtils::read10xCounts()</a></code>. The gene count matrix is
originally a <code>DelayedArray</code>, so the data is not all loaded
into memory. For now, the matrix has been converted into an in memory
<code>dgCMatrix</code>. However, for the next release, we would like to
write another vignette on on disk analyses. The challenge is
representing <code>sf</code> data frames on disk, perhaps with <a href="https://github.com/apache/incubator-sedona" class="external-link"><code>sedona</code></a>
and <a href="https://bioconductor.org/packages/release/bioc/html/SQLDataFrame.html" class="external-link"><code>SQLDataFrame</code></a>.</p>
<p>The cell metadata (including centroid coordinates) and cell
segmentation polygons were downloaded as <code>parquet</code> files, a
more compact way to store columnar data than CSV, and read into R as
data frames with <code><a href="https://arrow.apache.org/docs/r/reference/read_parquet.html" class="external-link">arrow::read_parquet()</a></code>. The cell polygons
were converted into <code>sf</code> data frame with
<code><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/df2sf.html" class="external-link">SpatialFeatureExperiment::df2sf()</a></code>. Then the SCE object was
converted into <code>SpatialFeatureExperiment</code> (SFE) and the
polygon geometry was added to the SFE object, which is in the
<code>SFEData</code> package.</p>
<p>Here we load the packages used in this vignette.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/voyager" class="external-link">Voyager</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; The legacy packages maptools, rgdal, and rgeos, underpinning the sp package,</span></span>
<span><span class="co">#&gt; which was just loaded, will retire in October 2023.</span></span>
<span><span class="co">#&gt; Please refer to R-spatial evolution reports for details, especially</span></span>
<span><span class="co">#&gt; https://r-spatial.org/r/2023/05/15/evolution4.html.</span></span>
<span><span class="co">#&gt; It may be desirable to make the sf package available;</span></span>
<span><span class="co">#&gt; package maintainers should consider adding sf to Suggests:.</span></span>
<span><span class="co">#&gt; The sp package is now running under evolution status 2</span></span>
<span><span class="co">#&gt;      (status 2 uses the sf package in place of rgdal)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SFEData" class="external-link">SFEData</a></span><span class="op">)</span> </span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: SummarizedExperiment</span></span>
<span><span class="co">#&gt; Loading required package: MatrixGenerics</span></span>
<span><span class="co">#&gt; Loading required package: matrixStats</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'MatrixGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span></span>
<span><span class="co">#&gt;     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span></span>
<span><span class="co">#&gt;     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span></span>
<span><span class="co">#&gt;     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span></span>
<span><span class="co">#&gt;     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span></span>
<span><span class="co">#&gt;     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span></span>
<span><span class="co">#&gt;     colWeightedMeans, colWeightedMedians, colWeightedSds,</span></span>
<span><span class="co">#&gt;     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span></span>
<span><span class="co">#&gt;     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span></span>
<span><span class="co">#&gt;     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span></span>
<span><span class="co">#&gt;     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span></span>
<span><span class="co">#&gt;     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span></span>
<span><span class="co">#&gt;     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span></span>
<span><span class="co">#&gt;     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span></span>
<span><span class="co">#&gt;     rowWeightedSds, rowWeightedVars</span></span>
<span><span class="co">#&gt; Loading required package: GenomicRanges</span></span>
<span><span class="co">#&gt; Loading required package: stats4</span></span>
<span><span class="co">#&gt; Loading required package: BiocGenerics</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'BiocGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">#&gt;     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">#&gt;     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,</span></span>
<span><span class="co">#&gt;     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span></span>
<span><span class="co">#&gt;     Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,</span></span>
<span><span class="co">#&gt;     table, tapply, union, unique, unsplit, which.max, which.min</span></span>
<span><span class="co">#&gt; Loading required package: S4Vectors</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'S4Vectors'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:utils':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     findMatches</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     expand.grid, I, unname</span></span>
<span><span class="co">#&gt; Loading required package: IRanges</span></span>
<span><span class="co">#&gt; Loading required package: GenomeInfoDb</span></span>
<span><span class="co">#&gt; Loading required package: Biobase</span></span>
<span><span class="co">#&gt; Welcome to Bioconductor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     Vignettes contain introductory material; view with</span></span>
<span><span class="co">#&gt;     'browseVignettes()'. To cite Bioconductor, see</span></span>
<span><span class="co">#&gt;     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'Biobase'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:MatrixGenerics':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     rowMedians</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyMissing, rowMedians</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drighelli/SpatialExperiment" class="external-link">SpatialExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SpatialFeatureExperiment" class="external-link">SpatialFeatureExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'ggplot2'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:SpatialFeatureExperiment':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     unit</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org" class="external-link">stringr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span> </span>
<span><span class="co">#&gt; Loading required package: scuttle</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scuttle</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/LTLA/BiocSingular" class="external-link">BiocSingular</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/JanesickBreastData.html" class="external-link">JanesickBreastData</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"rep2"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; see ?SFEData and browseVignettes('SFEData') for documentation</span></span>
<span><span class="co">#&gt; downloading 1 resources</span></span>
<span><span class="co">#&gt; retrieving 1 resource</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 541 118708 </span></span>
<span><span class="co">#&gt; metadata(1): Samples</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(541): ABCC11 ACTA2 ... BLANK_0497 BLANK_0499</span></span>
<span><span class="co">#&gt; rowData names(6): ID Symbol ... vars cv2</span></span>
<span><span class="co">#&gt; colnames: NULL</span></span>
<span><span class="co">#&gt; colData names(10): Sample Barcode ... nCounts nGenes</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : x_centroid y_centroid</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit:</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: centroids (POINT), cellSeg (POLYGON), nucSeg (GEOMETRY) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01:</span></span></code></pre></div>
<p>There are 118708 cells in this dataset, a little more than in the
CosMX dataset.</p>
<p>The SFE object doesn’t have column names (i.e. cell IDs). Here we
assign cell IDs.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>This is what the tissue, with the cell outlines, looks like</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotGeometry.html">plotGeometry</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"cellSeg"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-5-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Plot cell density in space</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotCellBin2D.html">plotCellBin2D</a></span><span class="op">(</span><span class="va">sfe</span>, hex <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-6-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="quality-control">Quality control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h2>
<div class="section level3">
<h3 id="cells">Cells<a class="anchor" aria-label="anchor" href="#cells"></a>
</h3>
<p>Some QC metrics are precomputed and are stored in
<code>colData</code></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "Sample"                  "Barcode"                </span></span>
<span><span class="co">#&gt;  [3] "transcript_counts"       "control_probe_counts"   </span></span>
<span><span class="co">#&gt;  [5] "control_codeword_counts" "cell_area"              </span></span>
<span><span class="co">#&gt;  [7] "nucleus_area"            "sample_id"              </span></span>
<span><span class="co">#&gt;  [9] "nCounts"                 "nGenes"</span></span></code></pre></div>
<p>Since there’re more cells, it would be better to plot the tissue
larger, so we’ll plot the histogram of QC metrics and the spatial plots
separately, unlike in the CosMx vignette.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n_panel</span> <span class="op">&lt;-</span> <span class="fl">313</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">nCounts_normed</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">$</span><span class="va">nCounts</span><span class="op">/</span><span class="va">n_panel</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">nGenes_normed</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">$</span><span class="va">nGenes</span><span class="op">/</span><span class="va">n_panel</span></span></code></pre></div>
<p>Here we divided nCounts by the total number of genes probed, so this
histogram is comparable to those from other smFISH-based datasets.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotColDataHistogram.html">plotColDataHistogram</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts_normed"</span>, <span class="st">"nGenes_normed"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-9-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Compared to the <a href="https://pachterlab.github.io/voyager/articles/vig4_cosmx.html#cells">FFPE
CosMX non-small cell lung cancer dataset</a>, more transcripts per gene
on average and a larger proportion of all genes are detected in this
dataset, which is also FFPE. However, this should be interpreted with
care, since these two datasets are from different tissues and have
different gene panels, so this may or may not indicate that Xenium has
better detection efficiency than CosMX.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nCounts"</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-10-1.png" width="700" style="display: block; margin: auto;"></p>
<p>There seem to be FOV artifacts. However, the cell ID and FOV
information were unavailable so we cannot examine them.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nGenes"</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-11-1.png" width="700" style="display: block; margin: auto;"></p>
<p>A standard examination is to look at the relationship between nCounts
and nGenes:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotColDataBin2D.html">plotColDataBin2D</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nCounts"</span>, <span class="st">"nGenes"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: plotCol/RowDataBin2D() was deprecated in Voyager 1.2.0.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Please use scater::plotCol/RowData() instead.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Set `bins` argument to enable binning.</span></span>
<span><span class="co">#&gt; <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">generated.</span></span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-12-1.png" width="700" style="display: block; margin: auto;"></p>
<p>There appear to be two branches.</p>
<p>Here we plot the distribution of cell area</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotColDataHistogram.html">plotColDataHistogram</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cell_area"</span>, <span class="st">"nucleus_area"</span><span class="op">)</span>, scales <span class="op">=</span> <span class="st">"free_y"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-13-1.png" width="700" style="display: block; margin: auto;"></p>
<p>That should be in pixels. There’s a very long tail. The nuclei are
much smaller than the cells.</p>
<p>How is cell area distributed in space?</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"cell_area"</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-14-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Cells in the sparse region tend to be larger than those in the dense
region. This may be biological or an artifact of the cell segmentation
algorithm or both.</p>
<p>Here the nuclei segmentations are plotted instead of cell
segmentation. The nuclei are much smaller to the extent that they are
difficult to see.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nucleus_area"</span>, colGeometryName <span class="op">=</span> <span class="st">"nucSeg"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-15-1.png" width="700" style="display: block; margin: auto;"></p>
<p>There’s an outlier near the right edge of the section, throwing off
the dynamic range of the plot. Upon inspection of the H&amp;E image, the
outlier is a bit of tissue debris that doesn’t look like a cell. But we
can still that cells in the dense, gland like regions tend to have
larger nuclei. This may be biological, or that nuclei are so densely
packed in those regions that they are more likely to be undersegmented,
i.e. when multiple nuclei are counted as one by the nuclei segmentation
program, or both.</p>
<p>These observations motivate an examination of the relationship
between cell area and nuclei area:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotColDataBin2D.html">plotColDataBin2D</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"cell_area"</span>, <span class="st">"nucleus_area"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_fill_viridis_c</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">fill</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">fill</span>, which will replace the existing scale.</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-16-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Again, there are two branches, probably related to cell density and
cell type. The nucleus outlier also has large cell area, though it is
not as much an outlier in cell area. However, it is a spatial outlier as
it’s unusually large compared to its neighbors (scroll up two plots
back).</p>
<p>Next we calculate the proportion of cell in this z-plane taken up by
the nucleus, and examine the distribution:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">prop_nuc</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">$</span><span class="va">nucleus_area</span> <span class="op">/</span> <span class="va">sfe</span><span class="op">$</span><span class="va">cell_area</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotColDataHistogram.html">plotColDataHistogram</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"prop_nuc"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-18-1.png" width="700" style="display: block; margin: auto;"></p>
<p>This distribution could have been generated from two peaks that were
combined. From the histogram, there do not seem to be cells without
nuclei or segmentation artifacts where the nucleus is larger than the
cell. However, there are so many cells in this dataset and it is
possible that just a few cells would not be visible on this histogram.
We double check:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># No nucleus</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">$</span><span class="va">nucleus_area</span> <span class="op">&lt;</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0</span></span>
<span><span class="co"># Nucleus larger than cell</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">$</span><span class="va">nucleus_area</span> <span class="op">&gt;</span> <span class="va">sfe</span><span class="op">$</span><span class="va">cell_area</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0</span></span></code></pre></div>
<p>So there are no cells without nuclei or nuclei larger than their
cells. Here we plot the nuclei proportion in space:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"prop_nuc"</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-20-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Cells in some histological regions have larger proportions occupied
by the nuclei. It is interesting to check, controlling for cell type,
how cell area, nucleus area, and the proportion of cell occupied by
nucleus relate to gene expression. However, a problem in performing such
an analysis is that cell segmentation is only available for one z-plane
here and these areas also relate to where this z-plane intersects each
cell.</p>
<p>Below we plot a 2D histogram to better show the density of points on
this plot:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotColDataBin2D.html">plotColDataBin2D</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"cell_area"</span>, <span class="st">"prop_nuc"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_fill_viridis_c</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">fill</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">fill</span>, which will replace the existing scale.</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-21-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Smaller cells tend to have higher proportion occupied by the nucleus.
This can be related to cell type, or it could be a limitation in how
small the nuclei can be in this tissue.</p>
<p>We also examine the relationship between nucleus area and the
proportion of cell occupied by the nucleus:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotColDataBin2D.html">plotColDataBin2D</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nucleus_area"</span>, <span class="st">"prop_nuc"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_fill_viridis_c</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">fill</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">fill</span>, which will replace the existing scale.</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-22-1.png" width="700" style="display: block; margin: auto;"></p>
<p>The outlier is obvious. There are more cells with both small nuclei
and low proportion of area occupied by the nucleus.</p>
</div>
<div class="section level3">
<h3 id="negative-controls">Negative controls<a class="anchor" aria-label="anchor" href="#negative-controls"></a>
</h3>
<p>Since there are only a few hundred genes plus negative control
probes, all row names of the SFE object can be printed out to find what
the negative control probes are called.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span>
<span><span class="co">#&gt;   [1] "ABCC11"                  "ACTA2"                  </span></span>
<span><span class="co">#&gt;   [3] "ACTG2"                   "ADAM9"                  </span></span>
<span><span class="co">#&gt;   [5] "ADGRE5"                  "ADH1B"                  </span></span>
<span><span class="co">#&gt;   [7] "ADIPOQ"                  "AGR3"                   </span></span>
<span><span class="co">#&gt;   [9] "AHSP"                    "AIF1"                   </span></span>
<span><span class="co">#&gt;  [11] "AKR1C1"                  "AKR1C3"                 </span></span>
<span><span class="co">#&gt;  [13] "ALDH1A3"                 "ANGPT2"                 </span></span>
<span><span class="co">#&gt;  [15] "ANKRD28"                 "ANKRD29"                </span></span>
<span><span class="co">#&gt;  [17] "ANKRD30A"                "APOBEC3A"               </span></span>
<span><span class="co">#&gt;  [19] "APOBEC3B"                "APOC1"                  </span></span>
<span><span class="co">#&gt;  [21] "AQP1"                    "AQP3"                   </span></span>
<span><span class="co">#&gt;  [23] "AR"                      "AVPR1A"                 </span></span>
<span><span class="co">#&gt;  [25] "BACE2"                   "BANK1"                  </span></span>
<span><span class="co">#&gt;  [27] "BASP1"                   "BTNL9"                  </span></span>
<span><span class="co">#&gt;  [29] "C15orf48"                "C1QA"                   </span></span>
<span><span class="co">#&gt;  [31] "C1QC"                    "C2orf42"                </span></span>
<span><span class="co">#&gt;  [33] "C5orf46"                 "C6orf132"               </span></span>
<span><span class="co">#&gt;  [35] "CAV1"                    "CAVIN2"                 </span></span>
<span><span class="co">#&gt;  [37] "CCDC6"                   "CCDC80"                 </span></span>
<span><span class="co">#&gt;  [39] "CCL20"                   "CCL5"                   </span></span>
<span><span class="co">#&gt;  [41] "CCL8"                    "CCND1"                  </span></span>
<span><span class="co">#&gt;  [43] "CCPG1"                   "CCR7"                   </span></span>
<span><span class="co">#&gt;  [45] "CD14"                    "CD163"                  </span></span>
<span><span class="co">#&gt;  [47] "CD19"                    "CD1C"                   </span></span>
<span><span class="co">#&gt;  [49] "CD247"                   "CD27"                   </span></span>
<span><span class="co">#&gt;  [51] "CD274"                   "CD3D"                   </span></span>
<span><span class="co">#&gt;  [53] "CD3E"                    "CD3G"                   </span></span>
<span><span class="co">#&gt;  [55] "CD4"                     "CD68"                   </span></span>
<span><span class="co">#&gt;  [57] "CD69"                    "CD79A"                  </span></span>
<span><span class="co">#&gt;  [59] "CD79B"                   "CD80"                   </span></span>
<span><span class="co">#&gt;  [61] "CD83"                    "CD86"                   </span></span>
<span><span class="co">#&gt;  [63] "CD8A"                    "CD8B"                   </span></span>
<span><span class="co">#&gt;  [65] "CD9"                     "CD93"                   </span></span>
<span><span class="co">#&gt;  [67] "CDC42EP1"                "CDH1"                   </span></span>
<span><span class="co">#&gt;  [69] "CEACAM6"                 "CEACAM8"                </span></span>
<span><span class="co">#&gt;  [71] "CENPF"                   "CLCA2"                  </span></span>
<span><span class="co">#&gt;  [73] "CLDN4"                   "CLDN5"                  </span></span>
<span><span class="co">#&gt;  [75] "CLEC14A"                 "CLEC9A"                 </span></span>
<span><span class="co">#&gt;  [77] "CLECL1"                  "CLIC6"                  </span></span>
<span><span class="co">#&gt;  [79] "CPA3"                    "CRHBP"                  </span></span>
<span><span class="co">#&gt;  [81] "CRISPLD2"                "CSF3"                   </span></span>
<span><span class="co">#&gt;  [83] "CTH"                     "CTLA4"                  </span></span>
<span><span class="co">#&gt;  [85] "CTSG"                    "CTTN"                   </span></span>
<span><span class="co">#&gt;  [87] "CX3CR1"                  "CXCL12"                 </span></span>
<span><span class="co">#&gt;  [89] "CXCL16"                  "CXCL5"                  </span></span>
<span><span class="co">#&gt;  [91] "CXCR4"                   "CYP1A1"                 </span></span>
<span><span class="co">#&gt;  [93] "CYTIP"                   "DAPK3"                  </span></span>
<span><span class="co">#&gt;  [95] "DERL3"                   "DMKN"                   </span></span>
<span><span class="co">#&gt;  [97] "DNAAF1"                  "DNTTIP1"                </span></span>
<span><span class="co">#&gt;  [99] "DPT"                     "DSC2"                   </span></span>
<span><span class="co">#&gt; [101] "DSP"                     "DST"                    </span></span>
<span><span class="co">#&gt; [103] "DUSP2"                   "DUSP5"                  </span></span>
<span><span class="co">#&gt; [105] "EDN1"                    "EDNRB"                  </span></span>
<span><span class="co">#&gt; [107] "EGFL7"                   "EGFR"                   </span></span>
<span><span class="co">#&gt; [109] "EIF4EBP1"                "ELF3"                   </span></span>
<span><span class="co">#&gt; [111] "ELF5"                    "ENAH"                   </span></span>
<span><span class="co">#&gt; [113] "EPCAM"                   "ERBB2"                  </span></span>
<span><span class="co">#&gt; [115] "ERN1"                    "ESM1"                   </span></span>
<span><span class="co">#&gt; [117] "ESR1"                    "FAM107B"                </span></span>
<span><span class="co">#&gt; [119] "FAM49A"                  "FASN"                   </span></span>
<span><span class="co">#&gt; [121] "FBLIM1"                  "FBLN1"                  </span></span>
<span><span class="co">#&gt; [123] "FCER1A"                  "FCER1G"                 </span></span>
<span><span class="co">#&gt; [125] "FCGR3A"                  "FGL2"                   </span></span>
<span><span class="co">#&gt; [127] "FLNB"                    "FOXA1"                  </span></span>
<span><span class="co">#&gt; [129] "FOXC2"                   "FOXP3"                  </span></span>
<span><span class="co">#&gt; [131] "FSTL3"                   "GATA3"                  </span></span>
<span><span class="co">#&gt; [133] "GJB2"                    "GLIPR1"                 </span></span>
<span><span class="co">#&gt; [135] "GNLY"                    "GPR183"                 </span></span>
<span><span class="co">#&gt; [137] "GZMA"                    "GZMB"                   </span></span>
<span><span class="co">#&gt; [139] "GZMK"                    "HAVCR2"                 </span></span>
<span><span class="co">#&gt; [141] "HDC"                     "HMGA1"                  </span></span>
<span><span class="co">#&gt; [143] "HOOK2"                   "HOXD8"                  </span></span>
<span><span class="co">#&gt; [145] "HOXD9"                   "HPX"                    </span></span>
<span><span class="co">#&gt; [147] "IGF1"                    "IGSF6"                  </span></span>
<span><span class="co">#&gt; [149] "IL2RA"                   "IL2RG"                  </span></span>
<span><span class="co">#&gt; [151] "IL3RA"                   "IL7R"                   </span></span>
<span><span class="co">#&gt; [153] "ITGAM"                   "ITGAX"                  </span></span>
<span><span class="co">#&gt; [155] "ITM2C"                   "JUP"                    </span></span>
<span><span class="co">#&gt; [157] "KARS"                    "KDR"                    </span></span>
<span><span class="co">#&gt; [159] "KIT"                     "KLF5"                   </span></span>
<span><span class="co">#&gt; [161] "KLRB1"                   "KLRC1"                  </span></span>
<span><span class="co">#&gt; [163] "KLRD1"                   "KLRF1"                  </span></span>
<span><span class="co">#&gt; [165] "KRT14"                   "KRT15"                  </span></span>
<span><span class="co">#&gt; [167] "KRT16"                   "KRT23"                  </span></span>
<span><span class="co">#&gt; [169] "KRT5"                    "KRT6B"                  </span></span>
<span><span class="co">#&gt; [171] "KRT7"                    "KRT8"                   </span></span>
<span><span class="co">#&gt; [173] "LAG3"                    "LARS"                   </span></span>
<span><span class="co">#&gt; [175] "LDHB"                    "LEP"                    </span></span>
<span><span class="co">#&gt; [177] "LGALSL"                  "LIF"                    </span></span>
<span><span class="co">#&gt; [179] "LILRA4"                  "LPL"                    </span></span>
<span><span class="co">#&gt; [181] "LPXN"                    "LRRC15"                 </span></span>
<span><span class="co">#&gt; [183] "LTB"                     "LUM"                    </span></span>
<span><span class="co">#&gt; [185] "LY86"                    "LYPD3"                  </span></span>
<span><span class="co">#&gt; [187] "LYZ"                     "MAP3K8"                 </span></span>
<span><span class="co">#&gt; [189] "MDM2"                    "MEDAG"                  </span></span>
<span><span class="co">#&gt; [191] "MKI67"                   "MLPH"                   </span></span>
<span><span class="co">#&gt; [193] "MMP1"                    "MMP12"                  </span></span>
<span><span class="co">#&gt; [195] "MMP2"                    "MMRN2"                  </span></span>
<span><span class="co">#&gt; [197] "MNDA"                    "MPO"                    </span></span>
<span><span class="co">#&gt; [199] "MRC1"                    "MS4A1"                  </span></span>
<span><span class="co">#&gt; [201] "MUC6"                    "MYBPC1"                 </span></span>
<span><span class="co">#&gt; [203] "MYH11"                   "MYLK"                   </span></span>
<span><span class="co">#&gt; [205] "MYO5B"                   "MZB1"                   </span></span>
<span><span class="co">#&gt; [207] "NARS"                    "NCAM1"                  </span></span>
<span><span class="co">#&gt; [209] "NDUFA4L2"                "NKG7"                   </span></span>
<span><span class="co">#&gt; [211] "NOSTRIN"                 "NPM3"                   </span></span>
<span><span class="co">#&gt; [213] "OCIAD2"                  "OPRPN"                  </span></span>
<span><span class="co">#&gt; [215] "OXTR"                    "PCLAF"                  </span></span>
<span><span class="co">#&gt; [217] "PCOLCE"                  "PDCD1"                  </span></span>
<span><span class="co">#&gt; [219] "PDCD1LG2"                "PDE4A"                  </span></span>
<span><span class="co">#&gt; [221] "PDGFRA"                  "PDGFRB"                 </span></span>
<span><span class="co">#&gt; [223] "PDK4"                    "PECAM1"                 </span></span>
<span><span class="co">#&gt; [225] "PELI1"                   "PGR"                    </span></span>
<span><span class="co">#&gt; [227] "PIGR"                    "PIM1"                   </span></span>
<span><span class="co">#&gt; [229] "PLD4"                    "POLR2J3"                </span></span>
<span><span class="co">#&gt; [231] "POSTN"                   "PPARG"                  </span></span>
<span><span class="co">#&gt; [233] "PRDM1"                   "PRF1"                   </span></span>
<span><span class="co">#&gt; [235] "PTGDS"                   "PTN"                    </span></span>
<span><span class="co">#&gt; [237] "PTPRC"                   "PTRHD1"                 </span></span>
<span><span class="co">#&gt; [239] "QARS"                    "RAB30"                  </span></span>
<span><span class="co">#&gt; [241] "RAMP2"                   "RAPGEF3"                </span></span>
<span><span class="co">#&gt; [243] "REXO4"                   "RHOH"                   </span></span>
<span><span class="co">#&gt; [245] "RORC"                    "RTKN2"                  </span></span>
<span><span class="co">#&gt; [247] "RUNX1"                   "S100A14"                </span></span>
<span><span class="co">#&gt; [249] "S100A4"                  "S100A8"                 </span></span>
<span><span class="co">#&gt; [251] "SCD"                     "SCGB2A1"                </span></span>
<span><span class="co">#&gt; [253] "SDC4"                    "SEC11C"                 </span></span>
<span><span class="co">#&gt; [255] "SEC24A"                  "SELL"                   </span></span>
<span><span class="co">#&gt; [257] "SERHL2"                  "SERPINA3"               </span></span>
<span><span class="co">#&gt; [259] "SERPINB9"                "SFRP1"                  </span></span>
<span><span class="co">#&gt; [261] "SFRP4"                   "SH3YL1"                 </span></span>
<span><span class="co">#&gt; [263] "SLAMF1"                  "SLAMF7"                 </span></span>
<span><span class="co">#&gt; [265] "SLC25A37"                "SLC4A1"                 </span></span>
<span><span class="co">#&gt; [267] "SLC5A6"                  "SMAP2"                  </span></span>
<span><span class="co">#&gt; [269] "SMS"                     "SNAI1"                  </span></span>
<span><span class="co">#&gt; [271] "SOX17"                   "SOX18"                  </span></span>
<span><span class="co">#&gt; [273] "SPIB"                    "SQLE"                   </span></span>
<span><span class="co">#&gt; [275] "SRPK1"                   "SSTR2"                  </span></span>
<span><span class="co">#&gt; [277] "STC1"                    "SVIL"                   </span></span>
<span><span class="co">#&gt; [279] "TAC1"                    "TACSTD2"                </span></span>
<span><span class="co">#&gt; [281] "TCEAL7"                  "TCF15"                  </span></span>
<span><span class="co">#&gt; [283] "TCF4"                    "TCF7"                   </span></span>
<span><span class="co">#&gt; [285] "TCIM"                    "TCL1A"                  </span></span>
<span><span class="co">#&gt; [287] "TENT5C"                  "TFAP2A"                 </span></span>
<span><span class="co">#&gt; [289] "THAP2"                   "TIFA"                   </span></span>
<span><span class="co">#&gt; [291] "TIGIT"                   "TIMP4"                  </span></span>
<span><span class="co">#&gt; [293] "TMEM147"                 "TNFRSF17"               </span></span>
<span><span class="co">#&gt; [295] "TOMM7"                   "TOP2A"                  </span></span>
<span><span class="co">#&gt; [297] "TPD52"                   "TPSAB1"                 </span></span>
<span><span class="co">#&gt; [299] "TRAC"                    "TRAF4"                  </span></span>
<span><span class="co">#&gt; [301] "TRAPPC3"                 "TRIB1"                  </span></span>
<span><span class="co">#&gt; [303] "TUBA4A"                  "TUBB2B"                 </span></span>
<span><span class="co">#&gt; [305] "TYROBP"                  "UCP1"                   </span></span>
<span><span class="co">#&gt; [307] "USP53"                   "VOPP1"                  </span></span>
<span><span class="co">#&gt; [309] "VWF"                     "WARS"                   </span></span>
<span><span class="co">#&gt; [311] "ZEB1"                    "ZEB2"                   </span></span>
<span><span class="co">#&gt; [313] "ZNF562"                  "NegControlProbe_00042"  </span></span>
<span><span class="co">#&gt; [315] "NegControlProbe_00041"   "NegControlProbe_00039"  </span></span>
<span><span class="co">#&gt; [317] "NegControlProbe_00035"   "NegControlProbe_00034"  </span></span>
<span><span class="co">#&gt; [319] "NegControlProbe_00033"   "NegControlProbe_00031"  </span></span>
<span><span class="co">#&gt; [321] "NegControlProbe_00025"   "NegControlProbe_00024"  </span></span>
<span><span class="co">#&gt; [323] "NegControlProbe_00022"   "NegControlProbe_00019"  </span></span>
<span><span class="co">#&gt; [325] "NegControlProbe_00017"   "NegControlProbe_00016"  </span></span>
<span><span class="co">#&gt; [327] "NegControlProbe_00014"   "NegControlProbe_00013"  </span></span>
<span><span class="co">#&gt; [329] "NegControlProbe_00012"   "NegControlProbe_00009"  </span></span>
<span><span class="co">#&gt; [331] "NegControlProbe_00004"   "NegControlProbe_00003"  </span></span>
<span><span class="co">#&gt; [333] "NegControlProbe_00002"   "antisense_PROKR2"       </span></span>
<span><span class="co">#&gt; [335] "antisense_ULK3"          "antisense_SCRIB"        </span></span>
<span><span class="co">#&gt; [337] "antisense_TRMU"          "antisense_MYLIP"        </span></span>
<span><span class="co">#&gt; [339] "antisense_LGI3"          "antisense_BCL2L15"      </span></span>
<span><span class="co">#&gt; [341] "antisense_ADCY4"         "NegControlCodeword_0500"</span></span>
<span><span class="co">#&gt; [343] "NegControlCodeword_0501" "NegControlCodeword_0502"</span></span>
<span><span class="co">#&gt; [345] "NegControlCodeword_0503" "NegControlCodeword_0504"</span></span>
<span><span class="co">#&gt; [347] "NegControlCodeword_0505" "NegControlCodeword_0506"</span></span>
<span><span class="co">#&gt; [349] "NegControlCodeword_0507" "NegControlCodeword_0508"</span></span>
<span><span class="co">#&gt; [351] "NegControlCodeword_0509" "NegControlCodeword_0510"</span></span>
<span><span class="co">#&gt; [353] "NegControlCodeword_0511" "NegControlCodeword_0512"</span></span>
<span><span class="co">#&gt; [355] "NegControlCodeword_0513" "NegControlCodeword_0514"</span></span>
<span><span class="co">#&gt; [357] "NegControlCodeword_0515" "NegControlCodeword_0516"</span></span>
<span><span class="co">#&gt; [359] "NegControlCodeword_0517" "NegControlCodeword_0518"</span></span>
<span><span class="co">#&gt; [361] "NegControlCodeword_0519" "NegControlCodeword_0520"</span></span>
<span><span class="co">#&gt; [363] "NegControlCodeword_0521" "NegControlCodeword_0522"</span></span>
<span><span class="co">#&gt; [365] "NegControlCodeword_0523" "NegControlCodeword_0524"</span></span>
<span><span class="co">#&gt; [367] "NegControlCodeword_0525" "NegControlCodeword_0526"</span></span>
<span><span class="co">#&gt; [369] "NegControlCodeword_0527" "NegControlCodeword_0528"</span></span>
<span><span class="co">#&gt; [371] "NegControlCodeword_0529" "NegControlCodeword_0530"</span></span>
<span><span class="co">#&gt; [373] "NegControlCodeword_0531" "NegControlCodeword_0532"</span></span>
<span><span class="co">#&gt; [375] "NegControlCodeword_0533" "NegControlCodeword_0534"</span></span>
<span><span class="co">#&gt; [377] "NegControlCodeword_0535" "NegControlCodeword_0536"</span></span>
<span><span class="co">#&gt; [379] "NegControlCodeword_0537" "NegControlCodeword_0538"</span></span>
<span><span class="co">#&gt; [381] "NegControlCodeword_0539" "NegControlCodeword_0540"</span></span>
<span><span class="co">#&gt; [383] "BLANK_0006"              "BLANK_0013"             </span></span>
<span><span class="co">#&gt; [385] "BLANK_0037"              "BLANK_0069"             </span></span>
<span><span class="co">#&gt; [387] "BLANK_0072"              "BLANK_0087"             </span></span>
<span><span class="co">#&gt; [389] "BLANK_0110"              "BLANK_0114"             </span></span>
<span><span class="co">#&gt; [391] "BLANK_0120"              "BLANK_0147"             </span></span>
<span><span class="co">#&gt; [393] "BLANK_0180"              "BLANK_0186"             </span></span>
<span><span class="co">#&gt; [395] "BLANK_0272"              "BLANK_0278"             </span></span>
<span><span class="co">#&gt; [397] "BLANK_0319"              "BLANK_0321"             </span></span>
<span><span class="co">#&gt; [399] "BLANK_0337"              "BLANK_0350"             </span></span>
<span><span class="co">#&gt; [401] "BLANK_0351"              "BLANK_0352"             </span></span>
<span><span class="co">#&gt; [403] "BLANK_0353"              "BLANK_0354"             </span></span>
<span><span class="co">#&gt; [405] "BLANK_0355"              "BLANK_0356"             </span></span>
<span><span class="co">#&gt; [407] "BLANK_0357"              "BLANK_0358"             </span></span>
<span><span class="co">#&gt; [409] "BLANK_0359"              "BLANK_0360"             </span></span>
<span><span class="co">#&gt; [411] "BLANK_0361"              "BLANK_0362"             </span></span>
<span><span class="co">#&gt; [413] "BLANK_0363"              "BLANK_0364"             </span></span>
<span><span class="co">#&gt; [415] "BLANK_0365"              "BLANK_0366"             </span></span>
<span><span class="co">#&gt; [417] "BLANK_0367"              "BLANK_0368"             </span></span>
<span><span class="co">#&gt; [419] "BLANK_0369"              "BLANK_0370"             </span></span>
<span><span class="co">#&gt; [421] "BLANK_0371"              "BLANK_0372"             </span></span>
<span><span class="co">#&gt; [423] "BLANK_0373"              "BLANK_0374"             </span></span>
<span><span class="co">#&gt; [425] "BLANK_0375"              "BLANK_0376"             </span></span>
<span><span class="co">#&gt; [427] "BLANK_0377"              "BLANK_0378"             </span></span>
<span><span class="co">#&gt; [429] "BLANK_0379"              "BLANK_0380"             </span></span>
<span><span class="co">#&gt; [431] "BLANK_0381"              "BLANK_0382"             </span></span>
<span><span class="co">#&gt; [433] "BLANK_0383"              "BLANK_0384"             </span></span>
<span><span class="co">#&gt; [435] "BLANK_0385"              "BLANK_0386"             </span></span>
<span><span class="co">#&gt; [437] "BLANK_0387"              "BLANK_0388"             </span></span>
<span><span class="co">#&gt; [439] "BLANK_0389"              "BLANK_0390"             </span></span>
<span><span class="co">#&gt; [441] "BLANK_0391"              "BLANK_0392"             </span></span>
<span><span class="co">#&gt; [443] "BLANK_0393"              "BLANK_0394"             </span></span>
<span><span class="co">#&gt; [445] "BLANK_0395"              "BLANK_0396"             </span></span>
<span><span class="co">#&gt; [447] "BLANK_0397"              "BLANK_0398"             </span></span>
<span><span class="co">#&gt; [449] "BLANK_0399"              "BLANK_0400"             </span></span>
<span><span class="co">#&gt; [451] "BLANK_0401"              "BLANK_0402"             </span></span>
<span><span class="co">#&gt; [453] "BLANK_0403"              "BLANK_0404"             </span></span>
<span><span class="co">#&gt; [455] "BLANK_0405"              "BLANK_0406"             </span></span>
<span><span class="co">#&gt; [457] "BLANK_0407"              "BLANK_0408"             </span></span>
<span><span class="co">#&gt; [459] "BLANK_0409"              "BLANK_0410"             </span></span>
<span><span class="co">#&gt; [461] "BLANK_0411"              "BLANK_0412"             </span></span>
<span><span class="co">#&gt; [463] "BLANK_0413"              "BLANK_0414"             </span></span>
<span><span class="co">#&gt; [465] "BLANK_0415"              "BLANK_0416"             </span></span>
<span><span class="co">#&gt; [467] "BLANK_0417"              "BLANK_0418"             </span></span>
<span><span class="co">#&gt; [469] "BLANK_0419"              "BLANK_0420"             </span></span>
<span><span class="co">#&gt; [471] "BLANK_0421"              "BLANK_0422"             </span></span>
<span><span class="co">#&gt; [473] "BLANK_0423"              "BLANK_0424"             </span></span>
<span><span class="co">#&gt; [475] "BLANK_0425"              "BLANK_0426"             </span></span>
<span><span class="co">#&gt; [477] "BLANK_0427"              "BLANK_0428"             </span></span>
<span><span class="co">#&gt; [479] "BLANK_0429"              "BLANK_0430"             </span></span>
<span><span class="co">#&gt; [481] "BLANK_0431"              "BLANK_0432"             </span></span>
<span><span class="co">#&gt; [483] "BLANK_0433"              "BLANK_0434"             </span></span>
<span><span class="co">#&gt; [485] "BLANK_0435"              "BLANK_0436"             </span></span>
<span><span class="co">#&gt; [487] "BLANK_0437"              "BLANK_0438"             </span></span>
<span><span class="co">#&gt; [489] "BLANK_0439"              "BLANK_0440"             </span></span>
<span><span class="co">#&gt; [491] "BLANK_0441"              "BLANK_0442"             </span></span>
<span><span class="co">#&gt; [493] "BLANK_0443"              "BLANK_0444"             </span></span>
<span><span class="co">#&gt; [495] "BLANK_0445"              "BLANK_0446"             </span></span>
<span><span class="co">#&gt; [497] "BLANK_0447"              "BLANK_0448"             </span></span>
<span><span class="co">#&gt; [499] "BLANK_0449"              "BLANK_0450"             </span></span>
<span><span class="co">#&gt; [501] "BLANK_0451"              "BLANK_0452"             </span></span>
<span><span class="co">#&gt; [503] "BLANK_0453"              "BLANK_0454"             </span></span>
<span><span class="co">#&gt; [505] "BLANK_0455"              "BLANK_0456"             </span></span>
<span><span class="co">#&gt; [507] "BLANK_0457"              "BLANK_0458"             </span></span>
<span><span class="co">#&gt; [509] "BLANK_0459"              "BLANK_0460"             </span></span>
<span><span class="co">#&gt; [511] "BLANK_0461"              "BLANK_0462"             </span></span>
<span><span class="co">#&gt; [513] "BLANK_0463"              "BLANK_0464"             </span></span>
<span><span class="co">#&gt; [515] "BLANK_0465"              "BLANK_0466"             </span></span>
<span><span class="co">#&gt; [517] "BLANK_0467"              "BLANK_0468"             </span></span>
<span><span class="co">#&gt; [519] "BLANK_0469"              "BLANK_0470"             </span></span>
<span><span class="co">#&gt; [521] "BLANK_0471"              "BLANK_0472"             </span></span>
<span><span class="co">#&gt; [523] "BLANK_0473"              "BLANK_0474"             </span></span>
<span><span class="co">#&gt; [525] "BLANK_0475"              "BLANK_0476"             </span></span>
<span><span class="co">#&gt; [527] "BLANK_0477"              "BLANK_0478"             </span></span>
<span><span class="co">#&gt; [529] "BLANK_0479"              "BLANK_0480"             </span></span>
<span><span class="co">#&gt; [531] "BLANK_0481"              "BLANK_0482"             </span></span>
<span><span class="co">#&gt; [533] "BLANK_0483"              "BLANK_0484"             </span></span>
<span><span class="co">#&gt; [535] "BLANK_0485"              "BLANK_0486"             </span></span>
<span><span class="co">#&gt; [537] "BLANK_0487"              "BLANK_0488"             </span></span>
<span><span class="co">#&gt; [539] "BLANK_0489"              "BLANK_0497"             </span></span>
<span><span class="co">#&gt; [541] "BLANK_0499"</span></span></code></pre></div>
<p>According to the Xenium paper <span class="citation">(Janesick et al.
2022)</span>, there are 3 types of controls:</p>
<blockquote>
<ol style="list-style-type: decimal">
<li>probe controls to assess non-specific binding to RNA,</li>
<li>decoding controls to assess misassigned genes, and</li>
<li>genomic DNA (gDNA) controls to ensure the signal is from RNA.</li>
</ol>
</blockquote>
<p>The paper does not explain in detail how those control probes were
designed, nor explain what the blank probes are. But the blank probes
can be used as a negative control.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">is_blank</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span>, <span class="st">"^BLANK_"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">is_blank</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 159</span></span></code></pre></div>
<p>This should be number 1, the probe control</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">is_neg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span>, <span class="st">"^NegControlProbe"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">is_neg</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 20</span></span></code></pre></div>
<p>This should be number 2, the decoding control</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">is_neg2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span>, <span class="st">"^NegControlCodeword"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">is_neg2</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 41</span></span></code></pre></div>
<p>This must be number 3, gDNA control</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">is_anti</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span>, <span class="st">"^antisense"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">is_anti</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 8</span></span></code></pre></div>
<p>Also make an indicator of whether a feature is any sort of negative
control</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">is_any_neg</span> <span class="op">&lt;-</span> <span class="va">is_blank</span> <span class="op">|</span> <span class="va">is_neg</span> <span class="op">|</span> <span class="va">is_neg2</span> <span class="op">|</span> <span class="va">is_anti</span></span></code></pre></div>
<p>The <code><a href="https://rdrr.io/pkg/scuttle/man/addPerCellQCMetrics.html" class="external-link">addPerCellQCMetrics()</a></code> function in the
<code>scuttle</code> package can conveniently add transcript counts,
proportion of total counts, and number of features detected for any
subset of features to the SCE object. Here we do this for the SFE
object, as SFE inherits from SCE.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/addPerCellQCMetrics.html" class="external-link">addPerCellQCMetrics</a></span><span class="op">(</span><span class="va">sfe</span>, subsets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>blank <span class="op">=</span> <span class="va">is_blank</span>,</span>
<span>                                               negProbe <span class="op">=</span> <span class="va">is_neg</span>,</span>
<span>                                               negCodeword <span class="op">=</span> <span class="va">is_neg2</span>,</span>
<span>                                               anti <span class="op">=</span> <span class="va">is_anti</span>,</span>
<span>                                               any_neg <span class="op">=</span> <span class="va">is_any_neg</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "Sample"                       "Barcode"                     </span></span>
<span><span class="co">#&gt;  [3] "transcript_counts"            "control_probe_counts"        </span></span>
<span><span class="co">#&gt;  [5] "control_codeword_counts"      "cell_area"                   </span></span>
<span><span class="co">#&gt;  [7] "nucleus_area"                 "sample_id"                   </span></span>
<span><span class="co">#&gt;  [9] "nCounts"                      "nGenes"                      </span></span>
<span><span class="co">#&gt; [11] "nCounts_normed"               "nGenes_normed"               </span></span>
<span><span class="co">#&gt; [13] "prop_nuc"                     "sum"                         </span></span>
<span><span class="co">#&gt; [15] "detected"                     "subsets_blank_sum"           </span></span>
<span><span class="co">#&gt; [17] "subsets_blank_detected"       "subsets_blank_percent"       </span></span>
<span><span class="co">#&gt; [19] "subsets_negProbe_sum"         "subsets_negProbe_detected"   </span></span>
<span><span class="co">#&gt; [21] "subsets_negProbe_percent"     "subsets_negCodeword_sum"     </span></span>
<span><span class="co">#&gt; [23] "subsets_negCodeword_detected" "subsets_negCodeword_percent" </span></span>
<span><span class="co">#&gt; [25] "subsets_anti_sum"             "subsets_anti_detected"       </span></span>
<span><span class="co">#&gt; [27] "subsets_anti_percent"         "subsets_any_neg_sum"         </span></span>
<span><span class="co">#&gt; [29] "subsets_any_neg_detected"     "subsets_any_neg_percent"     </span></span>
<span><span class="co">#&gt; [31] "total"</span></span></code></pre></div>
<p>Next we plot the proportion of transcript counts coming from any
negative control.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cols_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span>, <span class="st">"_percent$"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/plotColDataHistogram.html">plotColDataHistogram</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">cols_use</span>, bins <span class="op">=</span> <span class="fl">100</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 285 rows containing non-finite values (`stat_bin()`).</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-31-1.png" width="700" style="display: block; margin: auto;"></p>
<p>The histogram is dominated by the bin at zero and there are some
extreme outliers too few to be seen but evident from the scale of the x
axis. We also plot the histogram only for cells with at least 1 count
from a negative control. The NA’s come from cells that got segmented but
have no transcripts detected.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotColDataHistogram.html">plotColDataHistogram</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">cols_use</span>, bins <span class="op">=</span> <span class="fl">100</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotation_logticks.html" class="external-link">annotation_logticks</a></span><span class="op">(</span>sides <span class="op">=</span> <span class="st">"b"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Transformation introduced infinite values in continuous x-axis</span></span>
<span><span class="co">#&gt; Warning: Removed 565577 rows containing non-finite values</span></span>
<span><span class="co">#&gt; (`stat_bin()`).</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-32-1.png" width="700" style="display: block; margin: auto;"></p>
<p>The vast majority of these cells have less than 1% of transcript
counts from negative controls, but there are outliers with up to
50%.</p>
<p>Next we plot the distribution of the number of negative control
counts per cell:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cols_use2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span>, <span class="st">"_detected$"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/plotColDataHistogram.html">plotColDataHistogram</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">cols_use2</span>, bins <span class="op">=</span> <span class="fl">20</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="co"># Avoid decimal breaks on x axis unless there're too few breaks</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/breaks_extended.html" class="external-link">breaks_extended</a></span><span class="op">(</span>Q <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-33-1.png" width="700" style="display: block; margin: auto;"></p>
<p>The counts are low, mostly zero, but there are outliers with up to 10
counts of all types aggregated. Then the outlier with 50% of counts from
negative controls must have very low total real transcript counts to
begin with.</p>
<p>The <code>scuttle</code> package can detect outliers, but by default
it assigns anything above zero as an outlier, since that is over 3
median absolute deviations (MADs) away from the median, which is 0, and
the MAD is 0 since the vast majority of cells don’t have any negative
control count. But it makes sense to allow a small proportion of
negative controls. Here we use the distribution just for cells with at
least 1 negative control count to find outliers. This distribution has a
very long tail and some definite outliers.</p>
<p>The code below extracts the outliers, based only on cells with at
least one negative control count</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">get_neg_ctrl_outliers</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">col</span>, <span class="va">sfe</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">inds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">nCounts</span> <span class="op">&gt;</span> <span class="fl">0</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[[</span><span class="va">col</span><span class="op">]</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">0</span></span>
<span>    <span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span><span class="va">inds</span>,<span class="op">]</span></span>
<span>    <span class="va">outlier_inds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html" class="external-link">isOutlier</a></span><span class="op">(</span><span class="va">df</span><span class="op">[[</span><span class="va">col</span><span class="op">]</span><span class="op">]</span>, type <span class="op">=</span> <span class="st">"higher"</span><span class="op">)</span></span>
<span>    <span class="va">outliers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span><span class="op">[</span><span class="va">outlier_inds</span><span class="op">]</span></span>
<span>    <span class="va">col2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html" class="external-link">str_remove</a></span><span class="op">(</span><span class="va">col</span>, <span class="st">"^subsets_"</span><span class="op">)</span></span>
<span>    <span class="va">col2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html" class="external-link">str_remove</a></span><span class="op">(</span><span class="va">col2</span>, <span class="st">"_percent$"</span><span class="op">)</span></span>
<span>    <span class="va">new_colname</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"is"</span>, <span class="va">col2</span>, <span class="st">"outlier"</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[[</span><span class="va">new_colname</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">outliers</span></span>
<span>    <span class="va">sfe</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cols_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span>, <span class="st">"_percent$"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">n</span> <span class="kw">in</span> <span class="va">cols_use</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu">get_neg_ctrl_outliers</span><span class="op">(</span><span class="va">n</span>, <span class="va">sfe</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "Sample"                       "Barcode"                     </span></span>
<span><span class="co">#&gt;  [3] "transcript_counts"            "control_probe_counts"        </span></span>
<span><span class="co">#&gt;  [5] "control_codeword_counts"      "cell_area"                   </span></span>
<span><span class="co">#&gt;  [7] "nucleus_area"                 "sample_id"                   </span></span>
<span><span class="co">#&gt;  [9] "nCounts"                      "nGenes"                      </span></span>
<span><span class="co">#&gt; [11] "nCounts_normed"               "nGenes_normed"               </span></span>
<span><span class="co">#&gt; [13] "prop_nuc"                     "sum"                         </span></span>
<span><span class="co">#&gt; [15] "detected"                     "subsets_blank_sum"           </span></span>
<span><span class="co">#&gt; [17] "subsets_blank_detected"       "subsets_blank_percent"       </span></span>
<span><span class="co">#&gt; [19] "subsets_negProbe_sum"         "subsets_negProbe_detected"   </span></span>
<span><span class="co">#&gt; [21] "subsets_negProbe_percent"     "subsets_negCodeword_sum"     </span></span>
<span><span class="co">#&gt; [23] "subsets_negCodeword_detected" "subsets_negCodeword_percent" </span></span>
<span><span class="co">#&gt; [25] "subsets_anti_sum"             "subsets_anti_detected"       </span></span>
<span><span class="co">#&gt; [27] "subsets_anti_percent"         "subsets_any_neg_sum"         </span></span>
<span><span class="co">#&gt; [29] "subsets_any_neg_detected"     "subsets_any_neg_percent"     </span></span>
<span><span class="co">#&gt; [31] "total"                        "is_blank_outlier"            </span></span>
<span><span class="co">#&gt; [33] "is_negProbe_outlier"          "is_negCodeword_outlier"      </span></span>
<span><span class="co">#&gt; [35] "is_anti_outlier"              "is_any_neg_outlier"</span></span></code></pre></div>
<p>Below we examine where the outliers are located in space:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"is_blank_outlier"</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-37-1.png" width="700" style="display: block; margin: auto;"></p>
<p>We find that the outliers are difficult to see:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe</span>, y <span class="op">=</span> <span class="st">"is_blank_outlier"</span>, x <span class="op">=</span> <span class="st">"cell_area"</span>, </span>
<span>            point_fun <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">...</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> </span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-38-1.png" width="700" style="display: block; margin: auto;"></p>
<p>The analysis reveals that the outliers seem to be smaller. Outliers
for negative probe controls and negative codeword controls are also hard
to see on the plot, so their plots are skipped here. But the top left
region in the tissue tends to have more counts from antisense
controls.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"is_anti_outlier"</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-39-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Now that we have identified the outliers, we can remove them along
with empty cells before proceeding to further analysis:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">inds_keep</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">$</span><span class="va">nCounts</span> <span class="op">&gt;</span> <span class="fl">0</span> <span class="op">&amp;</span> <span class="va">sfe</span><span class="op">$</span><span class="va">nucleus_area</span> <span class="op">&lt;</span> <span class="fl">400</span> <span class="op">&amp;</span> <span class="op">!</span><span class="va">sfe</span><span class="op">$</span><span class="va">is_anti_outlier</span> <span class="op">&amp;</span></span>
<span>    <span class="op">!</span><span class="va">sfe</span><span class="op">$</span><span class="va">is_blank_outlier</span> <span class="op">&amp;</span> <span class="op">!</span><span class="va">sfe</span><span class="op">$</span><span class="va">is_negCodeword_outlier</span> <span class="op">&amp;</span> <span class="op">!</span><span class="va">sfe</span><span class="op">$</span><span class="va">is_negProbe_outlier</span></span>
<span><span class="op">(</span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">[</span>,<span class="va">inds_keep</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 541 117503 </span></span>
<span><span class="co">#&gt; metadata(1): Samples</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(541): ABCC11 ACTA2 ... BLANK_0497 BLANK_0499</span></span>
<span><span class="co">#&gt; rowData names(6): ID Symbol ... vars cv2</span></span>
<span><span class="co">#&gt; colnames(117503): 1 2 ... 118707 118708</span></span>
<span><span class="co">#&gt; colData names(36): Sample Barcode ... is_anti_outlier</span></span>
<span><span class="co">#&gt;   is_any_neg_outlier</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : x_centroid y_centroid</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit:</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: centroids (POINT), cellSeg (POLYGON), nucSeg (GEOMETRY) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01:</span></span></code></pre></div>
<p>Over 1000 cells were removed.</p>
<p>Next we check how many negative control features are detected per
cell:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotColDataHistogram.html">plotColDataHistogram</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">cols_use2</span>, bins <span class="op">=</span> <span class="fl">20</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="co"># Avoid decimal breaks on x axis unless there're too few breaks</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/breaks_extended.html" class="external-link">breaks_extended</a></span><span class="op">(</span><span class="fl">3</span>, Q <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-41-1.png" width="700" style="display: block; margin: auto;"></p>
<p>There are at most 3 counts per cell per type. For the non-outliers,
each type is at most around 1%, so this data looks good.</p>
</div>
<div class="section level3">
<h3 id="genes">Genes<a class="anchor" aria-label="anchor" href="#genes"></a>
</h3>
<p>Here we look at the mean and variance of each gene</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">means</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowMeans.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">vars</span> <span class="op">&lt;-</span> <span class="fu">rowVars</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Real genes generally have higher mean expression across cells than
negative controls.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">is_neg</span> <span class="op">&lt;-</span> <span class="va">is_any_neg</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotRowData.html" class="external-link">plotRowData</a></span><span class="op">(</span><span class="va">sfe</span>, x <span class="op">=</span> <span class="st">"means"</span>, y <span class="op">=</span> <span class="st">"is_neg"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotation_logticks.html" class="external-link">annotation_logticks</a></span><span class="op">(</span>sides <span class="op">=</span> <span class="st">"b"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-43-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Here the real genes and negative controls are plotted in different
colors</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotColDataBin2D.html">plotRowDataBin2D</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"means"</span>, <span class="st">"vars"</span>, subset <span class="op">=</span> <span class="st">"is_neg"</span>, </span>
<span>                 name_true <span class="op">=</span> <span class="st">"Counts (negative controls)"</span>, </span>
<span>                 name_false <span class="op">=</span> <span class="st">"Counts (real genes)"</span>, bins <span class="op">=</span> <span class="fl">50</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span>slope <span class="op">=</span> <span class="fl">1</span>, intercept <span class="op">=</span> <span class="fl">0</span>, color <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotation_logticks.html" class="external-link">annotation_logticks</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_equal</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-44-1.png" width="700" style="display: block; margin: auto;"></p>
<p>The red line <span class="math inline">\(y = x\)</span> is expected
if the data follows a Poisson distribution. Negative controls and real
genes form mostly separate clusters. Negative controls stick close to
the line, while real genes are overdispersed. Unlike in the <a href="https://pachterlab.github.io/voyager/articles/vig4_cosmx.html#genes">CosMX
dataset</a>, the negative controls don’t seem overdispersed.</p>
</div>
</div>
<div class="section level2">
<h2 id="spatial-autocorrelation-of-qc-metrics">Spatial autocorrelation of QC metrics<a class="anchor" aria-label="anchor" href="#spatial-autocorrelation-of-qc-metrics"></a>
</h2>
<p>There’s a sparse and a dense region. This poses the question of what
type of neighborhood graph to use, e.g. it is conceivable that cells in
the sparse region should just be singletons. Furthermore, it is unclear
what the length scale of their influence might be. It might depend on
the cell type and how contact and secreted signals are used in the cell
type, and length scale of the influence. If k nearest neighbors are
used, then the neighbors in the dense region are much closer together
than those in the sparse region. If distance based neighbors are used,
then cells in the dense region will have more neighbors than cells in
the sparse region, and the sparse region can break into multiple
compartments if the distance cutoff is not long enough.</p>
<p>For the purpose of demonstration, we use k nearest neighbors with
<span class="math inline">\(k = 5\)</span>, with inverse distance
weighting. Note that using more neighbors leads to longer computation
time of spatial autocorrelation metrics.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"knn5"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findSpatialNeighbors-SpatialFeatureExperiment-method.html" class="external-link">findSpatialNeighbors</a></span><span class="op">(</span><span class="va">sfe</span>, method <span class="op">=</span> <span class="st">"knearneigh"</span>, </span>
<span>                                                  dist_type <span class="op">=</span> <span class="st">"idw"</span>, k <span class="op">=</span> <span class="fl">5</span>, </span>
<span>                                                  style <span class="op">=</span> <span class="st">"W"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt;  11.095   0.048  11.150</span></span></code></pre></div>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">colDataMoransI</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span>, <span class="st">"cell_area"</span>, <span class="st">"nucleus_area"</span><span class="op">)</span>,</span>
<span>                      colGraphName <span class="op">=</span> <span class="st">"knn5"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/colFeatureData.html">colFeatureData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span>, <span class="st">"cell_area"</span>, <span class="st">"nucleus_area"</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="co">#&gt; DataFrame with 4 rows and 2 columns</span></span>
<span><span class="co">#&gt;              moran_sample01 K_sample01</span></span>
<span><span class="co">#&gt;                   &lt;numeric&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; nCounts            0.422387    5.55509</span></span>
<span><span class="co">#&gt; nGenes             0.401395    3.13694</span></span>
<span><span class="co">#&gt; cell_area          0.628837    7.57098</span></span>
<span><span class="co">#&gt; nucleus_area       0.377248    6.88331</span></span></code></pre></div>
<p>Global Moran’s I indicatse positive spatial autocorrelation. As the
strength of spatial autocorrelation can vary spatially, we also run
local Moran’s I.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">colDataUnivariate</a></span><span class="op">(</span><span class="va">sfe</span>, type <span class="op">=</span> <span class="st">"localmoran"</span>, </span>
<span>                         features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span>, <span class="st">"cell_area"</span>, </span>
<span>                                      <span class="st">"nucleus_area"</span><span class="op">)</span>,</span>
<span>                         colGraphName <span class="op">=</span> <span class="st">"knn5"</span>, BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The <code>pointsize</code> argument adjusts the point size in
<code>scattermore</code>. The default is 0, meaning single pixels, but
since the cells in the sparse region are hard to see that way, we
increase <code>pointsize</code>. We would still plot the polygons in
larger single panel plots, but use <code>scattermore</code> in
multi-panel plots where the polygons in each panel are invisible anyway
due to the small size to save some time.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"localmoran"</span>,</span>
<span>                features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span>, <span class="st">"cell_area"</span>, <span class="st">"nucleus_area"</span><span class="op">)</span>,</span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>, scattermore <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                divergent <span class="op">=</span> <span class="cn">TRUE</span>, diverge_center <span class="op">=</span> <span class="fl">0</span>, pointsize <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-49-1.png" width="864" style="display: block; margin: auto;"></p>
<p>Interestingly, nCounts is more homogeneous in the interior of the
dense region, while nGenes is more homogeneous by the edge of the dense
region. As expected, cell area is more homogeneous in the sparse region.
However, the nucleus area is more homogeneous in the interior of the
dense region.</p>
<p>Moran plot for nCounts</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">colDataUnivariate</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"moran.plot"</span>, <span class="st">"nCounts"</span>, colGraphName <span class="op">=</span> <span class="st">"knn5"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/moranPlot.html">moranPlot</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nCounts"</span>, binned <span class="op">=</span> <span class="cn">TRUE</span>, plot_influential <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> </span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/moranPlot.html">moranPlot</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"nCounts"</span>, binned <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">/</span> <span class="va">p2</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-51-1.png" width="768" style="display: block; margin: auto;"></p>
<p>There are no obvious clusters here. In the lower panel, the 2D
histogram of influential points is plotted in red.</p>
</div>
<div class="section level2">
<h2 id="morans-i">Moran’s I<a class="anchor" aria-label="anchor" href="#morans-i"></a>
</h2>
<p>By default, for gene expression, the log normalized counts are used
in spatial autocorrelation metrics, so before running Moran’s I, we
normalize the data.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span></code></pre></div>
<p>Use more cores if available to speed this up.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span></span>
<span>    <span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">runMoransI</a></span><span class="op">(</span><span class="va">sfe</span>, colGraphName <span class="op">=</span> <span class="st">"knn5"</span>, BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt;  38.092   3.954  21.959</span></span></code></pre></div>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">is_neg</span> <span class="op">&lt;-</span> <span class="va">is_any_neg</span></span></code></pre></div>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotRowData.html" class="external-link">plotRowData</a></span><span class="op">(</span><span class="va">sfe</span>, x <span class="op">=</span> <span class="st">"moran_sample01"</span>, y <span class="op">=</span> <span class="st">"is_neg"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-55-1.png" width="700" style="display: block; margin: auto;"></p>
<p>As expected, generally the negative controls are tightly clustered
around 0, while the real genes have positive Moran’s I, which means
there is generally no technical artifact spatial trend. No significantly
negative Moran’s I is observed. Why is negative spatial autocorrelation
so rare in gene expression?</p>
<p>What are the two negative controls with a sizable Moran’s I?</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ord</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">moran_sample01</span><span class="op">[</span><span class="va">is_any_neg</span><span class="op">]</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="va">top_neg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span><span class="va">is_any_neg</span><span class="op">]</span><span class="op">[</span><span class="va">ord</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">top_neg</span>, colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>,</span>
<span>                   scattermore <span class="op">=</span> <span class="cn">TRUE</span>, pointsize <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-56-1.png" width="864" style="display: block; margin: auto;"></p>
<p>There is somewhat a spatial trend for that antisense probe, with more
detected in the upper left. However, this might not significantly affect
other results since there are at most 2 counts and at most about 1% of
all counts in each cell. The negative control codeword has at most 1
count per cell and the cells with this negative control detected seem to
be few and far between.</p>
<p>These are the most detected negative controls, and the most detected
one is also the one with the highest Moran’s I among negative controls.
However, the other negative control with higher Moran’s I is not among
the most detected.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">means</span><span class="op">[</span><span class="va">is_any_neg</span><span class="op">]</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="fl">15</span><span class="op">)</span></span>
<span><span class="co">#&gt;      antisense_PROKR2       antisense_SCRIB     antisense_BCL2L15 </span></span>
<span><span class="co">#&gt;          0.0192761036          0.0131741317          0.0066806805 </span></span>
<span><span class="co">#&gt;        antisense_TRMU       antisense_MYLIP        antisense_ULK3 </span></span>
<span><span class="co">#&gt;          0.0042041480          0.0030807724          0.0028169494 </span></span>
<span><span class="co">#&gt;            BLANK_0485       antisense_ADCY4        antisense_LGI3 </span></span>
<span><span class="co">#&gt;          0.0023403658          0.0019829281          0.0017871884 </span></span>
<span><span class="co">#&gt;            BLANK_0430 NegControlProbe_00035 NegControlProbe_00012 </span></span>
<span><span class="co">#&gt;          0.0015063445          0.0010978443          0.0010382714 </span></span>
<span><span class="co">#&gt; NegControlProbe_00033 NegControlProbe_00014            BLANK_0120 </span></span>
<span><span class="co">#&gt;          0.0009446567          0.0009361463          0.0009276359</span></span></code></pre></div>
<p>What are the genes with the highest Moran’s I?</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_moran</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">moran_sample01</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">top_moran</span>, colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>,</span>
<span>                   scattermore <span class="op">=</span> <span class="cn">TRUE</span>, ncol <span class="op">=</span> <span class="fl">2</span>, pointsize <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-58-1.png" width="864" style="display: block; margin: auto;"></p>
<p>They all highlight the same histological regions, as in the <a href="https://pachterlab.github.io/voyager/articles/vig4_cosmx.html#morans-i">CosMX
vignette</a>. How does Moran’s I relate to gene expression level?</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotRowData.html" class="external-link">plotRowData</a></span><span class="op">(</span><span class="va">sfe</span>, x <span class="op">=</span> <span class="st">"means"</span>, y <span class="op">=</span> <span class="st">"moran_sample01"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-59-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Very highly expressed genes have higher Moran’s I, but there are some
less expressed genes with higher Moran’s I as well.</p>
</div>
<div class="section level2">
<h2 id="non-spatial-dimension-reduction-and-clustering">Non-spatial dimension reduction and clustering<a class="anchor" aria-label="anchor" href="#non-spatial-dimension-reduction-and-clustering"></a>
</h2>
<p>Here we run non-spatial PCA as for scRNA-seq data</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">29</span><span class="op">)</span></span>
<span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sfe</span>, ncomponents <span class="op">=</span> <span class="fl">30</span>, scale <span class="op">=</span> <span class="cn">TRUE</span>, BSPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/BiocSingularParam.html" class="external-link">IrlbaParam</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ElbowPlot.html">ElbowPlot</a></span><span class="op">(</span><span class="va">sfe</span>, ndims <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-61-1.png" width="700" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotDimLoadings.html">plotDimLoadings</a></span><span class="op">(</span><span class="va">sfe</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-62-1.png" width="700" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/spatialReducedDim.html">spatialReducedDim</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"PCA"</span>, <span class="fl">6</span>, colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>, divergent <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                  diverge_center <span class="op">=</span> <span class="fl">0</span>, ncol <span class="op">=</span> <span class="fl">2</span>, scattermore <span class="op">=</span> <span class="cn">TRUE</span>, pointsize <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-63-1.png" width="864" style="display: block; margin: auto;"></p>
<p>While spatial region is not explicitly used, the PC’s highlight
spatial regions due to spatial autocorrelation in gene expression and
histological regions with different cell types.</p>
<p>Non-spatial clustering and locating the clusters in space</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/clusterRows.html" class="external-link">clusterRows</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"PCA"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">15</span><span class="op">]</span>,</span>
<span>                                    BLUSPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/NNGraphParam-class.html" class="external-link">SNNGraphParam</a></span><span class="op">(</span></span>
<span>                                        cluster.fun <span class="op">=</span> <span class="st">"leiden"</span>,</span>
<span>                                        cluster.args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>                                            resolution_parameter <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                                            objective_function <span class="op">=</span> <span class="st">"modularity"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in (function (to_check, X, clust_centers, clust_info, dtype, nn, :</span></span>
<span><span class="co">#&gt; detected tied distances to neighbors, see ?'BiocNeighbors-ties'</span></span></code></pre></div>
<p>Now the <code>scater</code> can also rasterize the plots with lots of
points with the <code>rasterise</code> argument, but with a different
mechanism from <code>scattermore</code> that requires more system
dependencies.</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span><span class="va">sfe</span>, ncomponents <span class="op">=</span> <span class="fl">4</span>, colour_by <span class="op">=</span> <span class="st">"cluster"</span>, rasterise <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-65-1.png" width="768" style="display: block; margin: auto;"></p>
<p>Plot the location of the clusters in space</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"cluster"</span>, colGeometryName <span class="op">=</span> <span class="st">"cellSeg"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-66-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="differential-expression">Differential expression<a class="anchor" aria-label="anchor" href="#differential-expression"></a>
</h2>
<p>Cluster marker genes are found with Wilcoxon rank sum test as
commonly done for scRNA-seq.</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/findMarkers.html" class="external-link">findMarkers</a></span><span class="op">(</span><span class="va">sfe</span>, groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span>,</span>
<span>                       test.type <span class="op">=</span> <span class="st">"wilcox"</span>, pval.type <span class="op">=</span> <span class="st">"all"</span>, direction <span class="op">=</span> <span class="st">"up"</span><span class="op">)</span></span></code></pre></div>
<p>It’s already sorted by p-values:</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">markers</span><span class="op">[[</span><span class="fl">6</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt; DataFrame with 541 rows and 15 columns</span></span>
<span><span class="co">#&gt;              p.value          FDR summary.AUC     AUC.1     AUC.2     AUC.3</span></span>
<span><span class="co">#&gt;            &lt;numeric&gt;    &lt;numeric&gt;   &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; CD93    1.42443e-195 7.70618e-193    0.863187  0.900493  0.779389  0.869111</span></span>
<span><span class="co">#&gt; AQP1    2.09173e-186 5.65813e-184    0.831633  0.891634  0.861863  0.817567</span></span>
<span><span class="co">#&gt; VWF     8.74605e-176 1.57720e-173    0.821921  0.866937  0.849105  0.844943</span></span>
<span><span class="co">#&gt; CLEC14A 4.97396e-162 6.72728e-160    0.830159  0.832982  0.821664  0.820733</span></span>
<span><span class="co">#&gt; RAMP2   1.00945e-153 1.09223e-151    0.821457  0.821743  0.813303  0.802825</span></span>
<span><span class="co">#&gt; ...              ...          ...         ...       ...       ...       ...</span></span>
<span><span class="co">#&gt; TPD52              1            1    0.167334  0.167334  0.478863  0.491853</span></span>
<span><span class="co">#&gt; TRAC               1            1    0.098802  0.569643  0.455366  0.451757</span></span>
<span><span class="co">#&gt; TRAF4              1            1    0.211878  0.211878  0.521652  0.489904</span></span>
<span><span class="co">#&gt; USP53              1            1    0.232210  0.232210  0.507661  0.436882</span></span>
<span><span class="co">#&gt; ZNF562             1            1    0.342753  0.342753  0.479892  0.534452</span></span>
<span><span class="co">#&gt;             AUC.4     AUC.5     AUC.7     AUC.8     AUC.9    AUC.10    AUC.11</span></span>
<span><span class="co">#&gt;         &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; CD93     0.867422  0.868915  0.902953  0.889624  0.876993  0.855931  0.871322</span></span>
<span><span class="co">#&gt; AQP1     0.860940  0.853170  0.890048  0.827669  0.887978  0.831633  0.866044</span></span>
<span><span class="co">#&gt; VWF      0.842964  0.853794  0.875926  0.873847  0.865796  0.821921  0.848069</span></span>
<span><span class="co">#&gt; CLEC14A  0.819943  0.824260  0.835086  0.832718  0.828647  0.816394  0.824287</span></span>
<span><span class="co">#&gt; RAMP2    0.809337  0.812654  0.825618  0.822806  0.823262  0.806178  0.813241</span></span>
<span><span class="co">#&gt; ...           ...       ...       ...       ...       ...       ...       ...</span></span>
<span><span class="co">#&gt; TPD52    0.443396 0.0883431 0.0962249  0.262814  0.413792  0.467724  0.314196</span></span>
<span><span class="co">#&gt; TRAC     0.098802 0.4370783 0.5757724  0.554619  0.565067  0.437688  0.350594</span></span>
<span><span class="co">#&gt; TRAF4    0.534264 0.5244890 0.2952734  0.438645  0.508543  0.507428  0.497875</span></span>
<span><span class="co">#&gt; USP53    0.526431 0.4934801 0.3118112  0.413622  0.512321  0.484612  0.546671</span></span>
<span><span class="co">#&gt; ZNF562   0.574561 0.5522572 0.3578757  0.450013  0.564940  0.531835  0.571869</span></span>
<span><span class="co">#&gt;            AUC.12    AUC.13</span></span>
<span><span class="co">#&gt;         &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; CD93     0.863187  0.848465</span></span>
<span><span class="co">#&gt; AQP1     0.882190  0.825609</span></span>
<span><span class="co">#&gt; VWF      0.853322  0.822751</span></span>
<span><span class="co">#&gt; CLEC14A  0.830159  0.812106</span></span>
<span><span class="co">#&gt; RAMP2    0.821457  0.780951</span></span>
<span><span class="co">#&gt; ...           ...       ...</span></span>
<span><span class="co">#&gt; TPD52    0.476575  0.538739</span></span>
<span><span class="co">#&gt; TRAC     0.351554  0.525100</span></span>
<span><span class="co">#&gt; TRAF4    0.374053  0.512868</span></span>
<span><span class="co">#&gt; USP53    0.538777  0.354077</span></span>
<span><span class="co">#&gt; ZNF562   0.593679  0.520011</span></span></code></pre></div>
<p>The code below extracts the significant markers for each cluster:</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes_use</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">vapply</a></span><span class="op">(</span><span class="va">markers</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, FUN.VALUE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotExpression.html" class="external-link">plotExpression</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">genes_use</span>, x <span class="op">=</span> <span class="st">"cluster"</span>, point_fun <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">...</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-69-1.png" width="700" style="display: block; margin: auto;"></p>
<p>This allows for plotting more top marker genes in a heatmap:</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes_use2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">markers</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotGroupedHeatmap.html" class="external-link">plotGroupedHeatmap</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">genes_use2</span>, group <span class="op">=</span> <span class="st">"cluster"</span>, colour <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/viridis_pal.html" class="external-link">viridis_pal</a></span><span class="op">(</span><span class="op">)</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-70-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="local-spatial-statistics-of-marker-genes">Local spatial statistics of marker genes<a class="anchor" aria-label="anchor" href="#local-spatial-statistics-of-marker-genes"></a>
</h2>
<p>First we plot those genes in space as a reference</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">genes_use</span>, colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>, ncol <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                   pointsize <span class="op">=</span> <span class="fl">0.3</span>, scattermore <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-71-1.png" width="864" style="display: block; margin: auto;"></p>
<p>Global Moran’s I of these marker genes is shown below:</p>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">[</span><span class="va">genes_use</span>, <span class="st">"moran_sample01"</span><span class="op">]</span>, <span class="va">genes_use</span><span class="op">)</span></span>
<span><span class="co">#&gt;     FOXA1      FGL2       LUM      CD3E    TENT5C      CD93     GATA3     KRT14 </span></span>
<span><span class="co">#&gt; 0.7421765 0.2604219 0.6812312 0.4015241 0.2806543 0.3250982 0.6558350 0.5464716 </span></span>
<span><span class="co">#&gt;     APOC1      CPA3     MS4A1    LILRA4     ADH1B </span></span>
<span><span class="co">#&gt; 0.2696177 0.1904912 0.2144728 0.1092981 0.6108921</span></span></code></pre></div>
<p>All these marker genes have positive spatial autocorrelation, but
some stronger than others.</p>
<p>Local Moran’s I of these marker genes is shown below:</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">runUnivariate</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"localmoran"</span>, features <span class="op">=</span> <span class="va">genes_use</span>, colGraphName <span class="op">=</span> <span class="st">"knn5"</span>,</span>
<span>                     BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"localmoran"</span>, features <span class="op">=</span> <span class="va">genes_use</span>, </span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>, ncol <span class="op">=</span> <span class="fl">3</span>, divergent <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                diverge_center <span class="op">=</span> <span class="fl">0</span>, scattermore <span class="op">=</span> <span class="cn">TRUE</span>, pointsize <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-74-1.png" width="864" style="display: block; margin: auto;"></p>
<p>It seems that some histological regions tend to be more spatially
homogenous in gene expression than others. The epithelial region tends
to be more homogenous. For some genes, regions with higher expression
also have higher local Moran’s I, such as FOXA1 and GATA3, while for
some genes, this is not the case, such as FGL2 and LUM.</p>
<p>Finally, we assess local spatial heteroscdasticity (LOSH) for these
marker genes to find local heterogeneity:</p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">runUnivariate</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"LOSH"</span>, features <span class="op">=</span> <span class="va">genes_use</span>, colGraphName <span class="op">=</span> <span class="st">"knn5"</span>,</span>
<span>                     BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"LOSH"</span>, features <span class="op">=</span> <span class="va">genes_use</span>, </span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>, ncol <span class="op">=</span> <span class="fl">3</span>, scattermore <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                pointsize <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig5_xenium_files/figure-html/unnamed-chunk-76-1.png" width="864" style="display: block; margin: auto;"></p>
<p>Again, just like in the <a href="https://pachterlab.github.io/voyager/articles/vig4_cosmx.html#local-spatial-statistics-of-marker-genes">CosMX
dataset</a>, LOSH is higher where the gene is more highly expressed in
some (e.g. CD3E, LUM, TENT5C) but not all cases (e.g. FOXA1, GATA3).
This may be due to spatial distribution of different cell types.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">#&gt; Platform: x86_64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS Monterey 12.6.7</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] patchwork_1.1.2                scran_1.28.2                  </span></span>
<span><span class="co">#&gt;  [3] bluster_1.10.0                 BiocSingular_1.16.0           </span></span>
<span><span class="co">#&gt;  [5] BiocParallel_1.34.2            scater_1.28.0                 </span></span>
<span><span class="co">#&gt;  [7] scuttle_1.10.1                 stringr_1.5.0                 </span></span>
<span><span class="co">#&gt;  [9] ggplot2_3.4.2                  SpatialFeatureExperiment_1.2.2</span></span>
<span><span class="co">#&gt; [11] SpatialExperiment_1.10.0       SingleCellExperiment_1.22.0   </span></span>
<span><span class="co">#&gt; [13] SummarizedExperiment_1.30.2    Biobase_2.60.0                </span></span>
<span><span class="co">#&gt; [15] GenomicRanges_1.52.0           GenomeInfoDb_1.36.1           </span></span>
<span><span class="co">#&gt; [17] IRanges_2.34.1                 S4Vectors_0.38.1              </span></span>
<span><span class="co">#&gt; [19] BiocGenerics_0.46.0            MatrixGenerics_1.12.2         </span></span>
<span><span class="co">#&gt; [21] matrixStats_1.0.0              SFEData_1.2.0                 </span></span>
<span><span class="co">#&gt; [23] Voyager_1.2.4                 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] splines_4.3.1                 later_1.3.1                  </span></span>
<span><span class="co">#&gt;   [3] bitops_1.0-7                  filelock_1.0.2               </span></span>
<span><span class="co">#&gt;   [5] tibble_3.2.1                  R.oo_1.25.0                  </span></span>
<span><span class="co">#&gt;   [7] lifecycle_1.0.3               sf_1.0-14                    </span></span>
<span><span class="co">#&gt;   [9] edgeR_3.42.4                  rprojroot_2.0.3              </span></span>
<span><span class="co">#&gt;  [11] lattice_0.21-8                magrittr_2.0.3               </span></span>
<span><span class="co">#&gt;  [13] limma_3.56.2                  sass_0.4.7                   </span></span>
<span><span class="co">#&gt;  [15] rmarkdown_2.23                jquerylib_0.1.4              </span></span>
<span><span class="co">#&gt;  [17] yaml_2.3.7                    metapod_1.8.0                </span></span>
<span><span class="co">#&gt;  [19] httpuv_1.6.11                 sp_2.0-0                     </span></span>
<span><span class="co">#&gt;  [21] cowplot_1.1.1                 RColorBrewer_1.1-3           </span></span>
<span><span class="co">#&gt;  [23] DBI_1.1.3                     abind_1.4-5                  </span></span>
<span><span class="co">#&gt;  [25] zlibbioc_1.46.0               purrr_1.0.1                  </span></span>
<span><span class="co">#&gt;  [27] R.utils_2.12.2                RCurl_1.98-1.12              </span></span>
<span><span class="co">#&gt;  [29] rappdirs_0.3.3                GenomeInfoDbData_1.2.10      </span></span>
<span><span class="co">#&gt;  [31] ggrepel_0.9.3                 irlba_2.3.5.1                </span></span>
<span><span class="co">#&gt;  [33] terra_1.7-39                  pheatmap_1.0.12              </span></span>
<span><span class="co">#&gt;  [35] units_0.8-2                   RSpectra_0.16-1              </span></span>
<span><span class="co">#&gt;  [37] dqrng_0.3.0                   pkgdown_2.0.7                </span></span>
<span><span class="co">#&gt;  [39] DelayedMatrixStats_1.22.1     codetools_0.2-19             </span></span>
<span><span class="co">#&gt;  [41] DropletUtils_1.20.0           DelayedArray_0.26.7          </span></span>
<span><span class="co">#&gt;  [43] tidyselect_1.2.0              farver_2.1.1                 </span></span>
<span><span class="co">#&gt;  [45] ScaledMatrix_1.8.1            viridis_0.6.4                </span></span>
<span><span class="co">#&gt;  [47] BiocFileCache_2.8.0           jsonlite_1.8.7               </span></span>
<span><span class="co">#&gt;  [49] BiocNeighbors_1.18.0          e1071_1.7-13                 </span></span>
<span><span class="co">#&gt;  [51] ellipsis_0.3.2                systemfonts_1.0.4            </span></span>
<span><span class="co">#&gt;  [53] tools_4.3.1                   ggnewscale_0.4.9             </span></span>
<span><span class="co">#&gt;  [55] ragg_1.2.5                    Rcpp_1.0.11                  </span></span>
<span><span class="co">#&gt;  [57] glue_1.6.2                    gridExtra_2.3                </span></span>
<span><span class="co">#&gt;  [59] mgcv_1.9-0                    xfun_0.39                    </span></span>
<span><span class="co">#&gt;  [61] dplyr_1.1.2                   HDF5Array_1.28.1             </span></span>
<span><span class="co">#&gt;  [63] withr_2.5.0                   BiocManager_1.30.21.1        </span></span>
<span><span class="co">#&gt;  [65] fastmap_1.1.1                 boot_1.3-28.1                </span></span>
<span><span class="co">#&gt;  [67] rhdf5filters_1.12.1           fansi_1.0.4                  </span></span>
<span><span class="co">#&gt;  [69] spData_2.3.0                  digest_0.6.33                </span></span>
<span><span class="co">#&gt;  [71] rsvd_1.0.5                    R6_2.5.1                     </span></span>
<span><span class="co">#&gt;  [73] mime_0.12                     textshaping_0.3.6            </span></span>
<span><span class="co">#&gt;  [75] colorspace_2.1-0              wk_0.7.3                     </span></span>
<span><span class="co">#&gt;  [77] scattermore_1.2               RSQLite_2.3.1                </span></span>
<span><span class="co">#&gt;  [79] R.methodsS3_1.8.2             hexbin_1.28.3                </span></span>
<span><span class="co">#&gt;  [81] utf8_1.2.3                    generics_0.1.3               </span></span>
<span><span class="co">#&gt;  [83] class_7.3-22                  httr_1.4.6                   </span></span>
<span><span class="co">#&gt;  [85] S4Arrays_1.0.5                spdep_1.2-8                  </span></span>
<span><span class="co">#&gt;  [87] pkgconfig_2.0.3               scico_1.4.0                  </span></span>
<span><span class="co">#&gt;  [89] gtable_0.3.3                  blob_1.2.4                   </span></span>
<span><span class="co">#&gt;  [91] XVector_0.40.0                htmltools_0.5.5              </span></span>
<span><span class="co">#&gt;  [93] scales_1.2.1                  png_0.1-8                    </span></span>
<span><span class="co">#&gt;  [95] knitr_1.43                    rjson_0.2.21                 </span></span>
<span><span class="co">#&gt;  [97] nlme_3.1-162                  curl_5.0.1                   </span></span>
<span><span class="co">#&gt;  [99] proxy_0.4-27                  cachem_1.0.8                 </span></span>
<span><span class="co">#&gt; [101] rhdf5_2.44.0                  BiocVersion_3.17.1           </span></span>
<span><span class="co">#&gt; [103] KernSmooth_2.23-22            parallel_4.3.1               </span></span>
<span><span class="co">#&gt; [105] vipor_0.4.5                   AnnotationDbi_1.62.2         </span></span>
<span><span class="co">#&gt; [107] desc_1.4.2                    s2_1.1.4                     </span></span>
<span><span class="co">#&gt; [109] pillar_1.9.0                  grid_4.3.1                   </span></span>
<span><span class="co">#&gt; [111] vctrs_0.6.3                   promises_1.2.0.1             </span></span>
<span><span class="co">#&gt; [113] dbplyr_2.3.3                  beachmat_2.16.0              </span></span>
<span><span class="co">#&gt; [115] xtable_1.8-4                  cluster_2.1.4                </span></span>
<span><span class="co">#&gt; [117] beeswarm_0.4.0                evaluate_0.21                </span></span>
<span><span class="co">#&gt; [119] magick_2.7.4                  cli_3.6.1                    </span></span>
<span><span class="co">#&gt; [121] locfit_1.5-9.8                compiler_4.3.1               </span></span>
<span><span class="co">#&gt; [123] rlang_1.1.1                   crayon_1.5.2                 </span></span>
<span><span class="co">#&gt; [125] labeling_0.4.2                classInt_0.4-9               </span></span>
<span><span class="co">#&gt; [127] fs_1.6.3                      ggbeeswarm_0.7.2             </span></span>
<span><span class="co">#&gt; [129] stringi_1.7.12                viridisLite_0.4.2            </span></span>
<span><span class="co">#&gt; [131] deldir_1.0-9                  munsell_0.5.0                </span></span>
<span><span class="co">#&gt; [133] Biostrings_2.68.1             Matrix_1.6-0                 </span></span>
<span><span class="co">#&gt; [135] ExperimentHub_2.8.1           sparseMatrixStats_1.12.2     </span></span>
<span><span class="co">#&gt; [137] bit64_4.0.5                   Rhdf5lib_1.22.0              </span></span>
<span><span class="co">#&gt; [139] statmod_1.5.0                 KEGGREST_1.40.0              </span></span>
<span><span class="co">#&gt; [141] shiny_1.7.4.1                 highr_0.10                   </span></span>
<span><span class="co">#&gt; [143] interactiveDisplayBase_1.38.0 AnnotationHub_3.8.0          </span></span>
<span><span class="co">#&gt; [145] igraph_1.5.0.1                memoise_2.0.1                </span></span>
<span><span class="co">#&gt; [147] bslib_0.5.0                   bit_4.0.5</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Janesick2022-rp" class="csl-entry">
Janesick, Amanda, Robert Shelansky, Andrew Gottscho, Florian Wagner,
Morgane Rouault, Ghezal Beliakoff, Michelli Faria de Oliveira, et al.
2022. <span>“High Resolution Mapping of the Breast Cancer Tumor
Microenvironment Using Integrated Single Cell, Spatial and in Situ
Analysis of <span>FFPE</span> Tissue.”</span> <em>bioRxiv</em>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Lambda Moses, Kayla Jackson, Laura Luebbert, Sina Booeshaghi, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
