<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Voyager">
<title>Spatial analysis of 10X example Visium dataset • Voyager</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Spatial analysis of 10X example Visium dataset">
<meta property="og:description" content="Voyager">
<meta name="robots" content="noindex">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">Voyager</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">1.0.10</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/install.html">Install</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-technologies">Technologies</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-technologies">
    <a class="dropdown-item" href="../articles/visium_landing.html">Visium</a>
    <a class="dropdown-item" href="../articles/slideseqV2_landing.html">Slide-seq v2</a>
    <a class="dropdown-item" href="../articles/cosmx_landing.html">CosMX</a>
    <a class="dropdown-item" href="../articles/xenium_landing.html">Xenium</a>
    <a class="dropdown-item" href="../articles/merfish_landing.html">MERFISH</a>
    <a class="dropdown-item" href="../articles/seqfish_landing.html">seqFISH</a>
    <a class="dropdown-item" href="../articles/chromium_landing.html">10X Chromium V3</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Spatial analysis of 10X example Visium dataset</h1>
                        <h4 data-toc-skip class="author">Lambda
Moses</h4>
            
            <h4 data-toc-skip class="date">2023-03-04</h4>
      
      
      <div class="d-none name"><code>visium_10x_spatial.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In a <a href="https://pachterlab.github.io/voyager/articles/visium_10x.html">more
introductory vignette</a>, we performed basic non-spatial analyses on a
mouse olfactory bulb Visium dataset from the 10X website. In this
vignette, we perform spatial analyses in histological space as well as
in gene expression space.</p>
<p>Here we load the packages used in this vignette:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">Voyager</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SpatialFeatureExperiment" class="external-link">SpatialFeatureExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: SummarizedExperiment</span></span>
<span><span class="co">#&gt; Loading required package: MatrixGenerics</span></span>
<span><span class="co">#&gt; Loading required package: matrixStats</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'MatrixGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span></span>
<span><span class="co">#&gt;     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span></span>
<span><span class="co">#&gt;     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span></span>
<span><span class="co">#&gt;     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span></span>
<span><span class="co">#&gt;     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span></span>
<span><span class="co">#&gt;     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span></span>
<span><span class="co">#&gt;     colWeightedMeans, colWeightedMedians, colWeightedSds,</span></span>
<span><span class="co">#&gt;     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span></span>
<span><span class="co">#&gt;     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span></span>
<span><span class="co">#&gt;     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span></span>
<span><span class="co">#&gt;     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span></span>
<span><span class="co">#&gt;     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span></span>
<span><span class="co">#&gt;     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span></span>
<span><span class="co">#&gt;     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span></span>
<span><span class="co">#&gt;     rowWeightedSds, rowWeightedVars</span></span>
<span><span class="co">#&gt; Loading required package: GenomicRanges</span></span>
<span><span class="co">#&gt; Loading required package: stats4</span></span>
<span><span class="co">#&gt; Loading required package: BiocGenerics</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'BiocGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">#&gt;     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">#&gt;     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,</span></span>
<span><span class="co">#&gt;     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span></span>
<span><span class="co">#&gt;     Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,</span></span>
<span><span class="co">#&gt;     table, tapply, union, unique, unsplit, which.max, which.min</span></span>
<span><span class="co">#&gt; Loading required package: S4Vectors</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'S4Vectors'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     expand.grid, I, unname</span></span>
<span><span class="co">#&gt; Loading required package: IRanges</span></span>
<span><span class="co">#&gt; Loading required package: GenomeInfoDb</span></span>
<span><span class="co">#&gt; Loading required package: Biobase</span></span>
<span><span class="co">#&gt; Welcome to Bioconductor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     Vignettes contain introductory material; view with</span></span>
<span><span class="co">#&gt;     'browseVignettes()'. To cite Bioconductor, see</span></span>
<span><span class="co">#&gt;     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'Biobase'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:MatrixGenerics':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     rowMedians</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyMissing, rowMedians</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: scuttle</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scuttle</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org" class="external-link">stringr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/alexcb/rjson" class="external-link">rjson</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/aoles/EBImage" class="external-link">EBImage</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'EBImage'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:SummarizedExperiment':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     resize</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:Biobase':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     channel</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:GenomicRanges':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     resize, tile</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:IRanges':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     resize, tile</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rspatial.org/terra/" class="external-link">terra</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; terra 1.7.3</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'terra'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:EBImage':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     flip, rotate</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:patchwork':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     area</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:SummarizedExperiment':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     distance, nearest, shift, trim, values, values&lt;-, width</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:GenomicRanges':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     distance, gaps, nearest, shift, trim, values, values&lt;-, width</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:IRanges':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     distance, gaps, nearest, shift, trim, width</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:S4Vectors':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     values, values&lt;-, width</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:BiocGenerics':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     width</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:SpatialFeatureExperiment':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     centroids, crop</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rlang.r-lib.org" class="external-link">rlang</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'rlang'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:Biobase':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     exprs</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-spatial.github.io/sf/" class="external-link">sf</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Linking to GEOS 3.10.2, GDAL 3.4.2, PROJ 8.2.1; sf_use_s2() is TRUE</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ateucher/rmapshaper" class="external-link">rmapshaper</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:terra':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, union</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:EBImage':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     combine</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:Biobase':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     combine</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:GenomicRanges':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, union</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:GenomeInfoDb':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:IRanges':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     collapse, desc, intersect, setdiff, slice, union</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:S4Vectors':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     first, intersect, rename, setdiff, setequal, union</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:BiocGenerics':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     combine, intersect, setdiff, union</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     count</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BiocNeighbors</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Here we download the data from the 10X website. This is the
unfiltered gene count matrix:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="st">"visium_ob.tar.gz"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://cf.10xgenomics.com/samples/spatial-exp/2.0.0/Visium_Mouse_Olfactory_Bulb/Visium_Mouse_Olfactory_Bulb_raw_feature_bc_matrix.tar.gz"</span>, </span>
<span>                  destfile <span class="op">=</span> <span class="st">"visium_ob.tar.gz"</span><span class="op">)</span></span></code></pre></div>
<p>This is the spatial information:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="st">"visium_ob_spatial.tar.gz"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://cf.10xgenomics.com/samples/spatial-exp/2.0.0/Visium_Mouse_Olfactory_Bulb/Visium_Mouse_Olfactory_Bulb_spatial.tar.gz"</span>, </span>
<span>                  destfile <span class="op">=</span> <span class="st">"visium_ob_spatial.tar.gz"</span><span class="op">)</span></span></code></pre></div>
<p>Decompress the downloaded content:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.exists</a></span><span class="op">(</span><span class="st">"outs"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="st">"outs"</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"tar -xvf visium_ob.tar.gz -C outs"</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"tar -xvf visium_ob_spatial.tar.gz -C outs"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Contents of the <code>outs</code> directory as from Space Ranger is
explained in <a href="https://pachterlab.github.io/voyager/articles/visium_10x.html">the
introductory vignette</a>.</p>
<p>Here we read the data into R as an SFE object.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/read10xVisiumSFE.html" class="external-link">read10xVisiumSFE</a></span><span class="op">(</span>samples <span class="op">=</span> <span class="st">"outs"</span>, type <span class="op">=</span> <span class="st">"sparse"</span>, data <span class="op">=</span> <span class="st">"raw"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: as(&lt;dgTMatrix&gt;, "dgCMatrix") is deprecated since Matrix 1.5-0; do</span></span>
<span><span class="co">#&gt; as(., "CsparseMatrix") instead</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 32285 4992 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(32285): ENSMUSG00000051951 ENSMUSG00000089699 ...</span></span>
<span><span class="co">#&gt;   ENSMUSG00000095019 ENSMUSG00000095041</span></span>
<span><span class="co">#&gt; rowData names(1): symbol</span></span>
<span><span class="co">#&gt; colnames(4992): AAACAACGAATAGTTC-1 AAACAAGTATCTCCCA-1 ...</span></span>
<span><span class="co">#&gt;   TTGTTTGTATTACACG-1 TTGTTTGTGTAAATTC-1</span></span>
<span><span class="co">#&gt; colData names(4): in_tissue array_row array_col sample_id</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : pxl_row_in_fullres pxl_col_in_fullres</span></span>
<span><span class="co">#&gt; imgData names(4): sample_id image_id data scaleFactor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: spotPoly (POLYGON) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01:</span></span></code></pre></div>
<p>Here we add QC metrics, already plotted in the introductory
vignette.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">is_mt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">symbol</span>, <span class="st">"^mt-"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/addPerCellQCMetrics.html" class="external-link">addPerCellQCMetrics</a></span><span class="op">(</span><span class="va">sfe</span>, subsets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>mito <span class="op">=</span> <span class="va">is_mt</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="tissue-segmentation">Tissue segmentation<a class="anchor" aria-label="anchor" href="#tissue-segmentation"></a>
</h2>
<p>While Space Ranger can automatically detect which spots are in tissue
and the Loupe browser can be used to manually annotate which spots are
in tissue, it may be interesting to get the tissue outline polygon, so
we would know how much each spot overlaps with the tissue and plot the
outline. The tissue boundary polygon can be manually annotated with
QuPath, which saves the polygon as a GeoJSON and can be directly read
into R with <code><a href="https://r-spatial.github.io/sf/reference/st_read.html" class="external-link">st_read()</a></code>.</p>
<p>Or we can segment the tissue computationally. R generally isn’t great
for image processing, but there are some packages that can perform the
segmentation, such as <a href="https://bioconductor.org/packages/release/bioc/html/EBImage.html" class="external-link"><code>EBImage</code></a>,
which is based on its own in house C and C++ code, and <a href="https://github.com/dahtah/imager/" class="external-link"><code>imager</code></a>, which
is based on <a href="http://cimg.eu/" class="external-link"><code>CImg</code></a>.</p>
<p>Here we don’t have the full resolution image. We will perform tissue
segmentation on the high resolution downsampled image and then scale it
to make the coordinates of the tissue boundary match those of the spots.
The <code>EBImage</code> package is used here. Compared to OpenCV,
<code>EBImage</code> is slow on the full resolution image, but should be
fine here for the downsized image.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">img</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/io.html" class="external-link">readImage</a></span><span class="op">(</span><span class="st">"outs/spatial/tissue_hires_image.png"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/display.html" class="external-link">display</a></span><span class="op">(</span><span class="va">img</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>When rendered as a static webpage, the image is static, but when run
interactively, this image will be shown in an interactive widget where
you can zoom and pan.</p>
<p>Here we show the RGB channels separately</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">img2</span> <span class="op">&lt;-</span> <span class="va">img</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/Image.html" class="external-link">colorMode</a></span><span class="op">(</span><span class="va">img2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">Grayscale</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/display.html" class="external-link">display</a></span><span class="op">(</span><span class="va">img2</span>, all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">img</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<p>The tissue can be discerned with thresholding. The tall peak on the
right is the background. The much lower peaks from around 0.6 to 0.85
must be the tissue. To capture the faint bluish region, the blue channel
is used for thresholding. The threshold here is chosen based on the
histogram and experimenting with nearby values.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mask</span> <span class="op">&lt;-</span> <span class="va">img2</span><span class="op">[</span>,,<span class="fl">3</span><span class="op">]</span> <span class="op">&lt;</span> <span class="fl">0.87</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/display.html" class="external-link">display</a></span><span class="op">(</span><span class="va">mask</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<p>Then we use an opening operation (erosion followed by dilation) to
denoise</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kern</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/morphology.html" class="external-link">makeBrush</a></span><span class="op">(</span><span class="fl">3</span>, shape<span class="op">=</span><span class="st">'disc'</span><span class="op">)</span></span>
<span><span class="va">mask_open</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/morphology.html" class="external-link">opening</a></span><span class="op">(</span><span class="va">mask</span>, <span class="va">kern</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/display.html" class="external-link">display</a></span><span class="op">(</span><span class="va">mask_open</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<p>There are some small holes in the tissue, which can be removed by a
closing operation (dilation followed by erosion):</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mask_close</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/morphology.html" class="external-link">closing</a></span><span class="op">(</span><span class="va">mask_open</span>, <span class="va">kern</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/display.html" class="external-link">display</a></span><span class="op">(</span><span class="va">mask_close</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<p>There are some larger holes in the tissue mask, which may be real
holes or faint regions with few nuclei missed by thresholding. They
might not be large enough to affect which Visium spots intersect the
tissue.</p>
<p>Now the main piece of tissue is clear. It must be the object with the
largest area. However, there are two small pieces that should belong to
the tissue at the top left. The debris and fiducials can be removed by
setting all pixels in the mask outside the bounding box of the main
piece to 0. Here we assign a different value to each contiguous object
with <code><a href="https://rdrr.io/pkg/EBImage/man/bwlabel.html" class="external-link">bwlabel()</a></code>, and use
<code><a href="https://rdrr.io/pkg/EBImage/man/computeFeatures.html" class="external-link">computeFeatures.shape()</a></code> to find the area among other shape
features (e.g. perimeter) of each object.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mask_label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/bwlabel.html" class="external-link">bwlabel</a></span><span class="op">(</span><span class="va">mask_close</span><span class="op">)</span></span>
<span><span class="va">fts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/computeFeatures.html" class="external-link">computeFeatures.shape</a></span><span class="op">(</span><span class="va">mask_label</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">fts</span><span class="op">)</span></span>
<span><span class="co">#&gt;   s.area s.perimeter s.radius.mean s.radius.sd s.radius.min s.radius.max</span></span>
<span><span class="co">#&gt; 1     39          25      3.428773   1.3542219    1.4176036     5.762777</span></span>
<span><span class="co">#&gt; 2     20          14      2.032665   0.3439068    1.5000000     2.500000</span></span>
<span><span class="co">#&gt; 3      8           8      1.144123   0.4370160    0.7071068     1.581139</span></span>
<span><span class="co">#&gt; 4     14          10      1.689175   0.2160726    1.5811388     2.121320</span></span>
<span><span class="co">#&gt; 5     15          12      1.716761   0.4684015    1.0000000     2.236068</span></span>
<span><span class="co">#&gt; 6      9           8      1.207107   0.2071068    1.0000000     1.414214</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">fts</span><span class="op">[</span>,<span class="st">"s.area"</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;     Min.  1st Qu.   Median     Mean  3rd Qu.     Max. </span></span>
<span><span class="co">#&gt;      8.0     14.0     51.0    595.9    345.0 496326.0</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">max_ind</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/terra/man/summarize-generics.html" class="external-link">which.max</a></span><span class="op">(</span><span class="va">fts</span><span class="op">[</span>,<span class="st">"s.area"</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">inds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/coerce.html" class="external-link">as.array</a></span><span class="op">(</span><span class="va">mask_label</span><span class="op">)</span> <span class="op">==</span> <span class="va">max_ind</span>, arr.ind <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">inds</span><span class="op">)</span></span>
<span><span class="co">#&gt;       row col</span></span>
<span><span class="co">#&gt; [1,] 1168 562</span></span>
<span><span class="co">#&gt; [2,] 1169 562</span></span>
<span><span class="co">#&gt; [3,] 1170 562</span></span>
<span><span class="co">#&gt; [4,] 1158 563</span></span>
<span><span class="co">#&gt; [5,] 1159 563</span></span>
<span><span class="co">#&gt; [6,] 1160 563</span></span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">row_inds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">inds</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">inds</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">+</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/pkg/terra/man/dimensions.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">mask_label</span><span class="op">)</span>, by <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">col_inds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">inds</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">inds</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">+</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/pkg/terra/man/dimensions.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">mask_label</span><span class="op">)</span>, by <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">mask_label</span><span class="op">[</span><span class="va">row_inds</span>, <span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span><span class="va">mask_label</span><span class="op">[</span>,<span class="va">col_inds</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/display.html" class="external-link">display</a></span><span class="op">(</span><span class="va">mask_label</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<p>Then remove the small pieces that are debris.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="va">mask_label</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1]   0 421 425 429 430 438 450 458 461 469 473 483 487 505 523 633 640 642 651</span></span>
<span><span class="co">#&gt; [20] 678 739 741 757 762 775 778 789 791 797 805 810 813 820 821 822 826 831 838</span></span>
<span><span class="co">#&gt; [39] 839 840 843 845 848 849 861 862 863</span></span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fts2</span> <span class="op">&lt;-</span> <span class="va">fts</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="va">mask_label</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span>,<span class="op">]</span></span>
<span><span class="va">fts2</span> <span class="op">&lt;-</span> <span class="va">fts2</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">fts2</span><span class="op">[</span>,<span class="st">"s.area"</span><span class="op">]</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,<span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">fts2</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span>, type <span class="op">=</span> <span class="st">"l"</span>, ylab <span class="op">=</span> <span class="st">"Area"</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">fts2</span>, <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt;     s.area s.perimeter s.radius.mean s.radius.sd s.radius.min s.radius.max</span></span>
<span><span class="co">#&gt; 421 496326        3151    395.118732  68.6493949  234.1605637   485.715835</span></span>
<span><span class="co">#&gt; 450    217          55      7.840627   1.2883202    5.0010247    10.458197</span></span>
<span><span class="co">#&gt; 849    211          63      7.961248   2.0228566    3.8569142    12.189753</span></span>
<span><span class="co">#&gt; 797    182          56      7.547362   2.3368359    3.0772370    11.839919</span></span>
<span><span class="co">#&gt; 461    136          54      7.186020   3.2751628    0.9255555    12.479805</span></span>
<span><span class="co">#&gt; 741     92          56      6.365661   2.8382829    1.3273276    11.653219</span></span>
<span><span class="co">#&gt; 840     69          33      4.503526   1.6417370    1.6026264     7.076974</span></span>
<span><span class="co">#&gt; 862     63          37      4.854424   2.4445530    0.6361407     8.837838</span></span>
<span><span class="co">#&gt; 839     45          25      3.305562   0.7074306    1.9320455     4.526897</span></span>
<span><span class="co">#&gt; 775     32          22      2.887407   1.1755375    0.5300865     4.543636</span></span></code></pre></div>
<p>Object number 797 is a piece of debris at the bottom left. The other
pieces with area over 100 pixels are tissue. Since debris really is
small bits of tissue, so the boundary between debris and tissue can be
blurry. Here the two are distinguished by morphology on the H&amp;E
image and proximity to the main tissue.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#display(mask_label == 797)</span></span></code></pre></div>
<p>Here we remove the debris from the mask</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mask_label</span><span class="op">[</span><span class="va">mask_label</span> <span class="op"><a href="https://rdrr.io/pkg/terra/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">797</span>, <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">fts2</span><span class="op">)</span><span class="op">[</span><span class="va">fts2</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span> <span class="op">&lt;</span> <span class="fl">100</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span></code></pre></div>
<p>Since most holes in the mask are faint regions of the tissue missed
by thresholding, the holes will be filled</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mask_label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/fillHull.html" class="external-link">fillHull</a></span><span class="op">(</span><span class="va">mask_label</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/display.html" class="external-link">display</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/paintObjects.html" class="external-link">paintObjects</a></span><span class="op">(</span><span class="va">mask_label</span>, <span class="va">img</span>, col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"red"</span>, <span class="st">"yellow"</span><span class="op">)</span>, opac<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0.3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-27-1.png" width="700"></p>
<p>This segmentation process took a lot of manual oversight, in choosing
the threshold, choosing kernel size and shape in the opening and closing
operations, deciding whether to fill the holes, and deciding what is
debris and what is tissue.</p>
<p>Finally, the image needs to be rotated to match the spots.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mask_label</span> <span class="op">&lt;-</span> <span class="fu">EBImage</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/spatial.html" class="external-link">rotate</a></span><span class="op">(</span><span class="va">mask_label</span>, <span class="fl">270</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/display.html" class="external-link">display</a></span><span class="op">(</span><span class="va">mask_label</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-28-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="convert-tissue-mask-to-polygon">Convert tissue mask to polygon<a class="anchor" aria-label="anchor" href="#convert-tissue-mask-to-polygon"></a>
</h2>
<p>Now we have the tissue mask, which we will convert to polygon. While
OpenCV can directly perform the conversion, as there isn’t a
comprehensive R wrapper of OpenCV, this conversion is more convoluted in
R. We first convert the <code>Image</code> object to a raster as
implemented in <code>terra</code>, the core R package for geospatial
raster data. Then <code>terra</code> can convert the raster to polygon.
As this image is downsized, the polygon will look quite pixelated. To
mitigate this pixelation and save memory, the <code><a href="https://rdrr.io/pkg/rmapshaper/man/ms_simplify.html" class="external-link">ms_simplify()</a></code>
function is used to simplify the polygon, only keeping a small
proportion of all vertices. The <code><a href="https://r-spatial.github.io/sf/reference/geos_unary.html" class="external-link">st_simplify()</a></code> function in
<code>sf</code> can also simplify the polygons, but we can’t specify
what proportion of vertices to keep.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">raster2polygon</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">seg</span>, <span class="va">keep</span> <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">r</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/terra/man/rast.html" class="external-link">rast</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/transpose.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/coerce.html" class="external-link">as.array</a></span><span class="op">(</span><span class="va">seg</span><span class="op">)</span><span class="op">)</span>, extent <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/terra/man/ext.html" class="external-link">ext</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/pkg/terra/man/dimensions.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">seg</span><span class="op">)</span>, <span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/pkg/terra/man/dimensions.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">seg</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">r</span><span class="op">[</span><span class="va">r</span> <span class="op">&lt;</span> <span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span>    <span class="va">contours</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_as_sf.html" class="external-link">st_as_sf</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/as.spatvector.html" class="external-link">as.polygons</a></span><span class="op">(</span><span class="va">r</span>, dissolve <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">simplified</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rmapshaper/man/ms_simplify.html" class="external-link">ms_simplify</a></span><span class="op">(</span><span class="va">contours</span>, keep <span class="op">=</span> <span class="va">keep</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>full <span class="op">=</span> <span class="va">contours</span>,</span>
<span>         simplified <span class="op">=</span> <span class="va">simplified</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tb</span> <span class="op">&lt;-</span> <span class="fu">raster2polygon</span><span class="op">(</span><span class="va">mask_label</span><span class="op">)</span></span></code></pre></div>
<p>Before adding the geometry to the SFE object, it needs to be scaled
to match the coordinates of the spots</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scale_factors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rjson/man/fromJSON.html" class="external-link">fromJSON</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"outs/spatial/scalefactors_json.json"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tb</span><span class="op">$</span><span class="va">simplified</span><span class="op">$</span><span class="va">geometry</span> <span class="op">&lt;-</span> <span class="va">tb</span><span class="op">$</span><span class="va">simplified</span><span class="op">$</span><span class="va">geometry</span> <span class="op">/</span> <span class="va">scale_factors</span><span class="op">$</span><span class="va">tissue_hires_scalef</span></span></code></pre></div>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotGeometries.html" class="external-link">tissueBoundary</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">tb</span><span class="op">$</span><span class="va">simplified</span></span></code></pre></div>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"sum"</span>, annotGeometryName <span class="op">=</span> <span class="st">"tissueBoundary"</span>, </span>
<span>                   annot_fixed <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="cn">NA</span>, size <span class="op">=</span> <span class="fl">0.5</span>, color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-34-1.png" width="700"></p>
<p>Then we can use geometric operations to find which spots intersect
tissue, which spots are covered by tissue, and how much of each spot
intersects tissue.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Which spots intersect tissue</span></span>
<span><span class="va">sfe</span><span class="op">$</span><span class="va">int_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotPred.html" class="external-link">annotPred</a></span><span class="op">(</span><span class="va">sfe</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, </span>
<span>                            annotGeometryName <span class="op">=</span> <span class="st">"tissueBoundary"</span>,</span>
<span>                            pred <span class="op">=</span> <span class="va">st_intersects</span><span class="op">)</span></span>
<span><span class="va">sfe</span><span class="op">$</span><span class="va">cov_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotPred.html" class="external-link">annotPred</a></span><span class="op">(</span><span class="va">sfe</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, </span>
<span>                            annotGeometryName <span class="op">=</span> <span class="st">"tissueBoundary"</span>,</span>
<span>                            pred <span class="op">=</span> <span class="va">st_covered_by</span><span class="op">)</span></span></code></pre></div>
<p>Discrepancies between Space Ranger’s annotation and the annotation
based on tissue segmentation here:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span><span class="op">$</span><span class="va">diff_sr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html" class="external-link">case_when</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">$</span><span class="va">in_tissue</span> <span class="op">==</span> <span class="va">sfe</span><span class="op">$</span><span class="va">int_tissue</span> <span class="op">~</span> <span class="st">"same"</span>,</span>
<span>                         <span class="va">sfe</span><span class="op">$</span><span class="va">in_tissue</span> <span class="op">&amp;</span> <span class="op">!</span><span class="va">sfe</span><span class="op">$</span><span class="va">int_tissue</span> <span class="op">~</span> <span class="st">"Space Ranger"</span>,</span>
<span>                         <span class="va">sfe</span><span class="op">$</span><span class="va">int_tissue</span> <span class="op">&amp;</span> <span class="op">!</span><span class="va">sfe</span><span class="op">$</span><span class="va">in_tissue</span> <span class="op">~</span> <span class="st">"segmentation"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span>levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Space Ranger"</span>, <span class="st">"same"</span>, <span class="st">"segmentation"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"diff_sr"</span>, </span>
<span>                   annotGeometryName <span class="op">=</span> <span class="st">"tissueBoundary"</span>, </span>
<span>                   annot_fixed <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="cn">NA</span>, size <span class="op">=</span> <span class="fl">0.5</span>, color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_brewer</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"div"</span>, palette <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">fill</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">fill</span>, which will replace the existing scale.</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-36-1.png" width="700"></p>
<p>Spots at the margin can intersect the tissue without being covered by
it.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span><span class="op">$</span><span class="va">diff_int_cov</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">$</span><span class="va">int_tissue</span> <span class="op">!=</span> <span class="va">sfe</span><span class="op">$</span><span class="va">cov_tissue</span></span>
<span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"diff_int_cov"</span>, </span>
<span>                   annotGeometryName <span class="op">=</span> <span class="st">"tissueBoundary"</span>, </span>
<span>                   annot_fixed <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="cn">NA</span>, size <span class="op">=</span> <span class="fl">0.5</span>, color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-37-1.png" width="700"></p>
<p>We can also get the geometries of the intersections between the
tissue and the Visium spots, and then calculate what percentage of each
spot is in tissue. However, this percentage may not be very useful if
the tissue segmentation is subject to error. This percentage may be more
useful for pathologist annotated histological regions or objects such as
nuclei and myofibers.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spot_ints</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotOp.html" class="external-link">annotOp</a></span><span class="op">(</span><span class="va">sfe</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, </span>
<span>                     annotGeometryName <span class="op">=</span> <span class="st">"tissueBoundary"</span>, op <span class="op">=</span> <span class="va">st_intersection</span><span class="op">)</span></span>
<span><span class="va">sfe</span><span class="op">$</span><span class="va">pct_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://r-spatial.github.io/sf/reference/geos_measures.html" class="external-link">st_area</a></span><span class="op">(</span><span class="va">spot_ints</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://r-spatial.github.io/sf/reference/geos_measures.html" class="external-link">st_area</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/dimGeometries.html" class="external-link">spotPoly</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span></span></code></pre></div>
<p>For spots that intersect tissue, does total counts relate to
percentage of the spot in tissue?</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">[</span>,<span class="va">sfe</span><span class="op">$</span><span class="va">int_tissue</span><span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotColDataBin2D.html">plotColDataBin2D</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, x <span class="op">=</span> <span class="st">"pct_tissue"</span>, y <span class="op">=</span> <span class="st">"sum"</span>, subset <span class="op">=</span> <span class="st">"diff_int_cov"</span>,</span>
<span>                name_true <span class="op">=</span> <span class="st">"Not covered"</span>, name_false <span class="op">=</span> <span class="st">"Covered"</span>, bins <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-40-1.png" width="700"></p>
<p>Spots that are not fully covered by tissue have lower total UMI
counts, which can be due to both that they are not fully in tissue and
the cell types with lower total counts in the histological region near
the edge, as some spots fully covered by tissue also have low UMI
counts.</p>
</div>
<div class="section level2">
<h2 id="spatial-autocorrelation-of-qc-metrics">Spatial autocorrelation of QC metrics<a class="anchor" aria-label="anchor" href="#spatial-autocorrelation-of-qc-metrics"></a>
</h2>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"visium"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findVisiumGraph.html" class="external-link">findVisiumGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qc_features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sum"</span>, <span class="st">"detected"</span>, <span class="st">"subsets_mito_percent"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">colDataUnivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"moran.mc"</span>, <span class="va">qc_features</span>, nsim <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotMoranMC.html">plotMoranMC</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="va">qc_features</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-44-1.png" width="700"></p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">colDataUnivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"sp.correlogram"</span>, <span class="va">qc_features</span>,</span>
<span>                                order <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotCorrelogram.html">plotCorrelogram</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="va">qc_features</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-46-1.png" width="700"></p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">colDataUnivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localmoran"</span>, <span class="va">qc_features</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localmoran"</span>, <span class="va">qc_features</span>, ncol <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, divergent <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                diverge_center <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-48-1.png" width="768"></p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">colDataUnivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"LOSH"</span>, <span class="va">qc_features</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"LOSH"</span>, <span class="va">qc_features</span>, ncol <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-50-1.png" width="768"></p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">colDataUnivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"moran.plot"</span>, <span class="va">qc_features</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/moranPlot.html">moranPlot</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"subsets_mito_percent"</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-52-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="spatial-autocorrelation-of-gene-expression">Spatial autocorrelation of gene expression<a class="anchor" aria-label="anchor" href="#spatial-autocorrelation-of-gene-expression"></a>
</h2>
<p>Normalize data with the <code>scran</code> method, and find highly
variable genes</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#clusters &lt;- quickCluster(sfe_tissue)</span></span>
<span><span class="co">#sfe_tissue &lt;- computeSumFactors(sfe_tissue, clusters=clusters)</span></span>
<span><span class="co">#sfe_tissue &lt;- sfe_tissue[, sizeFactors(sfe_tissue) &gt; 0]</span></span>
<span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span>
<span><span class="va">hvgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">dec</span>, n <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span></code></pre></div>
<p>Find Moran’s I for all highly variable genes</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">runMoransI</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, features <span class="op">=</span> <span class="va">hvgs</span>, BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotColDataHistogram.html">plotRowDataHistogram</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"moran_sample01"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 30285 rows containing non-finite values (`stat_bin()`).</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-56-1.png" width="700"></p>
<p>The vast majority of genes have positive Moran’s I. Here we plot the
genes with the highest Moran’s I.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_moran</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">moran_sample01</span>, </span>
<span>                                        decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">9</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="va">top_moran</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-58-1.png" width="864"></p>
<p>Here global Moran’s I seems to be more about tissue structure.</p>
<p>Some genes have negative Moran’s I that might not be statistically
significant</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">neg_moran</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">moran_sample01</span>, </span>
<span>                                        decreasing <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">9</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="va">neg_moran</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-60-1.png" width="864"></p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">runUnivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"moran.mc"</span>, <span class="va">neg_moran</span>, </span>
<span>                            colGraphName <span class="op">=</span> <span class="st">"visium"</span>, nsim <span class="op">=</span> <span class="fl">200</span>, alternative <span class="op">=</span> <span class="st">"less"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotMoranMC.html">plotMoranMC</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="va">neg_moran</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-62-1.png" width="700"></p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">[</span><span class="va">neg_moran</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"moran_sample01"</span>, <span class="st">"moran.mc_p.value_sample01"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">#&gt; DataFrame with 9 rows and 2 columns</span></span>
<span><span class="co">#&gt;                    moran_sample01 moran.mc_p.value_sample01</span></span>
<span><span class="co">#&gt;                         &lt;numeric&gt;                 &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; ENSMUSG00000041426     -0.0531915                0.00497512</span></span>
<span><span class="co">#&gt; ENSMUSG00000048277     -0.0451179                0.00497512</span></span>
<span><span class="co">#&gt; ENSMUSG00000021236     -0.0445148                0.01990050</span></span>
<span><span class="co">#&gt; ENSMUSG00000025241     -0.0419121                0.01990050</span></span>
<span><span class="co">#&gt; ENSMUSG00000026126     -0.0399917                0.00995025</span></span>
<span><span class="co">#&gt; ENSMUSG00000034342     -0.0393964                0.02487562</span></span>
<span><span class="co">#&gt; ENSMUSG00000047205     -0.0381599                0.00497512</span></span>
<span><span class="co">#&gt; ENSMUSG00000031431     -0.0369456                0.00995025</span></span>
<span><span class="co">#&gt; ENSMUSG00000037552     -0.0368969                0.01990050</span></span></code></pre></div>
<p>As there’re 2000 highly variable genes and 2000 tests, these would no
longer be significant after correcting for multiple testing.</p>
<p>Does global Moran’s I relate to gene expression level?</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/addPerCellQCMetrics.html" class="external-link">addPerFeatureQCMetrics</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "symbol"                        "moran_sample01"               </span></span>
<span><span class="co">#&gt;  [3] "K_sample01"                    "moran.mc_statistic_sample01"  </span></span>
<span><span class="co">#&gt;  [5] "moran.mc_parameter_sample01"   "moran.mc_p.value_sample01"    </span></span>
<span><span class="co">#&gt;  [7] "moran.mc_alternative_sample01" "moran.mc_method_sample01"     </span></span>
<span><span class="co">#&gt;  [9] "moran.mc_data.name_sample01"   "moran.mc_res_sample01"        </span></span>
<span><span class="co">#&gt; [11] "mean"                          "detected"</span></span></code></pre></div>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotRowData.html" class="external-link">plotRowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, x <span class="op">=</span> <span class="st">"mean"</span>, y <span class="op">=</span> <span class="st">"moran_sample01"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotation_logticks.html" class="external-link">annotation_logticks</a></span><span class="op">(</span>sides <span class="op">=</span> <span class="st">"b"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density_2d.html" class="external-link">geom_density2d</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Transformation introduced infinite values in continuous x-axis</span></span>
<span><span class="co">#&gt; Transformation introduced infinite values in continuous x-axis</span></span>
<span><span class="co">#&gt; Warning: Removed 30285 rows containing non-finite values</span></span>
<span><span class="co">#&gt; (`stat_density2d()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 30285 rows containing missing values (`geom_point()`).</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-65-1.png" width="700"></p>
<p>Genes that are more highly expressed overall tend to have higher
Moran’s I.</p>
</div>
<div class="section level2">
<h2 id="apply-spatial-analysis-methods-to-gene-expression-space">Apply spatial analysis methods to gene expression space<a class="anchor" aria-label="anchor" href="#apply-spatial-analysis-methods-to-gene-expression-space"></a>
</h2>
<p>Spatial statistics that require a spatial neighborhood graph can also
be applied to the k nearest neighbor graph not in histological space but
in gene expression space. This is done in more depth in <a href="https://pachterlab.github.io/voyager/articles/nonspatial.html">this
vignette</a>.</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, ncomponents <span class="op">=</span> <span class="fl">30</span>, subset_row <span class="op">=</span> <span class="va">hvgs</span>,</span>
<span>                     scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># scale as in Seurat</span></span></code></pre></div>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">foo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocNeighbors/man/findKNN-methods.html" class="external-link">findKNN</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"PCA"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>, k<span class="op">=</span><span class="fl">10</span>, BNPARAM<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocNeighbors/man/AnnoyParam.html" class="external-link">AnnoyParam</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Split by row</span></span>
<span><span class="va">foo_nb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/asplit.html" class="external-link">asplit</a></span><span class="op">(</span><span class="va">foo</span><span class="op">$</span><span class="va">index</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">dmat</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">/</span><span class="va">foo</span><span class="op">$</span><span class="va">distance</span></span>
<span><span class="co"># Row normalize the weights</span></span>
<span><span class="va">dmat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sweep.html" class="external-link">sweep</a></span><span class="op">(</span><span class="va">dmat</span>, <span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">dmat</span><span class="op">)</span>, FUN <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span></span>
<span><span class="va">glist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/asplit.html" class="external-link">asplit</a></span><span class="op">(</span><span class="va">dmat</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co"># Sort based on index</span></span>
<span><span class="va">ord</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">foo_nb</span>, <span class="va">order</span><span class="op">)</span></span>
<span><span class="va">foo_nb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">foo_nb</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="va">foo_nb</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">ord</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">foo_nb</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"nb"</span></span>
<span><span class="va">glist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">glist</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="va">glist</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">ord</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">listw</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>style <span class="op">=</span> <span class="st">"W"</span>,</span>
<span>              neighbours <span class="op">=</span> <span class="va">foo_nb</span>,</span>
<span>              weights <span class="op">=</span> <span class="va">glist</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">listw</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"listw"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">listw</span>, <span class="st">"region.id"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"knn10"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">listw</span></span></code></pre></div>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">runMoransI</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, features <span class="op">=</span> <span class="va">hvgs</span>, BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>,</span>
<span>                         colGraphName <span class="op">=</span> <span class="st">"knn10"</span><span class="op">)</span></span></code></pre></div>
<p>Caveat in the current version of Voyager: running Moran’s I again
with a different neighborhood graph will overwrite the existing Moran’s
I results. In the next version, users may specify new names for the same
statistics with different parameters and the parameters will be stored
in <code>rowData</code>.</p>
<p>These are the genes that tend to be more similar to their neighbors
in the 10 nearest neighbor graph in PCA space for gene expression rather
than in histological space:</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_moran2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">moran_sample01</span>, </span>
<span>                                        decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">9</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="va">top_moran2</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-71-1.png" width="864"></p>
<p>Although this Moran’s I was not computed in histological space, these
genes with the highest Moran’s I in PCA space also show spatial
structure, as different cell types reside in different spatial
regions.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.2.2 (2022-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS Big Sur ... 10.16</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRblas.0.dylib</span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] BiocNeighbors_1.16.0           BiocParallel_1.32.5           </span></span>
<span><span class="co">#&gt;  [3] dplyr_1.1.0                    rmapshaper_0.4.6              </span></span>
<span><span class="co">#&gt;  [5] sf_1.0-9                       rlang_1.0.6                   </span></span>
<span><span class="co">#&gt;  [7] terra_1.7-3                    EBImage_4.40.0                </span></span>
<span><span class="co">#&gt;  [9] rjson_0.2.21                   bluster_1.8.0                 </span></span>
<span><span class="co">#&gt; [11] patchwork_1.1.2                stringr_1.5.0                 </span></span>
<span><span class="co">#&gt; [13] scran_1.26.2                   scater_1.27.6                 </span></span>
<span><span class="co">#&gt; [15] scuttle_1.8.4                  ggplot2_3.4.1                 </span></span>
<span><span class="co">#&gt; [17] SingleCellExperiment_1.20.0    SummarizedExperiment_1.28.0   </span></span>
<span><span class="co">#&gt; [19] Biobase_2.58.0                 GenomicRanges_1.50.2          </span></span>
<span><span class="co">#&gt; [21] GenomeInfoDb_1.34.9            IRanges_2.32.0                </span></span>
<span><span class="co">#&gt; [23] S4Vectors_0.36.2               BiocGenerics_0.44.0           </span></span>
<span><span class="co">#&gt; [25] MatrixGenerics_1.10.0          matrixStats_0.63.0            </span></span>
<span><span class="co">#&gt; [27] SpatialFeatureExperiment_1.0.3 Voyager_1.0.10                </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] systemfonts_1.0.4         igraph_1.4.1             </span></span>
<span><span class="co">#&gt;   [3] lazyeval_0.2.2            splines_4.2.2            </span></span>
<span><span class="co">#&gt;   [5] sp_1.6-0                  jqr_1.2.3                </span></span>
<span><span class="co">#&gt;   [7] geojsonlint_0.4.0         digest_0.6.31            </span></span>
<span><span class="co">#&gt;   [9] htmltools_0.5.4           viridis_0.6.2            </span></span>
<span><span class="co">#&gt;  [11] magick_2.7.3              tiff_0.1-11              </span></span>
<span><span class="co">#&gt;  [13] fansi_1.0.4               magrittr_2.0.3           </span></span>
<span><span class="co">#&gt;  [15] memoise_2.0.1             geojsonsf_2.0.3          </span></span>
<span><span class="co">#&gt;  [17] ScaledMatrix_1.6.0        SpatialExperiment_1.8.0  </span></span>
<span><span class="co">#&gt;  [19] cluster_2.1.4             limma_3.54.2             </span></span>
<span><span class="co">#&gt;  [21] R.utils_2.12.2            pkgdown_2.0.7            </span></span>
<span><span class="co">#&gt;  [23] jpeg_0.1-10               colorspace_2.1-0         </span></span>
<span><span class="co">#&gt;  [25] ggrepel_0.9.3             jsonvalidate_1.3.2       </span></span>
<span><span class="co">#&gt;  [27] textshaping_0.3.6         xfun_0.37                </span></span>
<span><span class="co">#&gt;  [29] RCurl_1.98-1.10           jsonlite_1.8.4           </span></span>
<span><span class="co">#&gt;  [31] glue_1.6.2                gtable_0.3.1             </span></span>
<span><span class="co">#&gt;  [33] zlibbioc_1.44.0           XVector_0.38.0           </span></span>
<span><span class="co">#&gt;  [35] DelayedArray_0.24.0       V8_4.2.2                 </span></span>
<span><span class="co">#&gt;  [37] scico_1.3.1               BiocSingular_1.14.0      </span></span>
<span><span class="co">#&gt;  [39] DropletUtils_1.18.1       Rhdf5lib_1.20.0          </span></span>
<span><span class="co">#&gt;  [41] HDF5Array_1.26.0          abind_1.4-5              </span></span>
<span><span class="co">#&gt;  [43] scales_1.2.1              DBI_1.1.3                </span></span>
<span><span class="co">#&gt;  [45] edgeR_3.40.2              Rcpp_1.0.10              </span></span>
<span><span class="co">#&gt;  [47] isoband_0.2.7             viridisLite_0.4.1        </span></span>
<span><span class="co">#&gt;  [49] spData_2.2.2              units_0.8-1              </span></span>
<span><span class="co">#&gt;  [51] dqrng_0.3.0               foreign_0.8-84           </span></span>
<span><span class="co">#&gt;  [53] spdep_1.2-8               rsvd_1.0.5               </span></span>
<span><span class="co">#&gt;  [55] proxy_0.4-27              metapod_1.6.0            </span></span>
<span><span class="co">#&gt;  [57] htmlwidgets_1.6.1         RColorBrewer_1.1-3       </span></span>
<span><span class="co">#&gt;  [59] wk_0.7.1                  farver_2.1.1             </span></span>
<span><span class="co">#&gt;  [61] pkgconfig_2.0.3           R.methodsS3_1.8.2        </span></span>
<span><span class="co">#&gt;  [63] sass_0.4.5                deldir_1.0-6             </span></span>
<span><span class="co">#&gt;  [65] crul_1.3                  locfit_1.5-9.7           </span></span>
<span><span class="co">#&gt;  [67] utf8_1.2.3                labeling_0.4.2           </span></span>
<span><span class="co">#&gt;  [69] tidyselect_1.2.0          munsell_0.5.0            </span></span>
<span><span class="co">#&gt;  [71] tools_4.2.2               cachem_1.0.7             </span></span>
<span><span class="co">#&gt;  [73] cli_3.6.0                 dbscan_1.1-11            </span></span>
<span><span class="co">#&gt;  [75] generics_0.1.3            evaluate_0.20            </span></span>
<span><span class="co">#&gt;  [77] fastmap_1.1.1             fftwtools_0.9-11         </span></span>
<span><span class="co">#&gt;  [79] yaml_2.3.7                ragg_1.2.5               </span></span>
<span><span class="co">#&gt;  [81] knitr_1.42                fs_1.6.1                 </span></span>
<span><span class="co">#&gt;  [83] purrr_1.0.1               s2_1.1.2                 </span></span>
<span><span class="co">#&gt;  [85] nlme_3.1-162              sparseMatrixStats_1.10.0 </span></span>
<span><span class="co">#&gt;  [87] R.oo_1.25.0               compiler_4.2.2           </span></span>
<span><span class="co">#&gt;  [89] curl_5.0.0                beeswarm_0.4.0           </span></span>
<span><span class="co">#&gt;  [91] png_0.1-8                 e1071_1.7-13             </span></span>
<span><span class="co">#&gt;  [93] tibble_3.1.8              statmod_1.5.0            </span></span>
<span><span class="co">#&gt;  [95] bslib_0.4.2               stringi_1.7.12           </span></span>
<span><span class="co">#&gt;  [97] highr_0.10                desc_1.4.2               </span></span>
<span><span class="co">#&gt;  [99] rgeos_0.6-1               lattice_0.20-45          </span></span>
<span><span class="co">#&gt; [101] Matrix_1.5-3              classInt_0.4-9           </span></span>
<span><span class="co">#&gt; [103] vctrs_0.5.2               pillar_1.8.1             </span></span>
<span><span class="co">#&gt; [105] lifecycle_1.0.3           rhdf5filters_1.10.0      </span></span>
<span><span class="co">#&gt; [107] jquerylib_0.1.4           cowplot_1.1.1            </span></span>
<span><span class="co">#&gt; [109] maptools_1.1-6            geojsonio_0.10.0         </span></span>
<span><span class="co">#&gt; [111] bitops_1.0-7              irlba_2.3.5.1            </span></span>
<span><span class="co">#&gt; [113] R6_2.5.1                  KernSmooth_2.23-20       </span></span>
<span><span class="co">#&gt; [115] gridExtra_2.3             vipor_0.4.5              </span></span>
<span><span class="co">#&gt; [117] codetools_0.2-19          MASS_7.3-58.2            </span></span>
<span><span class="co">#&gt; [119] boot_1.3-28.1             rhdf5_2.42.0             </span></span>
<span><span class="co">#&gt; [121] rprojroot_2.0.3           httpcode_0.3.0           </span></span>
<span><span class="co">#&gt; [123] withr_2.5.0               GenomeInfoDbData_1.2.9   </span></span>
<span><span class="co">#&gt; [125] mgcv_1.8-42               geojson_0.3.4            </span></span>
<span><span class="co">#&gt; [127] parallel_4.2.2            grid_4.2.2               </span></span>
<span><span class="co">#&gt; [129] beachmat_2.14.0           class_7.3-21             </span></span>
<span><span class="co">#&gt; [131] rmarkdown_2.20            DelayedMatrixStats_1.20.0</span></span>
<span><span class="co">#&gt; [133] ggnewscale_0.4.8          ggbeeswarm_0.7.1</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Lambda Moses, Kayla Jackson, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
