<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Voyager">
<title>Spatial analysis of 10X example Visium dataset • Voyager</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Spatial analysis of 10X example Visium dataset">
<meta property="og:description" content="Voyager">
<meta name="robots" content="noindex">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">Voyager</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">1.2.3</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/install.html">Install</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-technologies">Technologies</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-technologies">
    <a class="dropdown-item" href="../articles/visium_landing.html">Visium</a>
    <a class="dropdown-item" href="../articles/slideseqV2_landing.html">Slide-seq v2</a>
    <a class="dropdown-item" href="../articles/cosmx_landing.html">CosMX</a>
    <a class="dropdown-item" href="../articles/xenium_landing.html">Xenium</a>
    <a class="dropdown-item" href="../articles/merfish_landing.html">MERFISH</a>
    <a class="dropdown-item" href="../articles/seqfish_landing.html">seqFISH</a>
    <a class="dropdown-item" href="../articles/codex_landing.html">CODEX/PhenoCycler</a>
    <a class="dropdown-item" href="../articles/chromium_landing.html">10X Chromium V3</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-methods">Methods</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-methods">
    <h6 class="dropdown-header" data-toc-skip>Univariate global</h6>
    <a class="dropdown-item" href="../articles/moran_landing.html">Moran's I</a>
    <a class="dropdown-item" href="../articles/geary_landing.html">Geary's C</a>
    <a class="dropdown-item" href="../articles/correlogram_landing.html">Correlogram</a>
    <a class="dropdown-item" href="../articles/variogram_landing.html">Variogram</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Univariate local</h6>
    <a class="dropdown-item" href="../articles/moranplot_landing.html">Moran scatter plot</a>
    <a class="dropdown-item" href="../articles/localmoran_landing.html">Local Moran's I</a>
    <a class="dropdown-item" href="../articles/losh_landing.html">Local spatial heteroscedasticity</a>
    <a class="dropdown-item" href="../articles/getisord_landing.html">Getis-Ord Gi*</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Bivariate</h6>
    <a class="dropdown-item" href="../articles/lee_landing.html">Lee's L</a>
    <a class="dropdown-item" href="../articles/crossvariogram_landing.html">Cross variogram</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Multivariate</h6>
    <a class="dropdown-item" href="../articles/multispati_landing.html">MULTISPATI PCA</a>
    <a class="dropdown-item" href="../articles/localc_landing.html">Multivariate local Geary's C</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../articles/sfemethod">Custom methods</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/pachterlab/voyager/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Spatial analysis of 10X example Visium dataset</h1>
                        <h4 data-toc-skip class="author">Lambda
Moses</h4>
            
            <h4 data-toc-skip class="date">2023-05-19</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pachterlab/voyager/blob/HEAD/vignettes/visium_10x_spatial.Rmd" class="external-link"><code>vignettes/visium_10x_spatial.Rmd</code></a></small>
      <div class="d-none name"><code>visium_10x_spatial.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In a <a href="https://pachterlab.github.io/voyager/articles/visium_10x.html">more
introductory vignette</a>, we performed basic non-spatial analyses on a
mouse olfactory bulb Visium dataset from the 10X website. In this
vignette, we perform spatial analyses in histological space as well as
in gene expression space.</p>
<p>Here we load the packages used in this vignette:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/voyager" class="external-link">Voyager</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: replacing previous import 'S4Arrays::read_block' by</span></span>
<span><span class="co">#&gt; 'DelayedArray::read_block' when loading 'SummarizedExperiment'</span></span>
<span><span class="co">#&gt; Warning: replacing previous import 'S4Arrays::read_block' by</span></span>
<span><span class="co">#&gt; 'DelayedArray::read_block' when loading 'HDF5Array'</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SpatialFeatureExperiment" class="external-link">SpatialFeatureExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: SummarizedExperiment</span></span>
<span><span class="co">#&gt; Loading required package: MatrixGenerics</span></span>
<span><span class="co">#&gt; Loading required package: matrixStats</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'MatrixGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span></span>
<span><span class="co">#&gt;     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span></span>
<span><span class="co">#&gt;     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span></span>
<span><span class="co">#&gt;     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span></span>
<span><span class="co">#&gt;     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span></span>
<span><span class="co">#&gt;     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span></span>
<span><span class="co">#&gt;     colWeightedMeans, colWeightedMedians, colWeightedSds,</span></span>
<span><span class="co">#&gt;     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span></span>
<span><span class="co">#&gt;     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span></span>
<span><span class="co">#&gt;     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span></span>
<span><span class="co">#&gt;     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span></span>
<span><span class="co">#&gt;     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span></span>
<span><span class="co">#&gt;     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span></span>
<span><span class="co">#&gt;     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span></span>
<span><span class="co">#&gt;     rowWeightedSds, rowWeightedVars</span></span>
<span><span class="co">#&gt; Loading required package: GenomicRanges</span></span>
<span><span class="co">#&gt; Loading required package: stats4</span></span>
<span><span class="co">#&gt; Loading required package: BiocGenerics</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'BiocGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">#&gt;     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">#&gt;     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,</span></span>
<span><span class="co">#&gt;     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span></span>
<span><span class="co">#&gt;     Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,</span></span>
<span><span class="co">#&gt;     table, tapply, union, unique, unsplit, which.max, which.min</span></span>
<span><span class="co">#&gt; Loading required package: S4Vectors</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'S4Vectors'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:utils':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     findMatches</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     expand.grid, I, unname</span></span>
<span><span class="co">#&gt; Loading required package: IRanges</span></span>
<span><span class="co">#&gt; Loading required package: GenomeInfoDb</span></span>
<span><span class="co">#&gt; Loading required package: Biobase</span></span>
<span><span class="co">#&gt; Welcome to Bioconductor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     Vignettes contain introductory material; view with</span></span>
<span><span class="co">#&gt;     'browseVignettes()'. To cite Bioconductor, see</span></span>
<span><span class="co">#&gt;     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'Biobase'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:MatrixGenerics':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     rowMedians</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyMissing, rowMedians</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'ggplot2'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:SpatialFeatureExperiment':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     unit</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: scuttle</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scuttle</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org" class="external-link">stringr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/alexcb/rjson" class="external-link">rjson</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/aoles/EBImage" class="external-link">EBImage</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'EBImage'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:SummarizedExperiment':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     resize</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:Biobase':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     channel</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:GenomicRanges':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     resize, tile</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:IRanges':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     resize, tile</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:SpatialFeatureExperiment':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     transpose</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rspatial.org/" class="external-link">terra</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; terra 1.7.29</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'terra'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:EBImage':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     flip, rotate</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:patchwork':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     area</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:SummarizedExperiment':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     distance, nearest, shift, trim, values, values&lt;-, width</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:GenomicRanges':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     distance, gaps, nearest, shift, trim, values, values&lt;-, width</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:IRanges':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     distance, gaps, nearest, shift, trim, width</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:S4Vectors':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     values, values&lt;-, width</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:BiocGenerics':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     width</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:SpatialFeatureExperiment':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     centroids, crop</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rlang.r-lib.org" class="external-link">rlang</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'rlang'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:Biobase':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     exprs</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-spatial.github.io/sf/" class="external-link">sf</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Linking to GEOS 3.11.0, GDAL 3.5.3, PROJ 9.1.0; sf_use_s2() is TRUE</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ateucher/rmapshaper" class="external-link">rmapshaper</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:terra':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, union</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:EBImage':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     combine</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:Biobase':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     combine</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:GenomicRanges':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, union</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:GenomeInfoDb':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:IRanges':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     collapse, desc, intersect, setdiff, slice, union</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:S4Vectors':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     first, intersect, rename, setdiff, setequal, union</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:BiocGenerics':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     combine, intersect, setdiff, union</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     count</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BiocNeighbors</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rstudio.github.io/reticulate/" class="external-link">reticulate</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Specify Python version to use gget</span></span>
<span><span class="va">PY_PATH</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"which python"</span>, intern <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rstudio.github.io/reticulate/reference/use_python.html" class="external-link">use_python</a></span><span class="op">(</span><span class="va">PY_PATH</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rstudio.github.io/reticulate/reference/py_config.html" class="external-link">py_config</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; python:         /Users/runner/hostedtoolcache/Python/3.8.16/x64/bin/python</span></span>
<span><span class="co">#&gt; libpython:      /Users/runner/hostedtoolcache/Python/3.8.16/x64/lib/libpython3.8.dylib</span></span>
<span><span class="co">#&gt; pythonhome:     /Users/runner/hostedtoolcache/Python/3.8.16/x64:/Users/runner/hostedtoolcache/Python/3.8.16/x64</span></span>
<span><span class="co">#&gt; version:        3.8.16 (default, Jan 11 2023, 00:27:57)  [Clang 12.0.0 (clang-1200.0.32.29)]</span></span>
<span><span class="co">#&gt; numpy:          /Users/runner/hostedtoolcache/Python/3.8.16/x64/lib/python3.8/site-packages/numpy</span></span>
<span><span class="co">#&gt; numpy_version:  1.24.3</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; NOTE: Python version was forced by use_python function</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gget</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rstudio.github.io/reticulate/reference/import.html" class="external-link">import</a></span><span class="op">(</span><span class="st">"gget"</span><span class="op">)</span></span></code></pre></div>
<p>Here we download the data from the 10X website. This is the
unfiltered gene count matrix:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="st">"visium_ob.tar.gz"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://cf.10xgenomics.com/samples/spatial-exp/2.0.0/Visium_Mouse_Olfactory_Bulb/Visium_Mouse_Olfactory_Bulb_raw_feature_bc_matrix.tar.gz"</span>, </span>
<span>                  destfile <span class="op">=</span> <span class="st">"visium_ob.tar.gz"</span><span class="op">)</span></span></code></pre></div>
<p>This is the spatial information:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="st">"visium_ob_spatial.tar.gz"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://cf.10xgenomics.com/samples/spatial-exp/2.0.0/Visium_Mouse_Olfactory_Bulb/Visium_Mouse_Olfactory_Bulb_spatial.tar.gz"</span>, </span>
<span>                  destfile <span class="op">=</span> <span class="st">"visium_ob_spatial.tar.gz"</span><span class="op">)</span></span></code></pre></div>
<p>Decompress the downloaded content:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.exists</a></span><span class="op">(</span><span class="st">"outs"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="st">"outs"</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"tar -xvf visium_ob.tar.gz -C outs"</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"tar -xvf visium_ob_spatial.tar.gz -C outs"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Contents of the <code>outs</code> directory as from Space Ranger is
explained in <a href="https://pachterlab.github.io/voyager/articles/visium_10x.html">the
introductory vignette</a>.</p>
<p>Here we read the data into R as an SFE object.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/read10xVisiumSFE.html" class="external-link">read10xVisiumSFE</a></span><span class="op">(</span>samples <span class="op">=</span> <span class="st">"."</span>, type <span class="op">=</span> <span class="st">"sparse"</span>, data <span class="op">=</span> <span class="st">"raw"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 32285 4992 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(32285): ENSMUSG00000051951 ENSMUSG00000089699 ...</span></span>
<span><span class="co">#&gt;   ENSMUSG00000095019 ENSMUSG00000095041</span></span>
<span><span class="co">#&gt; rowData names(8): symbol Feature.Type ...</span></span>
<span><span class="co">#&gt;   Median.Normalized.Average.Counts_sample01</span></span>
<span><span class="co">#&gt;   Barcodes.Detected.per.Feature_sample01</span></span>
<span><span class="co">#&gt; colnames(4992): AAACAACGAATAGTTC-1 AAACAAGTATCTCCCA-1 ...</span></span>
<span><span class="co">#&gt;   TTGTTTGTATTACACG-1 TTGTTTGTGTAAATTC-1</span></span>
<span><span class="co">#&gt; colData names(4): in_tissue array_row array_col sample_id</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : pxl_col_in_fullres pxl_row_in_fullres</span></span>
<span><span class="co">#&gt; imgData names(4): sample_id image_id data scaleFactor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: full_res_image_pixel</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: spotPoly (POLYGON) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01:</span></span></code></pre></div>
<p>Here we add QC metrics, already plotted in the introductory
vignette.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">is_mt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">symbol</span>, <span class="st">"^mt-"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/addPerCellQCMetrics.html" class="external-link">addPerCellQCMetrics</a></span><span class="op">(</span><span class="va">sfe</span>, subsets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>mito <span class="op">=</span> <span class="va">is_mt</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="tissue-segmentation">Tissue segmentation<a class="anchor" aria-label="anchor" href="#tissue-segmentation"></a>
</h2>
<p>While Space Ranger can automatically detect which spots are in tissue
and the Loupe browser can be used to manually annotate which spots are
in tissue, it may be interesting to get the tissue outline polygon, so
we would know how much each spot overlaps with the tissue and plot the
outline. The tissue boundary polygon can be manually annotated with
QuPath, which saves the polygon as a GeoJSON and can be directly read
into R with <code><a href="https://r-spatial.github.io/sf/reference/st_read.html" class="external-link">st_read()</a></code>.</p>
<p>Or we can segment the tissue computationally. R generally isn’t great
for image processing, but there are some packages that can perform the
segmentation, such as <a href="https://bioconductor.org/packages/release/bioc/html/EBImage.html" class="external-link"><code>EBImage</code></a>,
which is based on its own in house C and C++ code, and <a href="https://github.com/dahtah/imager/" class="external-link"><code>imager</code></a>, which
is based on <a href="http://cimg.eu/" class="external-link"><code>CImg</code></a>.</p>
<p>Here we don’t have the full resolution image. We will perform tissue
segmentation on the high resolution downsampled image and then scale it
to make the coordinates of the tissue boundary match those of the spots.
The <code>EBImage</code> package is used here. Compared to OpenCV,
<code>EBImage</code> is slow on the full resolution image, but should be
fine here for the downsized image.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">img</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/io.html" class="external-link">readImage</a></span><span class="op">(</span><span class="st">"outs/spatial/tissue_hires_image.png"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/display.html" class="external-link">display</a></span><span class="op">(</span><span class="va">img</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<p>When rendered as a static webpage, the image is static, but when run
interactively, this image will be shown in an interactive widget where
you can zoom and pan.</p>
<p>Here we show the RGB channels separately</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">img2</span> <span class="op">&lt;-</span> <span class="va">img</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/Image.html" class="external-link">colorMode</a></span><span class="op">(</span><span class="va">img2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">Grayscale</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/display.html" class="external-link">display</a></span><span class="op">(</span><span class="va">img2</span>, all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">img</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<p>The tissue can be discerned with thresholding. The tall peak on the
right is the background. The much lower peaks from around 0.6 to 0.85
must be the tissue. To capture the faint bluish region, the blue channel
is used for thresholding. The threshold here is chosen based on the
histogram and experimenting with nearby values.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mask</span> <span class="op">&lt;-</span> <span class="va">img2</span><span class="op">[</span>,,<span class="fl">3</span><span class="op">]</span> <span class="op">&lt;</span> <span class="fl">0.87</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/display.html" class="external-link">display</a></span><span class="op">(</span><span class="va">mask</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<p>Then we use an opening operation (erosion followed by dilation) to
denoise</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kern</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/morphology.html" class="external-link">makeBrush</a></span><span class="op">(</span><span class="fl">3</span>, shape<span class="op">=</span><span class="st">'disc'</span><span class="op">)</span></span>
<span><span class="va">mask_open</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/morphology.html" class="external-link">opening</a></span><span class="op">(</span><span class="va">mask</span>, <span class="va">kern</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/display.html" class="external-link">display</a></span><span class="op">(</span><span class="va">mask_open</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
<p>There are some small holes in the tissue, which can be removed by a
closing operation (dilation followed by erosion):</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mask_close</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/morphology.html" class="external-link">closing</a></span><span class="op">(</span><span class="va">mask_open</span>, <span class="va">kern</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/display.html" class="external-link">display</a></span><span class="op">(</span><span class="va">mask_close</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
<p>There are some larger holes in the tissue mask, which may be real
holes or faint regions with few nuclei missed by thresholding. They
might not be large enough to affect which Visium spots intersect the
tissue.</p>
<p>Now the main piece of tissue is clear. It must be the object with the
largest area. However, there are two small pieces that should belong to
the tissue at the top left. The debris and fiducials can be removed by
setting all pixels in the mask outside the bounding box of the main
piece to 0. Here we assign a different value to each contiguous object
with <code><a href="https://rdrr.io/pkg/EBImage/man/bwlabel.html" class="external-link">bwlabel()</a></code>, and use
<code><a href="https://rdrr.io/pkg/EBImage/man/computeFeatures.html" class="external-link">computeFeatures.shape()</a></code> to find the area among other shape
features (e.g. perimeter) of each object.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mask_label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/bwlabel.html" class="external-link">bwlabel</a></span><span class="op">(</span><span class="va">mask_close</span><span class="op">)</span></span>
<span><span class="va">fts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/computeFeatures.html" class="external-link">computeFeatures.shape</a></span><span class="op">(</span><span class="va">mask_label</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">fts</span><span class="op">)</span></span>
<span><span class="co">#&gt;   s.area s.perimeter s.radius.mean s.radius.sd s.radius.min s.radius.max</span></span>
<span><span class="co">#&gt; 1     39          25      3.428773   1.3542219    1.4176036     5.762777</span></span>
<span><span class="co">#&gt; 2     20          14      2.032665   0.3439068    1.5000000     2.500000</span></span>
<span><span class="co">#&gt; 3      8           8      1.144123   0.4370160    0.7071068     1.581139</span></span>
<span><span class="co">#&gt; 4     14          10      1.689175   0.2160726    1.5811388     2.121320</span></span>
<span><span class="co">#&gt; 5     15          12      1.716761   0.4684015    1.0000000     2.236068</span></span>
<span><span class="co">#&gt; 6      9           8      1.207107   0.2071068    1.0000000     1.414214</span></span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">fts</span><span class="op">[</span>,<span class="st">"s.area"</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;     Min.  1st Qu.   Median     Mean  3rd Qu.     Max. </span></span>
<span><span class="co">#&gt;      8.0     14.0     51.0    595.9    345.0 496326.0</span></span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">max_ind</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/terra/man/summarize-generics.html" class="external-link">which.max</a></span><span class="op">(</span><span class="va">fts</span><span class="op">[</span>,<span class="st">"s.area"</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">inds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/coerce.html" class="external-link">as.array</a></span><span class="op">(</span><span class="va">mask_label</span><span class="op">)</span> <span class="op">==</span> <span class="va">max_ind</span>, arr.ind <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">inds</span><span class="op">)</span></span>
<span><span class="co">#&gt;       row col</span></span>
<span><span class="co">#&gt; [1,] 1168 562</span></span>
<span><span class="co">#&gt; [2,] 1169 562</span></span>
<span><span class="co">#&gt; [3,] 1170 562</span></span>
<span><span class="co">#&gt; [4,] 1158 563</span></span>
<span><span class="co">#&gt; [5,] 1159 563</span></span>
<span><span class="co">#&gt; [6,] 1160 563</span></span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">row_inds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">inds</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">inds</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">+</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/pkg/terra/man/dimensions.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">mask_label</span><span class="op">)</span>, by <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">col_inds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">inds</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">inds</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">+</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/pkg/terra/man/dimensions.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">mask_label</span><span class="op">)</span>, by <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">mask_label</span><span class="op">[</span><span class="va">row_inds</span>, <span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span><span class="va">mask_label</span><span class="op">[</span>,<span class="va">col_inds</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/display.html" class="external-link">display</a></span><span class="op">(</span><span class="va">mask_label</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-23-1.png" width="700"></p>
<p>Then remove the small pieces that are debris.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="va">mask_label</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1]   0 421 425 429 430 438 450 458 461 469 473 483 487 505 523 633 640 642 651</span></span>
<span><span class="co">#&gt; [20] 678 739 741 757 762 775 778 789 791 797 805 810 813 820 821 822 826 831 838</span></span>
<span><span class="co">#&gt; [39] 839 840 843 845 848 849 861 862 863</span></span></code></pre></div>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fts2</span> <span class="op">&lt;-</span> <span class="va">fts</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="va">mask_label</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span>,<span class="op">]</span></span>
<span><span class="va">fts2</span> <span class="op">&lt;-</span> <span class="va">fts2</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">fts2</span><span class="op">[</span>,<span class="st">"s.area"</span><span class="op">]</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,<span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">fts2</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span>, type <span class="op">=</span> <span class="st">"l"</span>, ylab <span class="op">=</span> <span class="st">"Area"</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-26-1.png" width="700"></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">fts2</span>, <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt;     s.area s.perimeter s.radius.mean s.radius.sd s.radius.min s.radius.max</span></span>
<span><span class="co">#&gt; 421 496326        3151    395.118732  68.6493949  234.1605637   485.715835</span></span>
<span><span class="co">#&gt; 450    217          55      7.840627   1.2883202    5.0010247    10.458197</span></span>
<span><span class="co">#&gt; 849    211          63      7.961248   2.0228566    3.8569142    12.189753</span></span>
<span><span class="co">#&gt; 797    182          56      7.547362   2.3368359    3.0772370    11.839919</span></span>
<span><span class="co">#&gt; 461    136          54      7.186020   3.2751628    0.9255555    12.479805</span></span>
<span><span class="co">#&gt; 741     92          56      6.365661   2.8382829    1.3273276    11.653219</span></span>
<span><span class="co">#&gt; 840     69          33      4.503526   1.6417370    1.6026264     7.076974</span></span>
<span><span class="co">#&gt; 862     63          37      4.854424   2.4445530    0.6361407     8.837838</span></span>
<span><span class="co">#&gt; 839     45          25      3.305562   0.7074306    1.9320455     4.526897</span></span>
<span><span class="co">#&gt; 775     32          22      2.887407   1.1755375    0.5300865     4.543636</span></span></code></pre></div>
<p>Object number 797 is a piece of debris at the bottom left. The other
pieces with area over 100 pixels are tissue. Since debris really is
small bits of tissue, so the boundary between debris and tissue can be
blurry. Here the two are distinguished by morphology on the H&amp;E
image and proximity to the main tissue.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#display(mask_label == 797)</span></span></code></pre></div>
<p>Here we remove the debris from the mask</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mask_label</span><span class="op">[</span><span class="va">mask_label</span> <span class="op"><a href="https://rdrr.io/pkg/terra/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">797</span>, <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">fts2</span><span class="op">)</span><span class="op">[</span><span class="va">fts2</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span> <span class="op">&lt;</span> <span class="fl">100</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span></code></pre></div>
<p>Since most holes in the mask are faint regions of the tissue missed
by thresholding, the holes will be filled</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mask_label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/fillHull.html" class="external-link">fillHull</a></span><span class="op">(</span><span class="va">mask_label</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/display.html" class="external-link">display</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/paintObjects.html" class="external-link">paintObjects</a></span><span class="op">(</span><span class="va">mask_label</span>, <span class="va">img</span>, col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"red"</span>, <span class="st">"yellow"</span><span class="op">)</span>, opac<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0.3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-31-1.png" width="700"></p>
<p>This segmentation process took a lot of manual oversight, in choosing
the threshold, choosing kernel size and shape in the opening and closing
operations, deciding whether to fill the holes, and deciding what is
debris and what is tissue.</p>
</div>
<div class="section level2">
<h2 id="convert-tissue-mask-to-polygon">Convert tissue mask to polygon<a class="anchor" aria-label="anchor" href="#convert-tissue-mask-to-polygon"></a>
</h2>
<p>Now we have the tissue mask, which we will convert to polygon. While
OpenCV can directly perform the conversion, as there isn’t a
comprehensive R wrapper of OpenCV, this conversion is more convoluted in
R. We first convert the <code>Image</code> object to a raster as
implemented in <code>terra</code>, the core R package for geospatial
raster data. Then <code>terra</code> can convert the raster to polygon.
As this image is downsized, the polygon will look quite pixelated. To
mitigate this pixelation and save memory, the <code><a href="https://rdrr.io/pkg/rmapshaper/man/ms_simplify.html" class="external-link">ms_simplify()</a></code>
function is used to simplify the polygon, only keeping a small
proportion of all vertices. The <code><a href="https://r-spatial.github.io/sf/reference/geos_unary.html" class="external-link">st_simplify()</a></code> function in
<code>sf</code> can also simplify the polygons, but we can’t specify
what proportion of vertices to keep.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">raster2polygon</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">seg</span>, <span class="va">keep</span> <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">r</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/terra/man/rast.html" class="external-link">rast</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/coerce.html" class="external-link">as.array</a></span><span class="op">(</span><span class="va">seg</span><span class="op">)</span>, extent <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/terra/man/ext.html" class="external-link">ext</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/pkg/terra/man/dimensions.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">seg</span><span class="op">)</span>, <span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/pkg/terra/man/dimensions.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">seg</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>        <span class="fu"><a href="https://rdrr.io/pkg/terra/man/transpose.html" class="external-link">trans</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/EBImage/man/spatial.html" class="external-link">flip</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="va">r</span><span class="op">[</span><span class="va">r</span> <span class="op">&lt;</span> <span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span>    <span class="va">contours</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_as_sf.html" class="external-link">st_as_sf</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/as.spatvector.html" class="external-link">as.polygons</a></span><span class="op">(</span><span class="va">r</span>, dissolve <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">simplified</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rmapshaper/man/ms_simplify.html" class="external-link">ms_simplify</a></span><span class="op">(</span><span class="va">contours</span>, keep <span class="op">=</span> <span class="va">keep</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>full <span class="op">=</span> <span class="va">contours</span>,</span>
<span>         simplified <span class="op">=</span> <span class="va">simplified</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tb</span> <span class="op">&lt;-</span> <span class="fu">raster2polygon</span><span class="op">(</span><span class="va">mask_label</span><span class="op">)</span></span></code></pre></div>
<p>Before adding the geometry to the SFE object, it needs to be scaled
to match the coordinates of the spots</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scale_factors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rjson/man/fromJSON.html" class="external-link">fromJSON</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"outs/spatial/scalefactors_json.json"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tb</span><span class="op">$</span><span class="va">simplified</span><span class="op">$</span><span class="va">geometry</span> <span class="op">&lt;-</span> <span class="va">tb</span><span class="op">$</span><span class="va">simplified</span><span class="op">$</span><span class="va">geometry</span> <span class="op">/</span> <span class="va">scale_factors</span><span class="op">$</span><span class="va">tissue_hires_scalef</span></span></code></pre></div>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotGeometries.html" class="external-link">tissueBoundary</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">tb</span><span class="op">$</span><span class="va">simplified</span></span></code></pre></div>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"sum"</span>, annotGeometryName <span class="op">=</span> <span class="st">"tissueBoundary"</span>, </span>
<span>                   annot_fixed <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="cn">NA</span>, color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span>,</span>
<span>                   image_id <span class="op">=</span> <span class="st">"lowres"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-37-1.png" width="700"></p>
<p>The mouse olfactory bulb is conventionally plotted horizontally. The
entire SFE object can be transposed in histologial space to make the
olfactory bulb horizontal.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu">SpatialFeatureExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/SFE-transform.html" class="external-link">transpose</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"sum"</span>, annotGeometryName <span class="op">=</span> <span class="st">"tissueBoundary"</span>, </span>
<span>                   annot_fixed <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="cn">NA</span>, color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span>,</span>
<span>                   image_id <span class="op">=</span> <span class="st">"lowres"</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-39-1.png" width="700"></p>
<p>Then we can use geometric operations to find which spots intersect
tissue, which spots are covered by tissue, and how much of each spot
intersects tissue.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Which spots intersect tissue</span></span>
<span><span class="va">sfe</span><span class="op">$</span><span class="va">int_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotPred.html" class="external-link">annotPred</a></span><span class="op">(</span><span class="va">sfe</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, </span>
<span>                            annotGeometryName <span class="op">=</span> <span class="st">"tissueBoundary"</span>,</span>
<span>                            pred <span class="op">=</span> <span class="va">st_intersects</span><span class="op">)</span></span>
<span><span class="va">sfe</span><span class="op">$</span><span class="va">cov_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotPred.html" class="external-link">annotPred</a></span><span class="op">(</span><span class="va">sfe</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, </span>
<span>                            annotGeometryName <span class="op">=</span> <span class="st">"tissueBoundary"</span>,</span>
<span>                            pred <span class="op">=</span> <span class="va">st_covered_by</span><span class="op">)</span></span></code></pre></div>
<p>Discrepancies between Space Ranger’s annotation and the annotation
based on tissue segmentation here:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span><span class="op">$</span><span class="va">diff_sr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html" class="external-link">case_when</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">$</span><span class="va">in_tissue</span> <span class="op">==</span> <span class="va">sfe</span><span class="op">$</span><span class="va">int_tissue</span> <span class="op">~</span> <span class="st">"same"</span>,</span>
<span>                         <span class="va">sfe</span><span class="op">$</span><span class="va">in_tissue</span> <span class="op">&amp;</span> <span class="op">!</span><span class="va">sfe</span><span class="op">$</span><span class="va">int_tissue</span> <span class="op">~</span> <span class="st">"Space Ranger"</span>,</span>
<span>                         <span class="va">sfe</span><span class="op">$</span><span class="va">int_tissue</span> <span class="op">&amp;</span> <span class="op">!</span><span class="va">sfe</span><span class="op">$</span><span class="va">in_tissue</span> <span class="op">~</span> <span class="st">"segmentation"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span>levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Space Ranger"</span>, <span class="st">"same"</span>, <span class="st">"segmentation"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"diff_sr"</span>, </span>
<span>                   annotGeometryName <span class="op">=</span> <span class="st">"tissueBoundary"</span>, </span>
<span>                   annot_fixed <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="cn">NA</span>, size <span class="op">=</span> <span class="fl">0.5</span>, color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_brewer</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"div"</span>, palette <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">fill</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">fill</span>, which will replace the existing scale.</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-41-1.png" width="700"></p>
<p>Spots at the margin can intersect the tissue without being covered by
it.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span><span class="op">$</span><span class="va">diff_int_cov</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">$</span><span class="va">int_tissue</span> <span class="op">!=</span> <span class="va">sfe</span><span class="op">$</span><span class="va">cov_tissue</span></span>
<span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"diff_int_cov"</span>, </span>
<span>                   annotGeometryName <span class="op">=</span> <span class="st">"tissueBoundary"</span>, </span>
<span>                   annot_fixed <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="cn">NA</span>, size <span class="op">=</span> <span class="fl">0.5</span>, color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-42-1.png" width="700"></p>
<p>We can also get the geometries of the intersections between the
tissue and the Visium spots, and then calculate what percentage of each
spot is in tissue. However, this percentage may not be very useful if
the tissue segmentation is subject to error. This percentage may be more
useful for pathologist annotated histological regions or objects such as
nuclei and myofibers.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spot_ints</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/annotOp.html" class="external-link">annotOp</a></span><span class="op">(</span><span class="va">sfe</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, </span>
<span>                     annotGeometryName <span class="op">=</span> <span class="st">"tissueBoundary"</span>, op <span class="op">=</span> <span class="va">st_intersection</span><span class="op">)</span></span>
<span><span class="va">sfe</span><span class="op">$</span><span class="va">pct_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://r-spatial.github.io/sf/reference/geos_measures.html" class="external-link">st_area</a></span><span class="op">(</span><span class="va">spot_ints</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://r-spatial.github.io/sf/reference/geos_measures.html" class="external-link">st_area</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/dimGeometries.html" class="external-link">spotPoly</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span></span></code></pre></div>
<p>For spots that intersect tissue, does total counts relate to
percentage of the spot in tissue?</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">[</span>,<span class="va">sfe</span><span class="op">$</span><span class="va">int_tissue</span><span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotColDataBin2D.html">plotColDataBin2D</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, x <span class="op">=</span> <span class="st">"pct_tissue"</span>, y <span class="op">=</span> <span class="st">"sum"</span>, subset <span class="op">=</span> <span class="st">"diff_int_cov"</span>,</span>
<span>                name_true <span class="op">=</span> <span class="st">"Not covered"</span>, name_false <span class="op">=</span> <span class="st">"Covered"</span>, bins <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: plotCol/RowDataBin2D() was deprecated in Voyager 1.2.0.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Please use scater::plotCol/RowData() instead.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Set `bins` argument to enable binning.</span></span>
<span><span class="co">#&gt; <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">generated.</span></span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-45-1.png" width="700"></p>
<p>Spots that are not fully covered by tissue have lower total UMI
counts, which can be due to both that they are not fully in tissue and
the cell types with lower total counts in the histological region near
the edge, as some spots fully covered by tissue also have low UMI
counts.</p>
</div>
<div class="section level2">
<h2 id="spatial-autocorrelation-of-qc-metrics">Spatial autocorrelation of QC metrics<a class="anchor" aria-label="anchor" href="#spatial-autocorrelation-of-qc-metrics"></a>
</h2>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"visium"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findVisiumGraph.html" class="external-link">findVisiumGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qc_features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sum"</span>, <span class="st">"detected"</span>, <span class="st">"subsets_mito_percent"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">colDataUnivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"moran.mc"</span>, <span class="va">qc_features</span>, nsim <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotMoranMC.html">plotMoranMC</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="va">qc_features</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-49-1.png" width="700"></p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">colDataUnivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"sp.correlogram"</span>, <span class="va">qc_features</span>,</span>
<span>                                order <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotCorrelogram.html">plotCorrelogram</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="va">qc_features</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-51-1.png" width="700"></p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">colDataUnivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localmoran"</span>, <span class="va">qc_features</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"localmoran"</span>, <span class="va">qc_features</span>, ncol <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, divergent <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                diverge_center <span class="op">=</span> <span class="fl">0</span>, image_id <span class="op">=</span> <span class="st">"lowres"</span>, maxcell <span class="op">=</span> <span class="fl">5e4</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-53-1.png" width="768"></p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">colDataUnivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"LOSH"</span>, <span class="va">qc_features</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotLocalResult.html">plotLocalResult</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"LOSH"</span>, <span class="va">qc_features</span>, ncol <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, image_id <span class="op">=</span> <span class="st">"lowres"</span>, maxcell <span class="op">=</span> <span class="fl">5e4</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-55-1.png" width="768"></p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">colDataUnivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"moran.plot"</span>, <span class="va">qc_features</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/moranPlot.html">moranPlot</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"subsets_mito_percent"</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-57-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="spatial-autocorrelation-of-gene-expression">Spatial autocorrelation of gene expression<a class="anchor" aria-label="anchor" href="#spatial-autocorrelation-of-gene-expression"></a>
</h2>
<p>Normalize data with the <code>scran</code> method, and find highly
variable genes</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#clusters &lt;- quickCluster(sfe_tissue)</span></span>
<span><span class="co">#sfe_tissue &lt;- computeSumFactors(sfe_tissue, clusters=clusters)</span></span>
<span><span class="co">#sfe_tissue &lt;- sfe_tissue[, sizeFactors(sfe_tissue) &gt; 0]</span></span>
<span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span>
<span><span class="va">hvgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">dec</span>, n <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span></code></pre></div>
<p>Find Moran’s I for all highly variable genes:</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">runMoransI</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, features <span class="op">=</span> <span class="va">hvgs</span>, BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotColDataHistogram.html">plotRowDataHistogram</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"moran_sample01"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 30285 rows containing non-finite values (`stat_bin()`).</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-61-1.png" width="700"></p>
<p>The vast majority of genes have positive Moran’s I. Here we’ll find
the genes with the highest Moran’s I:</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_moran</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">moran_sample01</span>, </span>
<span>                                        decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">9</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p>We can use the <a href="https://pachterlab.github.io/gget/info.html" class="external-link">gget info</a> module
from the <a href="https://pachterlab.github.io/gget/" class="external-link">gget</a> package
to get additional information on these genes, such as their
descriptions, synonyms, transcripts and more from a collection of
reference databases including <a href="https://ensembl.org/" class="external-link">Ensembl</a>, <a href="https://www.uniprot.org/" class="external-link">UniProt</a> and <a href="https://www.ncbi.nlm.nih.gov/" class="external-link">NCBI</a> Here, we are showing their
gene descriptions from <a href="https://www.ncbi.nlm.nih.gov/" class="external-link">NCBI</a>:</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gget_info</span> <span class="op">&lt;-</span> <span class="va">gget</span><span class="op">$</span><span class="fu">info</span><span class="op">(</span><span class="va">top_moran</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">gget_info</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">gget_info</span><span class="op">$</span><span class="va">primary_gene_name</span></span>
<span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">gget_info</span>, <span class="va">ncbi_description</span><span class="op">)</span></span>
<span><span class="co">#&gt;                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         ncbi_description</span></span>
<span><span class="co">#&gt; S100a5                                                                                                                                                                                                                                                                                                                                                                                                                             Predicted to enable calcium-dependent protein binding activity; metal ion binding activity; and protein homodimerization activity. Located in neuronal cell body. Orthologous to human S100A5 (S100 calcium binding protein A5). [provided by Alliance of Genome Resources, Apr 2022]</span></span>
<span><span class="co">#&gt; Fabp7                                                                                                                                                                                                Predicted to enable fatty acid binding activity. Acts upstream of or within cell proliferation in forebrain; neurogenesis; and prepulse inhibition. Located in several cellular components, including cell projection; cell-cell junction; and neuronal cell body. Is expressed in several structures, including central nervous system; cranial nerve; gut; peripheral nervous system; and retina. Orthologous to human FABP7 (fatty acid binding protein 7). [provided by Alliance of Genome Resources, Apr 2022]</span></span>
<span><span class="co">#&gt; Apoe   This gene encodes a member of the apolipoprotein A1/A4/E family of proteins. This protein is involved in the transport of lipoproteins in the blood. It binds to a specific liver and peripheral cell receptor, and is essential for the normal catabolism of triglyceride-rich lipoprotein constituents. Homozygous knockout mice for this gene accumulate high levels of cholesterol in the blood and develop atherosclerosis. Different alleles of this gene have been associated with either increased risk or a protective effect for Alzheimer's disease in human patients. This gene maps to chromosome 7 in a cluster with the related apolipoprotein C1, C2 and C4 genes. [provided by RefSeq, Apr 2015]</span></span>
<span><span class="co">#&gt; Gng13                                                                                                                                                                                                                                                                                        Predicted to enable G-protein beta-subunit binding activity. Acts upstream of or within phospholipase C-activating G protein-coupled receptor signaling pathway and sensory perception of taste. Located in dendrite. Part of heterotrimeric G-protein complex. Is expressed in brain; gonad; gut; and liver. Orthologous to human GNG13 (G protein subunit gamma 13). [provided by Alliance of Genome Resources, Apr 2022]</span></span>
<span><span class="co">#&gt; Pcp4                                                      Enables calmodulin binding activity. Predicted to be involved in several processes, including calmodulin dependent kinase signaling pathway; negative regulation of protein kinase activity; and positive regulation of dopamine secretion. Predicted to be located in axon and neurofilament. Predicted to be part of protein-containing complex. Predicted to be active in cytoplasm. Is expressed in several structures, including alimentary system; central nervous system; genitourinary system; peripheral nervous system; and sensory organ. Orthologous to human PCP4 (Purkinje cell protein 4). [provided by Alliance of Genome Resources, Apr 2022]</span></span>
<span><span class="co">#&gt; Mtco2                                                                                                                                                                   Predicted to enable copper ion binding activity and oxidoreductase activity. Predicted to contribute to cytochrome-c oxidase activity. Predicted to be involved in ATP synthesis coupled electron transport; positive regulation of cellular biosynthetic process; and positive regulation of necrotic cell death. Located in mitochondrion. Is expressed in embryo; epiblast; heart; liver; and metanephros. Orthologous to several human genes including MTCO2P12 (MT-CO2 pseudogene 12). [provided by Alliance of Genome Resources, Apr 2022]</span></span>
<span><span class="co">#&gt; Ptgds                                                                                     Enables prostaglandin-D synthase activity and retinoid binding activity. Involved in prostaglandin biosynthetic process and regulation of circadian sleep/wake cycle, sleep. Acts upstream of or within negative regulation of male germ cell proliferation. Located in extracellular region. Is expressed in several structures, including alimentary system; genitourinary system; integumental system; nervous system; and sensory organ. Human ortholog(s) of this gene implicated in carotid artery disease. Orthologous to human PTGDS (prostaglandin D2 synthase). [provided by Alliance of Genome Resources, Apr 2022]</span></span>
<span><span class="co">#&gt; Mtnd4        Predicted to enable NADH dehydrogenase (ubiquinone) activity and ubiquinone binding activity. Predicted to contribute to NADH dehydrogenase activity. Predicted to be involved in several processes, including electron transport coupled proton transport; mitochondrial electron transport, NADH to ubiquinone; and mitochondrial respiratory chain complex I assembly. Located in mitochondrion. Human ortholog(s) of this gene implicated in Leber hereditary optic neuropathy; Parkinson's disease; macular degeneration; and schizophrenia. Orthologous to human MT-ND4 (mitochondrially encoded NADH:ubiquinone oxidoreductase core subunit 4). [provided by Alliance of Genome Resources, Apr 2022]</span></span></code></pre></div>
<p>Plot the genes with the highest Moran’s I:</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="va">top_moran</span>, ncol <span class="op">=</span> <span class="fl">3</span>, image_id <span class="op">=</span> <span class="st">"lowres"</span>,</span>
<span>                   maxcell <span class="op">=</span> <span class="fl">5e4</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-64-1.png" width="864"></p>
<p>Here global Moran’s I seems to be more about tissue structure.</p>
<p>Some genes have negative Moran’s I that might not be statistically
significant:</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">neg_moran</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">moran_sample01</span>, </span>
<span>                                        decreasing <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">9</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Display NCBI descriptions for these genes</span></span>
<span><span class="va">gget_info_neg</span> <span class="op">&lt;-</span> <span class="va">gget</span><span class="op">$</span><span class="fu">info</span><span class="op">(</span><span class="va">neg_moran</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">gget_info_neg</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">gget_info_neg</span><span class="op">$</span><span class="va">primary_gene_name</span></span>
<span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">gget_info_neg</span>, <span class="va">ncbi_description</span><span class="op">)</span></span>
<span><span class="co">#&gt;                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            ncbi_description</span></span>
<span><span class="co">#&gt; Hibch                                                                                                                                                                                                                                                                                                                                                                                                                                                                              Predicted to enable 3-hydroxyisobutyryl-CoA hydrolase activity. Predicted to be involved in valine catabolic process. Predicted to act upstream of or within branched-chain amino acid catabolic process. Located in mitochondrion. Orthologous to human HIBCH (3-hydroxyisobutyryl-CoA hydrolase). [provided by Alliance of Genome Resources, Apr 2022]</span></span>
<span><span class="co">#&gt; Syngr2                                                                                                                                                                                                                                                                                                                                                                                                                                           Predicted to be involved in regulated exocytosis and synaptic vesicle membrane organization. Predicted to act upstream of or within protein targeting. Located in synaptic vesicle. Is expressed in several structures, including genitourinary system; heart; liver; lung; and spleen. Orthologous to human SYNGR2 (synaptogyrin 2). [provided by Alliance of Genome Resources, Apr 2022]</span></span>
<span><span class="co">#&gt; Entpd5                                                                                                                                                                                                                                                                   Enables guanosine-diphosphatase activity and uridine-diphosphatase activity. Involved in several processes, including positive regulation of glycolytic process; protein N-linked glycosylation; and regulation of phosphatidylinositol 3-kinase signaling. Located in endoplasmic reticulum. Is expressed in several structures, including genitourinary system; gut; hemolymphoid system gland; liver; and nose. Orthologous to human ENTPD5 (ectonucleoside triphosphate diphosphohydrolase 5 (inactive)). [provided by Alliance of Genome Resources, Apr 2022]</span></span>
<span><span class="co">#&gt; Fyco1                                                                                                                                                                                                                                                            Predicted to enable metal ion binding activity. Predicted to be involved in plus-end-directed vesicle transport along microtubule and positive regulation of autophagosome maturation. Predicted to be located in Golgi apparatus. Predicted to be active in autophagosome; late endosome; and lysosome. Is expressed in central nervous system; early conceptus; and retina. Human ortholog(s) of this gene implicated in cataract 18. Orthologous to human FYCO1 (FYVE and coiled-coil domain autophagy adaptor 1). [provided by Alliance of Genome Resources, Apr 2022]</span></span>
<span><span class="co">#&gt; Ptpn18                                                                                                                                                                                                                                                                                                                                                                                             Enables non-membrane spanning protein tyrosine phosphatase activity. Acts upstream of or within blastocyst formation. Located in cytoplasm and nucleus. Is expressed in several structures, including alimentary system; brain; genitourinary system; immune system; and liver and biliary system. Orthologous to human PTPN18 (protein tyrosine phosphatase non-receptor type 18). [provided by Alliance of Genome Resources, Apr 2022]</span></span>
<span><span class="co">#&gt; Cbl                                                                                                                                                                                           Enables SH3 domain binding activity and ephrin receptor binding activity. Involved in regulation of platelet-derived growth factor receptor-alpha signaling pathway. Acts upstream of or within regulation of Rap protein signal transduction. Located in Golgi apparatus and cilium. Part of flotillin complex. Is expressed in male reproductive system and urinary system. Human ortholog(s) of this gene implicated in acute myeloid leukemia; juvenile myelomonocytic leukemia; lung non-small cell carcinoma; and myeloid neoplasm. Orthologous to human CBL (Cbl proto-oncogene). [provided by Alliance of Genome Resources, Apr 2022]</span></span>
<span><span class="co">#&gt; Dusp18  Predicted to enable protein tyrosine phosphatase activity and protein tyrosine/serine/threonine phosphatase activity. Predicted to be involved in peptidyl-threonine dephosphorylation and peptidyl-tyrosine dephosphorylation. Predicted to act upstream of or within protein targeting to membrane; protein targeting to mitochondrion; and response to antibiotic. Predicted to be located in mitochondrial inner membrane; mitochondrial intermembrane space; and nucleoplasm. Predicted to be extrinsic component of mitochondrial inner membrane and intrinsic component of mitochondrial inner membrane. Is expressed in central nervous system; dorsal root ganglion; olfactory epithelium; and retina. Orthologous to human DUSP18 (dual specificity phosphatase 18). [provided by Alliance of Genome Resources, Apr 2022]</span></span>
<span><span class="co">#&gt; Tsc22d3                                                                                                                                                                                                                                                        Enables MRF binding activity. Acts upstream of or within several processes, including negative regulation of activation-induced cell death of T cells; negative regulation of skeletal muscle tissue development; and negative regulation of transcription by RNA polymerase II. Located in cytoplasm and nucleus. Is expressed in several structures, including early conceptus; genitourinary system; nervous system; sensory organ; and viscerocranium. Orthologous to human TSC22D3 (TSC22 domain family member 3). [provided by Alliance of Genome Resources, Apr 2022]</span></span>
<span><span class="co">#&gt; Plekhg2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   Predicted to enable guanyl-nucleotide exchange factor activity. Predicted to be involved in regulation of actin filament polymerization. Orthologous to human PLEKHG2 (pleckstrin homology and RhoGEF domain containing G2). [provided by Alliance of Genome Resources, Apr 2022]</span></span></code></pre></div>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="va">neg_moran</span>, ncol <span class="op">=</span> <span class="fl">3</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span>,</span>
<span>                   image_id <span class="op">=</span> <span class="st">"lowres"</span>, maxcell <span class="op">=</span> <span class="fl">5e4</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-67-1.png" width="864"></p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">runUnivariate</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"moran.mc"</span>, <span class="va">neg_moran</span>, </span>
<span>                            colGraphName <span class="op">=</span> <span class="st">"visium"</span>, nsim <span class="op">=</span> <span class="fl">200</span>, alternative <span class="op">=</span> <span class="st">"less"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotMoranMC.html">plotMoranMC</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="va">neg_moran</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-69-1.png" width="700"></p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">[</span><span class="va">neg_moran</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"moran_sample01"</span>, <span class="st">"moran.mc_p.value_sample01"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">#&gt; DataFrame with 9 rows and 2 columns</span></span>
<span><span class="co">#&gt;                    moran_sample01 moran.mc_p.value_sample01</span></span>
<span><span class="co">#&gt;                         &lt;numeric&gt;                 &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; ENSMUSG00000041426     -0.0531915                0.00995025</span></span>
<span><span class="co">#&gt; ENSMUSG00000048277     -0.0451179                0.00497512</span></span>
<span><span class="co">#&gt; ENSMUSG00000021236     -0.0445148                0.00497512</span></span>
<span><span class="co">#&gt; ENSMUSG00000025241     -0.0419121                0.00497512</span></span>
<span><span class="co">#&gt; ENSMUSG00000026126     -0.0399917                0.01492537</span></span>
<span><span class="co">#&gt; ENSMUSG00000034342     -0.0393964                0.01492537</span></span>
<span><span class="co">#&gt; ENSMUSG00000047205     -0.0381599                0.00995025</span></span>
<span><span class="co">#&gt; ENSMUSG00000031431     -0.0369456                0.01492537</span></span>
<span><span class="co">#&gt; ENSMUSG00000037552     -0.0368969                0.01492537</span></span></code></pre></div>
<p>As there are 2000 highly variable genes and 2000 tests, these would
no longer be significant after correcting for multiple testing.</p>
<p>Does global Moran’s I relate to gene expression level?</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/addPerCellQCMetrics.html" class="external-link">addPerFeatureQCMetrics</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "symbol"                                       </span></span>
<span><span class="co">#&gt;  [2] "Feature.Type"                                 </span></span>
<span><span class="co">#&gt;  [3] "I_sample01"                                   </span></span>
<span><span class="co">#&gt;  [4] "P.value_sample01"                             </span></span>
<span><span class="co">#&gt;  [5] "Adjusted.p.value_sample01"                    </span></span>
<span><span class="co">#&gt;  [6] "Feature.Counts.in.Spots.Under.Tissue_sample01"</span></span>
<span><span class="co">#&gt;  [7] "Median.Normalized.Average.Counts_sample01"    </span></span>
<span><span class="co">#&gt;  [8] "Barcodes.Detected.per.Feature_sample01"       </span></span>
<span><span class="co">#&gt;  [9] "moran_sample01"                               </span></span>
<span><span class="co">#&gt; [10] "K_sample01"                                   </span></span>
<span><span class="co">#&gt; [11] "moran.mc_statistic_sample01"                  </span></span>
<span><span class="co">#&gt; [12] "moran.mc_parameter_sample01"                  </span></span>
<span><span class="co">#&gt; [13] "moran.mc_p.value_sample01"                    </span></span>
<span><span class="co">#&gt; [14] "moran.mc_alternative_sample01"                </span></span>
<span><span class="co">#&gt; [15] "moran.mc_method_sample01"                     </span></span>
<span><span class="co">#&gt; [16] "moran.mc_res_sample01"                        </span></span>
<span><span class="co">#&gt; [17] "mean"                                         </span></span>
<span><span class="co">#&gt; [18] "detected"</span></span></code></pre></div>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotRowData.html" class="external-link">plotRowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, x <span class="op">=</span> <span class="st">"mean"</span>, y <span class="op">=</span> <span class="st">"moran_sample01"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotation_logticks.html" class="external-link">annotation_logticks</a></span><span class="op">(</span>sides <span class="op">=</span> <span class="st">"b"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density_2d.html" class="external-link">geom_density2d</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Transformation introduced infinite values in continuous x-axis</span></span>
<span><span class="co">#&gt; Transformation introduced infinite values in continuous x-axis</span></span>
<span><span class="co">#&gt; Warning: Removed 30285 rows containing non-finite values</span></span>
<span><span class="co">#&gt; (`stat_density2d()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 30285 rows containing missing values (`geom_point()`).</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-72-1.png" width="700"></p>
<p>Genes that are more highly expressed overall tend to have higher
Moran’s I.</p>
</div>
<div class="section level2">
<h2 id="apply-spatial-analysis-methods-to-gene-expression-space">Apply spatial analysis methods to gene expression space<a class="anchor" aria-label="anchor" href="#apply-spatial-analysis-methods-to-gene-expression-space"></a>
</h2>
<p>Spatial statistics that require a spatial neighborhood graph can also
be applied to the k nearest neighbor graph not in histological space but
in gene expression space. This is done in more depth in <a href="https://pachterlab.github.io/voyager/articles/nonspatial.html">this
vignette</a>.</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, ncomponents <span class="op">=</span> <span class="fl">30</span>, subset_row <span class="op">=</span> <span class="va">hvgs</span>,</span>
<span>                     scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># scale as in Seurat</span></span></code></pre></div>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">foo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocNeighbors/man/findKNN-methods.html" class="external-link">findKNN</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"PCA"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>, k<span class="op">=</span><span class="fl">10</span>, BNPARAM<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocNeighbors/man/AnnoyParam.html" class="external-link">AnnoyParam</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Split by row</span></span>
<span><span class="va">foo_nb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/asplit.html" class="external-link">asplit</a></span><span class="op">(</span><span class="va">foo</span><span class="op">$</span><span class="va">index</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">dmat</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">/</span><span class="va">foo</span><span class="op">$</span><span class="va">distance</span></span>
<span><span class="co"># Row normalize the weights</span></span>
<span><span class="va">dmat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sweep.html" class="external-link">sweep</a></span><span class="op">(</span><span class="va">dmat</span>, <span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">dmat</span><span class="op">)</span>, FUN <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span></span>
<span><span class="va">glist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/asplit.html" class="external-link">asplit</a></span><span class="op">(</span><span class="va">dmat</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co"># Sort based on index</span></span>
<span><span class="va">ord</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">foo_nb</span>, <span class="va">order</span><span class="op">)</span></span>
<span><span class="va">foo_nb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">foo_nb</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="va">foo_nb</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">ord</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">foo_nb</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"nb"</span></span>
<span><span class="va">glist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">glist</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="va">glist</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">ord</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">listw</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>style <span class="op">=</span> <span class="st">"W"</span>,</span>
<span>              neighbours <span class="op">=</span> <span class="va">foo_nb</span>,</span>
<span>              weights <span class="op">=</span> <span class="va">glist</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">listw</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"listw"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">listw</span>, <span class="st">"region.id"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"knn10"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">listw</span></span></code></pre></div>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">runMoransI</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, features <span class="op">=</span> <span class="va">hvgs</span>, BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>,</span>
<span>                         colGraphName <span class="op">=</span> <span class="st">"knn10"</span>, name <span class="op">=</span> <span class="st">"moran_ns"</span><span class="op">)</span></span></code></pre></div>
<p>Here we store the results in “moran_ns”, not to be confused with
spatial Moran’s I results.</p>
<p>These are the genes that tend to be more similar to their neighbors
in the 10 nearest neighbor graph in PCA space for gene expression rather
than in histological space:</p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_moran2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">moran_ns_sample01</span>, </span>
<span>                                        decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">9</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Display NCBI descriptions for these genes</span></span>
<span><span class="va">gget_info2</span> <span class="op">&lt;-</span> <span class="va">gget</span><span class="op">$</span><span class="fu">info</span><span class="op">(</span><span class="va">top_moran2</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">gget_info2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">gget_info2</span><span class="op">$</span><span class="va">primary_gene_name</span></span>
<span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">gget_info2</span>, <span class="va">ncbi_description</span><span class="op">)</span></span>
<span><span class="co">#&gt;                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        ncbi_description</span></span>
<span><span class="co">#&gt; Mtco1                                                                                                                                                                                                                                                                                                                                                                                                                                                              Enables cytochrome-c oxidase activity. Predicted to be involved in electron transport coupled proton transport; mitochondrial electron transport, cytochrome c to oxygen; and response to oxidative stress. Located in mitochondrial inner membrane. Part of mitochondrial respiratory chain complex IV. Is expressed in several structures, including brown fat; heart; liver; metanephros; and skeletal muscle. Orthologous to human MT-CO1 (mitochondrially encoded cytochrome c oxidase I). [provided by Alliance of Genome Resources, Apr 2022]</span></span>
<span><span class="co">#&gt; Mtco2                                                                                                                                                                                                                                                                                                                                                                                                                                                                  Predicted to enable copper ion binding activity and oxidoreductase activity. Predicted to contribute to cytochrome-c oxidase activity. Predicted to be involved in ATP synthesis coupled electron transport; positive regulation of cellular biosynthetic process; and positive regulation of necrotic cell death. Located in mitochondrion. Is expressed in embryo; epiblast; heart; liver; and metanephros. Orthologous to several human genes including MTCO2P12 (MT-CO2 pseudogene 12). [provided by Alliance of Genome Resources, Apr 2022]</span></span>
<span><span class="co">#&gt; Apoe                                                                                                                                                                                                                                                                                                  This gene encodes a member of the apolipoprotein A1/A4/E family of proteins. This protein is involved in the transport of lipoproteins in the blood. It binds to a specific liver and peripheral cell receptor, and is essential for the normal catabolism of triglyceride-rich lipoprotein constituents. Homozygous knockout mice for this gene accumulate high levels of cholesterol in the blood and develop atherosclerosis. Different alleles of this gene have been associated with either increased risk or a protective effect for Alzheimer's disease in human patients. This gene maps to chromosome 7 in a cluster with the related apolipoprotein C1, C2 and C4 genes. [provided by RefSeq, Apr 2015]</span></span>
<span><span class="co">#&gt; Mtnd4                                                                                                                                                                                                                                                                                                       Predicted to enable NADH dehydrogenase (ubiquinone) activity and ubiquinone binding activity. Predicted to contribute to NADH dehydrogenase activity. Predicted to be involved in several processes, including electron transport coupled proton transport; mitochondrial electron transport, NADH to ubiquinone; and mitochondrial respiratory chain complex I assembly. Located in mitochondrion. Human ortholog(s) of this gene implicated in Leber hereditary optic neuropathy; Parkinson's disease; macular degeneration; and schizophrenia. Orthologous to human MT-ND4 (mitochondrially encoded NADH:ubiquinone oxidoreductase core subunit 4). [provided by Alliance of Genome Resources, Apr 2022]</span></span>
<span><span class="co">#&gt; Fabp7                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               Predicted to enable fatty acid binding activity. Acts upstream of or within cell proliferation in forebrain; neurogenesis; and prepulse inhibition. Located in several cellular components, including cell projection; cell-cell junction; and neuronal cell body. Is expressed in several structures, including central nervous system; cranial nerve; gut; peripheral nervous system; and retina. Orthologous to human FABP7 (fatty acid binding protein 7). [provided by Alliance of Genome Resources, Apr 2022]</span></span>
<span><span class="co">#&gt; mt-Nd2                                                                                                                                                                                                                                                                                                                                                  Predicted to enable NADH dehydrogenase (ubiquinone) activity; ionotropic glutamate receptor binding activity; and protein kinase binding activity. Acts upstream of or within reactive oxygen species metabolic process. Located in mitochondrion. Is expressed in early conceptus and secondary oocyte. Human ortholog(s) of this gene implicated in Leber hereditary optic neuropathy; multiple sclerosis; myocardial infarction; neurodegenerative disease (multiple); and urinary bladder cancer. Orthologous to human MT-ND2 (mitochondrially encoded NADH:ubiquinone oxidoreductase core subunit 2). [provided by Alliance of Genome Resources, Apr 2022]</span></span>
<span><span class="co">#&gt; Ptn                                                                                                                                                                                                                                                                                                                                   Predicted to enable several functions, including glycosaminoglycan binding activity; signaling receptor binding activity; and syndecan binding activity. Involved in several processes, including decidualization; learning or memory; and regulation of nervous system development. Acts upstream of or within bone mineralization. Located in extracellular region. Is expressed in several structures, including alimentary system; genitourinary system; nervous system; respiratory system; and sensory organ. Human ortholog(s) of this gene implicated in adrenal carcinoma. Orthologous to human PTN (pleiotrophin). [provided by Alliance of Genome Resources, Apr 2022]</span></span>
<span><span class="co">#&gt; Apod                                                                                                                                                                                                                                                                                                                                                                               The protein encoded by this gene is a component of high-density lipoprotein (HDL), but is unique in that it shares greater structural similarity to lipocalin than to other members of the apolipoprotein family, and has a wider tissue expression pattern. The encoded protein is involved in lipid metabolism, and ablation of this gene results in defects in triglyceride metabolism. Elevated levels of this gene product have been observed in multiple tissues of Niemann-Pick disease mouse models, as well as in some tumors. Alternative splicing results in multiple transcript variants. [provided by RefSeq, Aug 2014]</span></span>
<span><span class="co">#&gt; Atp1a2 Predicted to enable several functions, including ATP binding activity; ATP hydrolysis activity; and alkali metal ion binding activity. Involved in several processes, including cellular response to steroid hormone stimulus; locomotory exploration behavior; and response to auditory stimulus. Acts upstream of or within several processes, including forebrain development; regulation of blood circulation; and regulation of muscle contraction. Located in T-tubule; cell projection; and neuronal cell body. Is expressed in several structures, including genitourinary system; heart; musculature; nervous system; and sensory organ. Used to study familial hemiplegic migraine 2. Human ortholog(s) of this gene implicated in alternating hemiplegia of childhood; benign neonatal seizures; familial hemiplegic migraine 2; hypertension; and migraine with aura. Orthologous to human ATP1A2 (ATPase Na+/K+ transporting subunit alpha 2). [provided by Alliance of Genome Resources, Apr 2022]</span></span></code></pre></div>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="va">top_moran2</span>, ncol <span class="op">=</span> <span class="fl">3</span>, swap_rownames <span class="op">=</span> <span class="st">"symbol"</span>,</span>
<span>                   image_id <span class="op">=</span> <span class="st">"lowres"</span>, maxcell <span class="op">=</span> <span class="fl">5e4</span><span class="op">)</span></span></code></pre></div>
<p><img src="visium_10x_spatial_files/figure-html/unnamed-chunk-79-1.png" width="864"></p>
<p>Although this Moran’s I was not computed in histological space, these
genes with the highest Moran’s I in PCA space also show spatial
structure, as different cell types reside in different spatial
regions.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">#&gt; Platform: x86_64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS Monterey 12.6.5</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] reticulate_1.28                BiocNeighbors_1.18.0          </span></span>
<span><span class="co">#&gt;  [3] BiocParallel_1.34.1            dplyr_1.1.2                   </span></span>
<span><span class="co">#&gt;  [5] rmapshaper_0.5.0               sf_1.0-12                     </span></span>
<span><span class="co">#&gt;  [7] rlang_1.1.1                    terra_1.7-29                  </span></span>
<span><span class="co">#&gt;  [9] EBImage_4.42.0                 rjson_0.2.21                  </span></span>
<span><span class="co">#&gt; [11] bluster_1.10.0                 patchwork_1.1.2               </span></span>
<span><span class="co">#&gt; [13] stringr_1.5.0                  scran_1.28.1                  </span></span>
<span><span class="co">#&gt; [15] scater_1.28.0                  scuttle_1.10.1                </span></span>
<span><span class="co">#&gt; [17] ggplot2_3.4.2                  SingleCellExperiment_1.22.0   </span></span>
<span><span class="co">#&gt; [19] SummarizedExperiment_1.30.1    Biobase_2.60.0                </span></span>
<span><span class="co">#&gt; [21] GenomicRanges_1.52.0           GenomeInfoDb_1.36.0           </span></span>
<span><span class="co">#&gt; [23] IRanges_2.34.0                 S4Vectors_0.38.1              </span></span>
<span><span class="co">#&gt; [25] BiocGenerics_0.46.0            MatrixGenerics_1.12.0         </span></span>
<span><span class="co">#&gt; [27] matrixStats_0.63.0             SpatialFeatureExperiment_1.2.1</span></span>
<span><span class="co">#&gt; [29] Voyager_1.2.3                 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] RColorBrewer_1.1-3        jsonlite_1.8.4           </span></span>
<span><span class="co">#&gt;   [3] wk_0.7.3                  magrittr_2.0.3           </span></span>
<span><span class="co">#&gt;   [5] ggbeeswarm_0.7.2          magick_2.7.4             </span></span>
<span><span class="co">#&gt;   [7] farver_2.1.1              rmarkdown_2.21           </span></span>
<span><span class="co">#&gt;   [9] fs_1.6.2                  zlibbioc_1.46.0          </span></span>
<span><span class="co">#&gt;  [11] ragg_1.2.5                vctrs_0.6.2              </span></span>
<span><span class="co">#&gt;  [13] spdep_1.2-8               memoise_2.0.1            </span></span>
<span><span class="co">#&gt;  [15] DelayedMatrixStats_1.22.0 RCurl_1.98-1.12          </span></span>
<span><span class="co">#&gt;  [17] htmltools_0.5.5           S4Arrays_1.0.4           </span></span>
<span><span class="co">#&gt;  [19] curl_5.0.0                Rhdf5lib_1.22.0          </span></span>
<span><span class="co">#&gt;  [21] s2_1.1.4                  rhdf5_2.44.0             </span></span>
<span><span class="co">#&gt;  [23] sass_0.4.6                spData_2.2.2             </span></span>
<span><span class="co">#&gt;  [25] KernSmooth_2.23-21        bslib_0.4.2              </span></span>
<span><span class="co">#&gt;  [27] htmlwidgets_1.6.2         desc_1.4.2               </span></span>
<span><span class="co">#&gt;  [29] cachem_1.0.8              igraph_1.4.2             </span></span>
<span><span class="co">#&gt;  [31] lifecycle_1.0.3           pkgconfig_2.0.3          </span></span>
<span><span class="co">#&gt;  [33] rsvd_1.0.5                Matrix_1.5-4             </span></span>
<span><span class="co">#&gt;  [35] R6_2.5.1                  fastmap_1.1.1            </span></span>
<span><span class="co">#&gt;  [37] GenomeInfoDbData_1.2.10   digest_0.6.31            </span></span>
<span><span class="co">#&gt;  [39] colorspace_2.1-0          ggnewscale_0.4.8         </span></span>
<span><span class="co">#&gt;  [41] rprojroot_2.0.3           dqrng_0.3.0              </span></span>
<span><span class="co">#&gt;  [43] RSpectra_0.16-1           irlba_2.3.5.1            </span></span>
<span><span class="co">#&gt;  [45] textshaping_0.3.6         beachmat_2.16.0          </span></span>
<span><span class="co">#&gt;  [47] labeling_0.4.2            fansi_1.0.4              </span></span>
<span><span class="co">#&gt;  [49] mgcv_1.8-42               abind_1.4-5              </span></span>
<span><span class="co">#&gt;  [51] compiler_4.3.0            here_1.0.1               </span></span>
<span><span class="co">#&gt;  [53] proxy_0.4-27              withr_2.5.0              </span></span>
<span><span class="co">#&gt;  [55] tiff_0.1-11               viridis_0.6.3            </span></span>
<span><span class="co">#&gt;  [57] DBI_1.1.3                 highr_0.10               </span></span>
<span><span class="co">#&gt;  [59] HDF5Array_1.28.1          R.utils_2.12.2           </span></span>
<span><span class="co">#&gt;  [61] MASS_7.3-60               DelayedArray_0.26.2      </span></span>
<span><span class="co">#&gt;  [63] classInt_0.4-9            tools_4.3.0              </span></span>
<span><span class="co">#&gt;  [65] units_0.8-2               vipor_0.4.5              </span></span>
<span><span class="co">#&gt;  [67] beeswarm_0.4.0            R.oo_1.25.0              </span></span>
<span><span class="co">#&gt;  [69] glue_1.6.2                dbscan_1.1-11            </span></span>
<span><span class="co">#&gt;  [71] nlme_3.1-162              rhdf5filters_1.12.1      </span></span>
<span><span class="co">#&gt;  [73] grid_4.3.0                cluster_2.1.4            </span></span>
<span><span class="co">#&gt;  [75] generics_0.1.3            isoband_0.2.7            </span></span>
<span><span class="co">#&gt;  [77] gtable_0.3.3              R.methodsS3_1.8.2        </span></span>
<span><span class="co">#&gt;  [79] class_7.3-22              metapod_1.8.0            </span></span>
<span><span class="co">#&gt;  [81] BiocSingular_1.16.0       ScaledMatrix_1.8.1       </span></span>
<span><span class="co">#&gt;  [83] sp_1.6-0                  utf8_1.2.3               </span></span>
<span><span class="co">#&gt;  [85] XVector_0.40.0            ggrepel_0.9.3            </span></span>
<span><span class="co">#&gt;  [87] pillar_1.9.0              limma_3.56.1             </span></span>
<span><span class="co">#&gt;  [89] splines_4.3.0             lattice_0.21-8           </span></span>
<span><span class="co">#&gt;  [91] deldir_1.0-9              tidyselect_1.2.0         </span></span>
<span><span class="co">#&gt;  [93] locfit_1.5-9.7            knitr_1.42               </span></span>
<span><span class="co">#&gt;  [95] gridExtra_2.3             V8_4.3.0                 </span></span>
<span><span class="co">#&gt;  [97] edgeR_3.42.2              xfun_0.39                </span></span>
<span><span class="co">#&gt;  [99] statmod_1.5.0             DropletUtils_1.20.0      </span></span>
<span><span class="co">#&gt; [101] fftwtools_0.9-11          stringi_1.7.12           </span></span>
<span><span class="co">#&gt; [103] geojsonsf_2.0.3           yaml_2.3.7               </span></span>
<span><span class="co">#&gt; [105] boot_1.3-28.1             evaluate_0.21            </span></span>
<span><span class="co">#&gt; [107] codetools_0.2-19          tibble_3.2.1             </span></span>
<span><span class="co">#&gt; [109] cli_3.6.1                 systemfonts_1.0.4        </span></span>
<span><span class="co">#&gt; [111] munsell_0.5.0             jquerylib_0.1.4          </span></span>
<span><span class="co">#&gt; [113] Rcpp_1.0.10               png_0.1-8                </span></span>
<span><span class="co">#&gt; [115] parallel_4.3.0            pkgdown_2.0.7            </span></span>
<span><span class="co">#&gt; [117] jpeg_0.1-10               sparseMatrixStats_1.12.0 </span></span>
<span><span class="co">#&gt; [119] bitops_1.0-7              SpatialExperiment_1.10.0 </span></span>
<span><span class="co">#&gt; [121] viridisLite_0.4.2         scales_1.2.1             </span></span>
<span><span class="co">#&gt; [123] e1071_1.7-13              purrr_1.0.1              </span></span>
<span><span class="co">#&gt; [125] crayon_1.5.2              scico_1.3.1              </span></span>
<span><span class="co">#&gt; [127] cowplot_1.1.1</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Lambda Moses, Kayla Jackson, Laura Luebbert, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
