<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>MERFISH Processing Workflows with Voyager • Voyager</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="MERFISH Processing Workflows with Voyager">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Voyager</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.8.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/install.html">Install</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-technologies" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Technologies</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-technologies">
<li><a class="dropdown-item" href="../articles/visium_landing.html">Visium</a></li>
    <li><a class="dropdown-item" href="../articles/slideseqV2_landing.html">Slide-seq v2</a></li>
    <li><a class="dropdown-item" href="../articles/cosmx_landing.html">CosMX</a></li>
    <li><a class="dropdown-item" href="../articles/xenium_landing.html">Xenium</a></li>
    <li><a class="dropdown-item" href="../articles/merfish_landing.html">MERFISH</a></li>
    <li><a class="dropdown-item" href="../articles/seqfish_landing.html">seqFISH</a></li>
    <li><a class="dropdown-item" href="../articles/codex_landing.html">CODEX/PhenoCycler</a></li>
    <li><a class="dropdown-item" href="../articles/splitseq_landing.html">SPLiT-seq</a></li>
    <li><a class="dropdown-item" href="../articles/chromium_landing.html">10X Chromium V3</a></li>
    <li><a class="dropdown-item" href="../articles/10xnuclei_landing.html">10X Chromium Nuclei Isolation</a></li>
    <li><a class="dropdown-item" href="../articles/10xcrispr_landing.html">10X Chromium Single Cell CRISPR</a></li>
    <li><a class="dropdown-item" href="../articles/10xatac_landing.html">10X Single Cell ATAC</a></li>
    <li><a class="dropdown-item" href="../articles/10xmultiome_landing.html">10X Single Cell Multiome</a></li>
  </ul>
</li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-methods" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Methods</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-methods">
<li><h6 class="dropdown-header" data-toc-skip>Univariate global</h6></li>
    <li><a class="dropdown-item" href="../articles/moran_landing.html">Moran's I</a></li>
    <li><a class="dropdown-item" href="../articles/geary_landing.html">Geary's C</a></li>
    <li><a class="dropdown-item" href="../articles/correlogram_landing.html">Correlogram</a></li>
    <li><a class="dropdown-item" href="../articles/variogram_landing.html">Variogram</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Univariate local</h6></li>
    <li><a class="dropdown-item" href="../articles/moranplot_landing.html">Moran scatter plot</a></li>
    <li><a class="dropdown-item" href="../articles/localmoran_landing.html">Local Moran's I</a></li>
    <li><a class="dropdown-item" href="../articles/losh_landing.html">Local spatial heteroscedasticity</a></li>
    <li><a class="dropdown-item" href="../articles/getisord_landing.html">Getis-Ord Gi*</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bivariate</h6></li>
    <li><a class="dropdown-item" href="../articles/lee_landing.html">Lee's L</a></li>
    <li><a class="dropdown-item" href="../articles/crossvariogram_landing.html">Cross variogram</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Multivariate</h6></li>
    <li><a class="dropdown-item" href="../articles/multispati_landing.html">MULTISPATI PCA</a></li>
    <li><a class="dropdown-item" href="../articles/localc_landing.html">Multivariate local Geary's C</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/sfemethod">Custom methods</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/pachterlab/voyager/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>MERFISH Processing Workflows with Voyager</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/pachterlab/voyager/blob/documentation/vignettes/merfish_landing.Rmd" class="external-link"><code>vignettes/merfish_landing.Rmd</code></a></small>
      <div class="d-none name"><code>merfish_landing.Rmd</code></div>
    </div>

    
    
<p>The MERSCOPE Platform from <a href="https://vizgen.com/products/" class="external-link">Vizgen</a> is enabled by Multiplexed
Error-Robust Fluorescence in-situ Hybridization (MERFISH). MERFISH is a
spatial transcriptomics technology that expands the multiplexing
capabilities of single molecule fluorescent in-situ hybridization
(smFISH) via a combinatorial barcode coupled with sequential rounds of
imaging. The technology enables detection of hundreds to thousands of
single molecule targets at sub-cellular resolution. While MERFISH was
designed to be compatible with fresh or fixed frozen tissues, Vizgen has
demonstrated efficacy of the technology on FFPE tissue specimens.</p>
<div class="section level2">
<h2 id="pros-and-cons">Pros and cons<a class="anchor" aria-label="anchor" href="#pros-and-cons"></a>
</h2>
<p>Pros:</p>
<ul>
<li>Commercial kit</li>
<li>Single cell resolution</li>
<li>High detection efficiency</li>
<li>Formalin fixed, paraffin embedded (FFPE) tissue compatible</li>
<li>Provides subcellular transcript localization information</li>
<li>Compatible with histological staining including DAPI</li>
<li>Protein co-detection is supported</li>
</ul>
<p>Cons:</p>
<ul>
<li>A curated panel of usually a few hundred genes is required. However,
Vizgen provides curated gene panels for neuroscience and oncology, as
well as panel design services.</li>
<li>Data size is harder to manage for larger tissue areas and number of
samples. Not all spatial analysis methods can scale to hundreds of
thousands to millions of cells.</li>
</ul>
</div>
<div class="section level2">
<h2 id="getting-started">Getting Started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<p>Several MERFISH datasets have been generated on the MERSCOPE Platform
and are publicly available from Vizgen on their <a href="https://vizgen.com/data-release-program/" class="external-link">website</a>. We provide
examples available for processing the output of various spatial
transcriptomics technologies into a <a href="https://bioconductor.org/packages/devel/bioc/html/SpatialFeatureExperiment.html" class="external-link"><code>SpatialFeatureExperiment</code></a>(SFE)
object for use with <code>Voyager</code> in the vignette below.</p>
<p>The vignette provides technology specific notes for data downloaded
from Vizgen. Briefly, Vizgen provides cell metadata and the gene count
matrix as CSV files which can be read in quickly with the <a href="https://cran.r-project.org/web/packages/vroom/index.html" class="external-link"><code>vroom</code></a>
package. Cell segmentation data are provided as HDF5 files that are
delineated by field of view.</p>
<table class="table">
<colgroup>
<col width="28%">
<col width="24%">
<col width="47%">
</colgroup>
<thead><tr class="header">
<th>Vignette</th>
<th align="center">Colab Notebook</th>
<th>Description</th>
</tr></thead>
<tbody><tr class="odd">
<td><a href="https://pachterlab.github.io/voyager/articles/create_sfe.html">Create
a <code>SFE</code> object</a></td>
<td align="center"><a href="https://colab.research.google.com/github/pachterlab/voyager/blob/documentation/vignettes/create_sfe.ipynb" class="external-link">Colab
Notebook</a></td>
<td>Download data, create <code>SFE</code> object, Vizgen/MERFISH
specific notes</td>
</tr></tbody>
</table>
</div>
<div class="section level2">
<h2 id="analysis-workflows">Analysis Workflows<a class="anchor" aria-label="anchor" href="#analysis-workflows"></a>
</h2>
<p>The vignettes below demonstrate workflows that can be implemented
with <code>Voyager</code> using data generated with the MERFISH
technology. The publicly available MERFISH datasets profile hundreds of
genes in hundreds of thousands to millions of cells. Thus, the vignettes
linked below can provide some context on the capabilities of
<code>Voyager</code> for moderate to large datasets.</p>
<table class="table">
<colgroup>
<col width="28%">
<col width="24%">
<col width="47%">
</colgroup>
<thead><tr class="header">
<th>Vignette</th>
<th align="center">Colab Notebook</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><a href="https://pachterlab.github.io/voyager/articles/vig6_merfish.html">MERFISH
mouse liver analysis</a></td>
<td align="center"><a href="https://colab.research.google.com/github/pachterlab/voyager/blob/documentation/vignettes/vig6_merfish.ipynb" class="external-link">Colab
Notebook</a></td>
<td>Perform basic QC, compute Moran’s I, length scales of spatial
autocorrelation, binning, variogram models, PCA for large datasets</td>
</tr>
<tr class="even">
<td><a href="https://pachterlab.github.io/voyager/articles/multispati.html">MULTISPATI
PCA and negative spatial autocorrelation</a></td>
<td align="center"><a href="https://colab.research.google.com/github/pachterlab/voyager/blob/documentation/vignettes/multispati.ipynb" class="external-link">Colab
Notebook</a></td>
<td>Run MULTISPATI PCA on MERFISH mouse liver dataset, and compare the
results to those from non-spatial PCA</td>
</tr>
</tbody>
</table>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Lambda Moses, Alik Huseynov, Kayla Jackson, Laura Luebbert, Sina Booeshaghi, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
