<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>CODEX exploratory data analysis • Voyager</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="CODEX exploratory data analysis">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Voyager</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.8.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/install.html">Install</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-technologies" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Technologies</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-technologies">
<li><a class="dropdown-item" href="../articles/visium_landing.html">Visium</a></li>
    <li><a class="dropdown-item" href="../articles/slideseqV2_landing.html">Slide-seq v2</a></li>
    <li><a class="dropdown-item" href="../articles/cosmx_landing.html">CosMX</a></li>
    <li><a class="dropdown-item" href="../articles/xenium_landing.html">Xenium</a></li>
    <li><a class="dropdown-item" href="../articles/merfish_landing.html">MERFISH</a></li>
    <li><a class="dropdown-item" href="../articles/seqfish_landing.html">seqFISH</a></li>
    <li><a class="dropdown-item" href="../articles/codex_landing.html">CODEX/PhenoCycler</a></li>
    <li><a class="dropdown-item" href="../articles/splitseq_landing.html">SPLiT-seq</a></li>
    <li><a class="dropdown-item" href="../articles/chromium_landing.html">10X Chromium V3</a></li>
    <li><a class="dropdown-item" href="../articles/10xnuclei_landing.html">10X Chromium Nuclei Isolation</a></li>
    <li><a class="dropdown-item" href="../articles/10xcrispr_landing.html">10X Chromium Single Cell CRISPR</a></li>
    <li><a class="dropdown-item" href="../articles/10xatac_landing.html">10X Single Cell ATAC</a></li>
    <li><a class="dropdown-item" href="../articles/10xmultiome_landing.html">10X Single Cell Multiome</a></li>
  </ul>
</li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-methods" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Methods</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-methods">
<li><h6 class="dropdown-header" data-toc-skip>Univariate global</h6></li>
    <li><a class="dropdown-item" href="../articles/moran_landing.html">Moran's I</a></li>
    <li><a class="dropdown-item" href="../articles/geary_landing.html">Geary's C</a></li>
    <li><a class="dropdown-item" href="../articles/correlogram_landing.html">Correlogram</a></li>
    <li><a class="dropdown-item" href="../articles/variogram_landing.html">Variogram</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Univariate local</h6></li>
    <li><a class="dropdown-item" href="../articles/moranplot_landing.html">Moran scatter plot</a></li>
    <li><a class="dropdown-item" href="../articles/localmoran_landing.html">Local Moran's I</a></li>
    <li><a class="dropdown-item" href="../articles/losh_landing.html">Local spatial heteroscedasticity</a></li>
    <li><a class="dropdown-item" href="../articles/getisord_landing.html">Getis-Ord Gi*</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Bivariate</h6></li>
    <li><a class="dropdown-item" href="../articles/lee_landing.html">Lee's L</a></li>
    <li><a class="dropdown-item" href="../articles/crossvariogram_landing.html">Cross variogram</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Multivariate</h6></li>
    <li><a class="dropdown-item" href="../articles/multispati_landing.html">MULTISPATI PCA</a></li>
    <li><a class="dropdown-item" href="../articles/localc_landing.html">Multivariate local Geary's C</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/sfemethod">Custom methods</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/pachterlab/voyager/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>CODEX exploratory data analysis</h1>
                        <h4 data-toc-skip class="author">Kayla
Jackson</h4>
            
            <h4 data-toc-skip class="date">2024-11-23</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pachterlab/voyager/blob/documentation/vignettes/vig8_codex.Rmd" class="external-link"><code>vignettes/vig8_codex.Rmd</code></a></small>
      <div class="d-none name"><code>vig8_codex.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/voyager" class="external-link">Voyager</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drighelli/SpatialExperiment" class="external-link">SpatialExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pachterlab/SpatialFeatureExperiment" class="external-link">SpatialFeatureExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">batchelor</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MarioniLab/scran/" class="external-link">scran</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://glue.tidyverse.org/" class="external-link">glue</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/" class="external-link">purrr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://yutannihilation.github.io/gghighlight/" class="external-link">gghighlight</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/r-spatial/spdep/" class="external-link">spdep</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/sales-lab/spatialDE" class="external-link">spatialDE</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="dataset">Dataset<a class="anchor" aria-label="anchor" href="#dataset"></a>
</h2>
<p>The dataset used in this vignette is from the paper <a href="doi.org/10.3389/fimmu.2021.727626">Strategies for Accurate Cell
Type Identification in CODEX Multiplexed Imaging Data</a>(Hickey, et.al
2021). The data were collected as part of the HuBMap consortium which
seeks to characterize healthy human tissues and make data broadly
available. More specifically, this dataset characterizes 4 regions of
the large intestine (colon) from a single donor. This vignette will
focus on data from the sigmoid colon.</p>
<p>The intestinal sections were interrogated using the multiplexed
imaging method CO-Detection by indEXing (CODEX). CODEX involves cyclical
staining of a tissue with DNA-barcoded antibodies. At each round of
experimentation, fluoresently labeled probes hybridize to the tissue
bound DNA-conjugated antibodies are subsequently imaged and the stripped
from the tissue. At present, the technology quantifies up to 60 markers
in a single experiment. Raw images generated from this process are
subjected to image stitching, drift compensation, deconvolution, and
cycle concatenation using publicly avaialable <a href="https://github.com/nolanlab/CODEX" class="external-link">software</a>. The result of
this pre-processing is a matrix that contains the location of individual
cells and the quantified markers for each cell. Cell types were assigned
as described in the manuscript linked above. Briefly, the authors used a
hand-gating strategy to define cell types and create a standard to
compare the effect of normalization methods on clustering and cell
annotation.</p>
<p>The raw intensity data are available for download from HuBMAP with
identifier <a href="https://portal.hubmapconsortium.org/browse/dataset/e6fd525b837f4cf736c8af830f4f750f" class="external-link">HBM575.THQMM.284</a>
and the cell type annotations are provided as supplementary data in the
manuscript. The data relevant to this vignette have been converted to a
<code>SFE</code> object and are available to download <a href="https://caltech.box.com/public/static/zfr8l20450n2z28lnp0ugdj471ph9eyx" class="external-link">here</a>
from Box.</p>
<p>These data will be submitted to the <code>SFEData</code> package on
Bioconductor and will be available there in a future release.</p>
<p>We will begin by downloading the data and loading it in to R.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://caltech.box.com/public/static/zfr8l20450n2z28lnp0ugdj471ph9eyx"</span>,<span class="st">'./codex.Rds'</span>, mode<span class="op">=</span><span class="st">'wb'</span>, method <span class="op">=</span> <span class="st">'wget'</span>, quiet <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"./codex.Rds"</span><span class="op">)</span></span>
<span><span class="va">sfe</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 47 19724 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): protein</span></span>
<span><span class="co">#&gt; rownames(47): MUC2 SOX9 ... CD49a CD163</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(19724): 1 2 ... 182 184</span></span>
<span><span class="co">#&gt; colData names(9): cell_id cell_type ... fn sample_id</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : X Y</span></span>
<span><span class="co">#&gt; imgData names(0):</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; unit: full_res_image_pixels</span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: centroids (POINT) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; sample01:</span></span></code></pre></div>
<p>The rows in the count matrix correspond to the 47 barcoded genes
measured by CODEX. Additionally, the authors provide some metadata for
the cells, including the cell type.</p>
<p>It turns out the column names are not unique which will cause errors
in downstream analysis. We will update the column names below</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cellids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"{colData(sfe)$fn}_{colData(sfe)$cell_id}"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">cellids</span></span></code></pre></div>
<div class="section level3">
<h3 id="exploratory-data-analysis">Exploratory Data Analysis<a class="anchor" aria-label="anchor" href="#exploratory-data-analysis"></a>
</h3>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">celldensity</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotCellBin2D.html">plotCellBin2D</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span>
<span></span>
<span><span class="va">celldensity</span></span></code></pre></div>
<p><img src="vig8_codex_files/figure-html/unnamed-chunk-5-1.png" width="700" style="display: block; margin: auto;">
We can see from the figure above that the colonic epithelium is enriched
with cells while the loose connective tissue and muscle layers beneath
the epithelial layer are more sparsely populated. This is in line with
known colon histology. The epithelium is enriched with goblet cells and
has invaginations that project inwards towards the connective tissue.
Smooth muscle cells are also prominent in the colon, where bands of
muscle contract to move colonic contents towards the rectum.</p>
<p>We can visualize these cell types in space using the
<code><a href="../reference/plotSpatialFeature.html">plotSpatialFeature()</a></code> function. We will highlight Goblet and
smooth muscle cells to display their relative distribution in the
tissue. Since CODEX image processing relies on segmentation, each dot in
the plot represents a single cell. Here, each cell is represented by its
centroid, but can also be visualized as cell polygons in cases where the
segmentation mask is available.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, features<span class="op">=</span><span class="st">'cell_type'</span>, colGeometryName <span class="op">=</span> <span class="st">"centroids"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://yutannihilation.github.io/gghighlight/reference/gghighlight.html" class="external-link">gghighlight</a></span><span class="op">(</span><span class="va">cell_type</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Goblet"</span>, <span class="st">"SmoothMuscleME"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Tried to calculate with group_by(), but the calculation failed.</span></span>
<span><span class="co">#&gt; Falling back to ungrouped filter operation...</span></span>
<span></span>
<span><span class="va">spatial</span></span></code></pre></div>
<p><img src="vig8_codex_files/figure-html/unnamed-chunk-6-1.png" width="700" style="display: block; margin: auto;">
The goblet cells clearly define the epithelial border of the tissue and
the thick bands of smooth muscle cells are prominent below the
mucosa.</p>
<p>Next, we will compute some gene level metrics for each of the 47
barcoded genes. In contrast to RNA-based methods, the fields in the
matrix represent intensities rather than counts.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">mean</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">var</span> <span class="op">&lt;-</span> <span class="fu">rowVars</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">mean</span>, <span class="va">var</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig8_codex_files/figure-html/unnamed-chunk-7-1.png" width="700" style="display: block; margin: auto;">
There appears to be a sigmoid relationship between the mean and variance
of the protein expression. The pattern is reminiscent of what might be
expected if the intensity values were derived from a Gamma distribution,
the continuous analog of the Negative Binomial distribution that is
typically used to describe count data from scRNA-seq experiments. This
may have implications for how CODEX data is variance stabilized in the
future.</p>
<p>CODEX data is subject to noise from several sources including
segmentation artifacts, nonspecific staining, and imperfect tissue
processing. These are factors that can limit accurate quantification of
signal intensity and impede accurate cell annotation. The authors of the
dataset tested the effects of several normalization methods on cell type
annotation and clustering and found that Z-score normalization of each
marker resulted in accurate identification of both rare and common cell
types. In the cell below, we demonstrate how to accomplish this using
standard matrix operations. The normalized count matrix is typically
stored in the <code>logcounts</code> slot for scRNA-seq data, but we
will instead store the normalized matrix in a slot called
<code>normalizedIntensity</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mtx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">'protein'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">'normalizedIntensity'</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">mtx</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums-methods.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">mtx</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="fu">rowSds</span><span class="op">(</span><span class="va">mtx</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of length 2</span></span>
<span><span class="co">#&gt; names(2): protein normalizedIntensity</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="spatial-eda">Spatial EDA<a class="anchor" aria-label="anchor" href="#spatial-eda"></a>
</h2>
<p>Neighbor definition is a critical step in computation of metrics of
spatial dependency like Moran’s I and Geary’s C. The definition of
neighbors is complex, even when cell polygons are available. In the
latter case, the <code>poly2nb</code> method might be appropriate to
assign two cells as neighbors if they physically touch each other or
share a border. This may not be tenable in cases where cells are sparse
or cells are represented by their centroids, as in this dataset.</p>
<p>We will compute the spatial neighborhood graph using the
<code>knearestneigh</code> function as it is implemented in
<code>spdep</code>. In brief, Euclidean distances are computed between
each pair of cells and the <code>k</code> nearest cells are considered
neighbors. In the following code cell, we will consdier
<code>k=10</code> for speed purposes, but this may not be ideal in
general.</p>
<p>The weights of the neighborhood matrix are inverse-distance weighted,
such that the the weight of regions listed as neighbors increases as the
distance between pairs of points decreases. Setting
<code>style = "W"</code> ensures that the weights are row
standardized.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"knn10"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/findSpatialNeighbors-SpatialFeatureExperiment-method.html" class="external-link">findSpatialNeighbors</a></span><span class="op">(</span></span>
<span>    <span class="va">sfe</span>, method <span class="op">=</span> <span class="st">"knearneigh"</span>, dist_type <span class="op">=</span> <span class="st">"idw"</span>, </span>
<span>    k <span class="op">=</span> <span class="fl">10</span>, style <span class="op">=</span> <span class="st">"W"</span><span class="op">)</span></span></code></pre></div>
<p>The <code><a href="../reference/plotColGraph.html">plotColGraph()</a></code> function plots the graph in space
along with its corresponding <code>colGeometry</code>, but since there
are so many cells in this dataset, plotting the neighborhood graph may
not be as useful as many connections will be obscure by overlapping
lines. In any case, we will demonstrate use of the function below.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotColGraph.html">plotColGraph</a></span><span class="op">(</span><span class="va">sfe</span>, colGraphName <span class="op">=</span> <span class="st">"knn10"</span>, colGeometryName <span class="op">=</span> <span class="st">'centroids'</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig8_codex_files/figure-html/unnamed-chunk-10-1.png" width="700" style="display: block; margin: auto;">
Next, we will explore univariate metrics for global spatial
autocorrelation. Since few genes are quantified in this study, we will
compute the metrics for all genes. For larger datasets, it may be useful
to restrict analysis to the most variable genes.</p>
<p>We use the <code><a href="../reference/calculateUnivariate.html">runUnivariate()</a></code> function to compute the
spatial autocorrelation metrics and save the results in the SFE
object.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">runUnivariate</a></span><span class="op">(</span></span>
<span>    <span class="va">sfe</span>, type <span class="op">=</span> <span class="st">"moran.mc"</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span>,</span>
<span>    exprs_values <span class="op">=</span> <span class="st">"normalizedIntensity"</span>, colGraphName <span class="op">=</span> <span class="st">"knn10"</span>, nsim <span class="op">=</span> <span class="fl">100</span>,</span>
<span>    BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateUnivariate.html">runUnivariate</a></span><span class="op">(</span></span>
<span>    <span class="va">sfe</span>, type <span class="op">=</span> <span class="st">"moran.plot"</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span>,</span>
<span>    exprs_values <span class="op">=</span> <span class="st">"normalizedIntensity"</span>, colGraphName <span class="op">=</span> <span class="st">"knn10"</span><span class="op">)</span></span></code></pre></div>
<p>The results of these computations are accessible in the
<code>rowData</code> attribute of the SFE object.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "mean"                          "var"                          </span></span>
<span><span class="co">#&gt; [3] "moran.mc_statistic_sample01"   "moran.mc_parameter_sample01"  </span></span>
<span><span class="co">#&gt; [5] "moran.mc_p.value_sample01"     "moran.mc_alternative_sample01"</span></span>
<span><span class="co">#&gt; [7] "moran.mc_method_sample01"      "moran.mc_res_sample01"</span></span></code></pre></div>
<p>Next, we plot the results of the genes with the highest Moran’s I
statistic.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_moran</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">moran.mc_statistic_sample01</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">6</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">moran</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotMoranMC.html">plotMoranMC</a></span><span class="op">(</span><span class="va">sfe</span>, features <span class="op">=</span> <span class="va">top_moran</span>, facet_by <span class="op">=</span> <span class="st">'features'</span><span class="op">)</span></span>
<span><span class="va">moran</span></span></code></pre></div>
<p><img src="vig8_codex_files/figure-html/unnamed-chunk-13-1.png" width="700" style="display: block; margin: auto;">
The vertical line in each plot represents the observed Moran’s I while
the density represents the Moran’s I statistic for each of the random
permutations of the data. Each of these plots suggests that the Moran’s
I statistic is significant. We can plot the normalized intensity for
these genes in space.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatialFeature.html">plotSpatialFeature</a></span><span class="op">(</span></span>
<span>    <span class="va">sfe</span>, features<span class="op">=</span><span class="va">top_moran</span>, colGeometryName <span class="op">=</span> <span class="st">"centroids"</span>,</span>
<span>    exprs_values <span class="op">=</span> <span class="st">"normalizedIntensity"</span>, scattermore <span class="op">=</span> <span class="cn">TRUE</span>, pointsize <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig8_codex_files/figure-html/unnamed-chunk-14-1.png" width="1152" style="display: block; margin: auto;">
While most of these genes appear to have some spatial distribution, it
also seems that it may overlap with cell type. The cells that appear to
express the genes of interest seem to be spatially restricted to known
boundaries in the tissue.</p>
<p>The <code><a href="../reference/moranPlot.html">moranPlot()</a></code> function plots spatial data against its
spatially lagged values and enables users to assess how similar observed
values are to its neighbors. When the variable is centered, the plot is
divided into four quadrants defined by the horizontal line y = 0 and the
vertical line x = 0. Points in the upper right (or high-high) and lower
left (or low-low) quadrants indicate positive spatial association, and
points in the lower right (or high-low) and upper left (or low-high)
quadrants include observations that exhibit negative spatial
association.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/moranPlot.html">moranPlot</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">top_moran</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig8_codex_files/figure-html/unnamed-chunk-15-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="differential-expression">Differential Expression<a class="anchor" aria-label="anchor" href="#differential-expression"></a>
</h2>
<p>While Moran’s I and other global spatial autocorrelation metrics
provide insight to the spatial patterns of gene expression, it is
necessarily limited by the structure imposed by the spatial weights
matrix. A complimentary task might be to identify spatially variable
(SV) genes. One such method to do this is described in <a href="">SpatialDE: identification of spatially variable genes</a>. The
method described in the manuscript relies on Gaussian process regression
and decomposes variability in expression into spatial and non-spatial
components. In contrast to Moran’s I, the covariance between each pair
of cells is modeled as a function of the distance between them. Notably,
it does not require an explicit specification of hte neighborhood graph,
but rather the a parameter controls the decay in covariance as distance
increases.</p>
<p>The <code>spatialDE</code> package is implemented in R and requires a
normalized matrix as input. The <code><a href="https://rdrr.io/pkg/spatialDE/man/spatialDE.html" class="external-link">spatialDE()</a></code> function from
the package performs normalization steps before running the algorithm.
Because the data has already been normalized, we will use the
<code><a href="https://rdrr.io/pkg/spatialDE/man/run.html" class="external-link">run()</a></code> function directly to run spatialDE. We will first
have to convert the centroid coordinates to a data frame as required by
the function.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Store coordinates in a data frame object</span></span>
<span><span class="va">coords</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialFeatureExperiment/man/colGeometries.html" class="external-link">centroids</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">geometry</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map_dfr.html" class="external-link">map_dfr</a></span><span class="op">(</span>\<span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, y <span class="op">=</span> <span class="va">x</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># de_res &lt;- spatialDE::run(assay(sfe,"normalizedIntensity"), coords, verbose=TRUE)</span></span></code></pre></div>
<p>We can plot the normalized expression of the top 5 genes in
space.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># top_genes &lt;- de_res |&gt;</span></span>
<span><span class="co">#   arrange(pval) |&gt;</span></span>
<span><span class="co">#   slice_head(n=6) |&gt;</span></span>
<span><span class="co">#   pull(g)</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># plotSpatialFeature(sfe, top_genes, colGeometryName="centroids",</span></span>
<span><span class="co">#                    exprs_values = "normalizedIntensity")</span></span></code></pre></div>
<p>Perhaps unsurprisingly, the expression of the top DE genes seems to
highlight the spatial distribution of known cell types in the tissue
rather than identify spatially restricted gene expression. This is
related to the experimental design where the targeted genes were chosen
to differentiate cell types. Perhaps in genome-wide technologies, the
potential for discovery of neew gene expression patterns is more
plausible. There is an open question as to whether these results offer
new information compared to what is inferred by typical DE expression
methods.</p>
<p>These analyses represent a minority of the types of inferences that
can be made from protein expression data. It would be interested to
investigate how the protein expression results compare or inform data
from spatail scRNA-sequencing experiments. Already, there is work being
done to obtain multimodal spatial measurements on the same sample.
Importantly however, considerations should be made on the types of
biases each individual technology adds to the measurements. These are
active areas of research that are ripe for future exploration.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.2 (2024-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.5 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] BiocParallel_1.40.0            spatialDE_1.12.0              </span></span>
<span><span class="co">#&gt;  [3] spdep_1.3-6                    sf_1.0-19                     </span></span>
<span><span class="co">#&gt;  [5] spData_2.3.3                   patchwork_1.3.0               </span></span>
<span><span class="co">#&gt;  [7] gghighlight_0.4.1              dplyr_1.1.4                   </span></span>
<span><span class="co">#&gt;  [9] tidyr_1.3.1                    purrr_1.0.2                   </span></span>
<span><span class="co">#&gt; [11] glue_1.8.0                     bluster_1.16.0                </span></span>
<span><span class="co">#&gt; [13] scran_1.34.0                   scater_1.34.0                 </span></span>
<span><span class="co">#&gt; [15] ggplot2_3.5.1                  scuttle_1.16.0                </span></span>
<span><span class="co">#&gt; [17] batchelor_1.22.0               SpatialExperiment_1.16.0      </span></span>
<span><span class="co">#&gt; [19] SingleCellExperiment_1.28.1    SummarizedExperiment_1.36.0   </span></span>
<span><span class="co">#&gt; [21] Biobase_2.66.0                 GenomicRanges_1.58.0          </span></span>
<span><span class="co">#&gt; [23] GenomeInfoDb_1.42.0            IRanges_2.40.0                </span></span>
<span><span class="co">#&gt; [25] S4Vectors_0.44.0               BiocGenerics_0.52.0           </span></span>
<span><span class="co">#&gt; [27] MatrixGenerics_1.18.0          matrixStats_1.4.1             </span></span>
<span><span class="co">#&gt; [29] Voyager_1.8.1                  SpatialFeatureExperiment_1.9.4</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] splines_4.4.2             filelock_1.0.3           </span></span>
<span><span class="co">#&gt;   [3] bitops_1.0-9              tibble_3.2.1             </span></span>
<span><span class="co">#&gt;   [5] R.oo_1.27.0               basilisk.utils_1.18.0    </span></span>
<span><span class="co">#&gt;   [7] lifecycle_1.0.4           edgeR_4.4.0              </span></span>
<span><span class="co">#&gt;   [9] lattice_0.22-6            MASS_7.3-61              </span></span>
<span><span class="co">#&gt;  [11] backports_1.5.0           magrittr_2.0.3           </span></span>
<span><span class="co">#&gt;  [13] limma_3.62.1              sass_0.4.9               </span></span>
<span><span class="co">#&gt;  [15] rmarkdown_2.29            jquerylib_0.1.4          </span></span>
<span><span class="co">#&gt;  [17] yaml_2.3.10               metapod_1.14.0           </span></span>
<span><span class="co">#&gt;  [19] sp_2.1-4                  reticulate_1.40.0        </span></span>
<span><span class="co">#&gt;  [21] RColorBrewer_1.1-3        DBI_1.2.3                </span></span>
<span><span class="co">#&gt;  [23] ResidualMatrix_1.16.0     multcomp_1.4-26          </span></span>
<span><span class="co">#&gt;  [25] abind_1.4-8               spatialreg_1.3-5         </span></span>
<span><span class="co">#&gt;  [27] zlibbioc_1.52.0           R.utils_2.12.3           </span></span>
<span><span class="co">#&gt;  [29] RCurl_1.98-1.16           TH.data_1.1-2            </span></span>
<span><span class="co">#&gt;  [31] sandwich_3.1-1            GenomeInfoDbData_1.2.13  </span></span>
<span><span class="co">#&gt;  [33] ggrepel_0.9.6             irlba_2.3.5.1            </span></span>
<span><span class="co">#&gt;  [35] terra_1.7-83              units_0.8-5              </span></span>
<span><span class="co">#&gt;  [37] RSpectra_0.16-2           dqrng_0.4.1              </span></span>
<span><span class="co">#&gt;  [39] pkgdown_2.1.1             DelayedMatrixStats_1.28.0</span></span>
<span><span class="co">#&gt;  [41] codetools_0.2-20          DropletUtils_1.26.0      </span></span>
<span><span class="co">#&gt;  [43] DelayedArray_0.32.0       tidyselect_1.2.1         </span></span>
<span><span class="co">#&gt;  [45] UCSC.utils_1.2.0          memuse_4.2-3             </span></span>
<span><span class="co">#&gt;  [47] farver_2.1.2              ScaledMatrix_1.14.0      </span></span>
<span><span class="co">#&gt;  [49] viridis_0.6.5             jsonlite_1.8.9           </span></span>
<span><span class="co">#&gt;  [51] BiocNeighbors_2.0.0       e1071_1.7-16             </span></span>
<span><span class="co">#&gt;  [53] survival_3.7-0            systemfonts_1.1.0        </span></span>
<span><span class="co">#&gt;  [55] tools_4.4.2               ggnewscale_0.5.0         </span></span>
<span><span class="co">#&gt;  [57] ragg_1.3.3                Rcpp_1.0.13-1            </span></span>
<span><span class="co">#&gt;  [59] gridExtra_2.3             SparseArray_1.6.0        </span></span>
<span><span class="co">#&gt;  [61] mgcv_1.9-1                xfun_0.49                </span></span>
<span><span class="co">#&gt;  [63] EBImage_4.48.0            HDF5Array_1.34.0         </span></span>
<span><span class="co">#&gt;  [65] withr_3.0.2               fastmap_1.2.0            </span></span>
<span><span class="co">#&gt;  [67] basilisk_1.18.0           boot_1.3-31              </span></span>
<span><span class="co">#&gt;  [69] rhdf5filters_1.18.0       fansi_1.0.6              </span></span>
<span><span class="co">#&gt;  [71] digest_0.6.37             rsvd_1.0.5               </span></span>
<span><span class="co">#&gt;  [73] R6_2.5.1                  textshaping_0.4.0        </span></span>
<span><span class="co">#&gt;  [75] colorspace_2.1-1          wk_0.9.4                 </span></span>
<span><span class="co">#&gt;  [77] scattermore_1.2           LearnBayes_2.15.1        </span></span>
<span><span class="co">#&gt;  [79] jpeg_0.1-10               R.methodsS3_1.8.2        </span></span>
<span><span class="co">#&gt;  [81] utf8_1.2.4                generics_0.1.3           </span></span>
<span><span class="co">#&gt;  [83] data.table_1.16.2         class_7.3-22             </span></span>
<span><span class="co">#&gt;  [85] httr_1.4.7                htmlwidgets_1.6.4        </span></span>
<span><span class="co">#&gt;  [87] S4Arrays_1.6.0            pkgconfig_2.0.3          </span></span>
<span><span class="co">#&gt;  [89] scico_1.5.0               gtable_0.3.6             </span></span>
<span><span class="co">#&gt;  [91] XVector_0.46.0            htmltools_0.5.8.1        </span></span>
<span><span class="co">#&gt;  [93] fftwtools_0.9-11          scales_1.3.0             </span></span>
<span><span class="co">#&gt;  [95] png_0.1-8                 knitr_1.49               </span></span>
<span><span class="co">#&gt;  [97] rjson_0.2.23              checkmate_2.3.2          </span></span>
<span><span class="co">#&gt;  [99] coda_0.19-4.1             nlme_3.1-166             </span></span>
<span><span class="co">#&gt; [101] proxy_0.4-27              cachem_1.1.0             </span></span>
<span><span class="co">#&gt; [103] zoo_1.8-12                rhdf5_2.50.0             </span></span>
<span><span class="co">#&gt; [105] KernSmooth_2.23-24        parallel_4.4.2           </span></span>
<span><span class="co">#&gt; [107] vipor_0.4.7               desc_1.4.3               </span></span>
<span><span class="co">#&gt; [109] s2_1.1.7                  pillar_1.9.0             </span></span>
<span><span class="co">#&gt; [111] grid_4.4.2                vctrs_0.6.5              </span></span>
<span><span class="co">#&gt; [113] BiocSingular_1.22.0       beachmat_2.22.0          </span></span>
<span><span class="co">#&gt; [115] sfheaders_0.4.4           cluster_2.1.6            </span></span>
<span><span class="co">#&gt; [117] beeswarm_0.4.0            evaluate_1.0.1           </span></span>
<span><span class="co">#&gt; [119] isoband_0.2.7             zeallot_0.1.0            </span></span>
<span><span class="co">#&gt; [121] magick_2.8.5              mvtnorm_1.3-2            </span></span>
<span><span class="co">#&gt; [123] cli_3.6.3                 locfit_1.5-9.10          </span></span>
<span><span class="co">#&gt; [125] compiler_4.4.2            rlang_1.1.4              </span></span>
<span><span class="co">#&gt; [127] crayon_1.5.3              labeling_0.4.3           </span></span>
<span><span class="co">#&gt; [129] classInt_0.4-10           fs_1.6.5                 </span></span>
<span><span class="co">#&gt; [131] ggbeeswarm_0.7.2          viridisLite_0.4.2        </span></span>
<span><span class="co">#&gt; [133] deldir_2.0-4              munsell_0.5.1            </span></span>
<span><span class="co">#&gt; [135] tiff_0.1-12               Matrix_1.7-1             </span></span>
<span><span class="co">#&gt; [137] dir.expiry_1.14.0         sparseMatrixStats_1.18.0 </span></span>
<span><span class="co">#&gt; [139] Rhdf5lib_1.28.0           statmod_1.5.0            </span></span>
<span><span class="co">#&gt; [141] igraph_2.1.1              bslib_0.8.0</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Lambda Moses, Alik Huseynov, Kayla Jackson, Laura Luebbert, Sina Booeshaghi, Lior Pachter.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
